

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  

  <title>scirpy.tl.define_clonotype_clusters &mdash; scirpy  documentation</title>
  


  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/scanpy.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/typehints.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="scirpy.tl.clonotype_convergence" href="scirpy.tl.clonotype_convergence.html" />
    <link rel="prev" title="scirpy.tl.define_clonotypes" href="scirpy.tl.define_clonotypes.html" />
 


</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.9.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage-principles.html">Usage principles</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.io">Input/Output: <code class="docutils literal notranslate"><span class="pre">io</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.pp">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../api.html#tools-tl">Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code></a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../api.html#generic">Generic</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api.html#quality-control">Quality control</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../api.html#define-and-visualize-clonotypes">Define and visualize clonotypes</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="scirpy.tl.define_clonotypes.html">scirpy.tl.define_clonotypes</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">scirpy.tl.define_clonotype_clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="scirpy.tl.clonotype_convergence.html">scirpy.tl.clonotype_convergence</a></li>
<li class="toctree-l4"><a class="reference internal" href="scirpy.tl.clonotype_network.html">scirpy.tl.clonotype_network</a></li>
<li class="toctree-l4"><a class="reference internal" href="scirpy.tl.clonotype_network_igraph.html">scirpy.tl.clonotype_network_igraph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../api.html#analyse-clonal-diversity">Analyse clonal diversity</a></li>
<li class="toctree-l3"><a class="reference internal" href="../api.html#v-d-j-gene-usage">V(D)J gene usage</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.pl">Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.datasets">Datasets: <code class="docutils literal notranslate"><span class="pre">datasets</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.util">Utility functions: <code class="docutils literal notranslate"><span class="pre">util</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="../api.html#module-scirpy.ir_dist">IR distance utilities: <code class="docutils literal notranslate"><span class="pre">ir_dist</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zzz_bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">API</a> &raquo;</li>
        
      <li>scirpy.tl.define_clonotype_clusters</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <!-- User defined GitHub URL -->
              <a href="https://github.com/icbi-lab/scirpy/tree/master/scirpy/_tools/_clonotypes.py#L167-L315" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="scirpy-tl-define-clonotype-clusters">
<h1>scirpy.tl.define_clonotype_clusters<a class="headerlink" href="#scirpy-tl-define-clonotype-clusters" title="Permalink to this headline">¶</a></h1>
<dl class="py function">
<dt id="scirpy.tl.define_clonotype_clusters">
<code class="sig-prename descclassname">scirpy.tl.</code><code class="sig-name descname">define_clonotype_clusters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">adata</span></em>, <em class="sig-param"><span class="o">*</span></em>, <em class="sig-param"><span class="n">sequence</span><span class="o">=</span><span class="default_value">'aa'</span></em>, <em class="sig-param"><span class="n">metric</span><span class="o">=</span><span class="default_value">'identity'</span></em>, <em class="sig-param"><span class="n">receptor_arms</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">dual_ir</span><span class="o">=</span><span class="default_value">'any'</span></em>, <em class="sig-param"><span class="n">same_v_gene</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">within_group</span><span class="o">=</span><span class="default_value">'receptor_type'</span></em>, <em class="sig-param"><span class="n">key_added</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">partitions</span><span class="o">=</span><span class="default_value">'connected'</span></em>, <em class="sig-param"><span class="n">resolution</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">n_iterations</span><span class="o">=</span><span class="default_value">5</span></em>, <em class="sig-param"><span class="n">distance_key</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">chunksize</span><span class="o">=</span><span class="default_value">2000</span></em><span class="sig-paren">)</span><a class="headerlink" href="#scirpy.tl.define_clonotype_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Define <a class="reference internal" href="../glossary.html#term-Clonotype-cluster"><span class="xref std std-term">clonotype clusters</span></a>.</p>
<p>As opposed to <a class="reference internal" href="scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">define_clonotypes()</span></code></a> which employs a more stringent
definition of <a class="reference internal" href="../glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a>, this function flexibly defines
clonotype clusters based on amino acid or nucleic acid sequence identity or
similarity.</p>
<p>Requires running <a class="reference internal" href="scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-func docutils literal notranslate"><span class="pre">ir_dist()</span></code></a> with the same <code class="docutils literal notranslate"><span class="pre">sequence</span></code> and
<code class="docutils literal notranslate"><span class="pre">metric</span></code> values first.</p>
<dl class="simple">
<dt>Definition of clonotype(-clusters) follows roughly the following procedure:</dt><dd><ol class="arabic simple">
<li><p>Create a list of unique receptor configurations. This is useful to
collapse heavily expanded clonotypes, leading to many cells with identical
CDR3 sequences, to a single entry.</p></li>
<li><p>Compute a pairwise distance matrix of unique receptor configurations.
Unique receptor configurations are matched based on the pre-computed
VJ and VDJ distance matrices and the parameters of <code class="docutils literal notranslate"><span class="pre">receptor_arms</span></code>,
<code class="docutils literal notranslate"><span class="pre">dual_ir</span></code>, <code class="docutils literal notranslate"><span class="pre">same_v_gene</span></code> and <code class="docutils literal notranslate"><span class="pre">within_group</span></code>.</p></li>
<li><p>Find connected modules in the graph defined by this distance matrix. Each
connected module is considered a clonotype-cluster.</p></li>
</ol>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>adata</strong> : <span class="annotation terse"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.7.dev1+g33006dc)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span><span class="annotation full"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.7.dev1+g33006dc)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span></dt><dd><p>Annotated data matrix</p>
</dd>
<dt><strong>sequence</strong> : <span class="annotation terse">{‘aa’, ‘nt’}</span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[‘aa’, ‘nt’]</span> (default: <code class="docutils literal notranslate"><span class="pre">'aa'</span></code>)</dt><dd><p>The sequence parameter used when running :func:scirpy.pp.ir_dist`</p>
</dd>
<dt><strong>metric</strong> : <span class="annotation terse">{‘alignment’, ‘identity’, ‘levenshtein’, ‘hamming’} | <a class="reference internal" href="scirpy.ir_dist.metrics.DistanceCalculator.html#scirpy.ir_dist.metrics.DistanceCalculator" title="scirpy.ir_dist.metrics.DistanceCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">DistanceCalculator</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[‘alignment’, ‘identity’, ‘levenshtein’, ‘hamming’], <a class="reference internal" href="scirpy.ir_dist.metrics.DistanceCalculator.html#scirpy.ir_dist.metrics.DistanceCalculator" title="scirpy.ir_dist.metrics.DistanceCalculator"><code class="xref py py-class docutils literal notranslate"><span class="pre">DistanceCalculator</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">'identity'</span></code>)</dt><dd><p>The metric parameter used when running <a class="reference internal" href="scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pp.ir_dist()</span></code></a></p>
</dd>
<dt><strong>receptor_arms</strong> : <span class="annotation terse">{‘VJ’, ‘VDJ’, ‘all’, ‘any’}</span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[‘VJ’, ‘VDJ’, ‘all’, ‘any’]</span> (default: <code class="docutils literal notranslate"><span class="pre">'all'</span></code>)</dt><dd><p><dl class="simple">
<dt>One of the following options:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;VJ&quot;</span></code> - only consider <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ</span></a> sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;VDJ&quot;</span></code> - only consider <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VDJ</span></a> sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> - both VJ and VDJ need to match</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;any&quot;</span></code> - either VJ or VDJ need to match</p></li>
</ul>
</dd>
</dl>
<p>If <code class="docutils literal notranslate"><span class="pre">&quot;any&quot;</span></code>, two distances are combined by taking their minimum. If <code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code>,
two distances are combined by taking their maximum. This is motivated
by the hypothesis that a receptor recognizes the same antigen if it
has a distance smaller than a certain cutoff. If we require only one
of the receptors to match (<code class="docutils literal notranslate"><span class="pre">&quot;any&quot;</span></code>) the smaller distance is relevant.
If we require both receptors to match (<code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code>), the larger distance is
relevant.</p>
</p>
</dd>
<dt><strong>dual_ir</strong> : <span class="annotation terse">{‘primary_only’, ‘all’, ‘any’}</span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Literal</span></code></a>[‘primary_only’, ‘all’, ‘any’]</span> (default: <code class="docutils literal notranslate"><span class="pre">'any'</span></code>)</dt><dd><p><dl class="simple">
<dt>One of the following options:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;primary_only&quot;</span></code> - only consider most abundant pair of <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ/VDJ</span></a> chains</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;any&quot;</span></code> - consider both pairs of <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ/VDJ</span></a> sequences. Distance must be below
cutoff for any of the chains.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> - consider both pairs of <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ/VDJ</span></a> sequences. Distance must be below
cutoff for all of the chains.</p></li>
</ul>
</dd>
</dl>
<p>Distances are combined as for <code class="docutils literal notranslate"><span class="pre">receptor_arms</span></code>.</p>
<p>See also <a class="reference internal" href="../glossary.html#term-Dual-IR"><span class="xref std std-term">Dual IR</span></a>.</p>
</p>
</dd>
<dt><strong>same_v_gene</strong> : <span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a></span> (default: <code class="docutils literal notranslate"><span class="pre">False</span></code>)</dt><dd><p><p>Enforces clonotypes to have the same <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">V-genes</span></a>. This is useful
as the CDR1 and CDR2 regions are fully encoded in this gene.
See <a class="reference internal" href="../glossary.html#term-CDR"><span class="xref std std-term">CDR</span></a> for more details.</p>
<p>v genes are matched based on the behaviour defined with <code class="docutils literal notranslate"><span class="pre">receptor_arms</span></code> and
<code class="docutils literal notranslate"><span class="pre">dual_ir</span></code>.</p>
</p>
</dd>
<dt><strong>within_group</strong> : <span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>] | <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a> | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Union" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>], <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a>]</span> (default: <code class="docutils literal notranslate"><span class="pre">'receptor_type'</span></code>)</dt><dd><p>Enforces clonotypes to have the same group defined by one or multiple grouping
variables. Per default, this is set to <a class="reference internal" href="../glossary.html#term-Receptor-type"><span class="xref std std-term">receptor_type</span></a>,
i.e. clonotypes cannot comprise both B cells and T cells. Set this to
<a class="reference internal" href="../glossary.html#term-Receptor-subtype"><span class="xref std std-term">receptor_subtype</span></a> if you don’t want clonotypes to
be shared across e.g. gamma-delta and alpha-beta T-cells.
You can also set this to any other column in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> that contains
a grouping, or to <code class="docutils literal notranslate"><span class="pre">None</span></code>, if you want no constraints.</p>
</dd>
</dl>
</dd>
</dl>
<dl>
<dt>key_added</dt><dd><p>The column name under which the clonotype clusters and cluster sizes
will be stored in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> and under which the clonotype network will be
stored in <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code>.</p>
<blockquote>
<div><ul class="simple">
<li><p>Defaults to <code class="docutils literal notranslate"><span class="pre">cc_{sequence}_{metric}</span></code>, e.g. <code class="docutils literal notranslate"><span class="pre">cc_aa_levenshtein</span></code>,
where <code class="docutils literal notranslate"><span class="pre">cc</span></code> stands for “clonotype cluster”.</p></li>
<li><p>The clonotype sizes will be stored in <code class="docutils literal notranslate"><span class="pre">{key_added}_size</span></code>,
e.g. <code class="docutils literal notranslate"><span class="pre">cc_aa_levenshtein_size</span></code>.</p></li>
<li><p>The clonotype x clonotype network will be stored in <code class="docutils literal notranslate"><span class="pre">{key_added}_dist</span></code>,
e.g. <code class="docutils literal notranslate"><span class="pre">cc_aa_levenshtein_dist</span></code>.</p></li>
</ul>
</div></blockquote>
</dd>
<dt>partitions</dt><dd><p>How to find graph partitions that define a clonotype.
Possible values are <code class="docutils literal notranslate"><span class="pre">leiden</span></code>, for using the “Leiden” algorithm and
<code class="docutils literal notranslate"><span class="pre">connected</span></code> to find fully connected sub-graphs.</p>
<p>The difference is that the Leiden algorithm further divides
fully connected subgraphs into highly-connected modules.</p>
</dd>
<dt>resolution</dt><dd><p><code class="docutils literal notranslate"><span class="pre">resolution</span></code> parameter for the leiden algorithm.</p>
</dd>
<dt>n_iterations</dt><dd><p><code class="docutils literal notranslate"><span class="pre">n_iterations</span></code> parameter for the leiden algorithm.</p>
</dd>
<dt>distance_key</dt><dd><p>Key in <code class="docutils literal notranslate"><span class="pre">adata.uns</span></code> where the sequence distances are stored. This defaults
to <code class="docutils literal notranslate"><span class="pre">ir_dist_{sequence}_{metric}</span></code>.</p>
</dd>
<dt>inplace</dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, adds the results to anndata, otherwise returns them.</p>
</dd>
<dt>n_jobs</dt><dd><p>Number of CPUs to use for clonotype cluster calculation. Default: use all cores.
If the number of cells is smaller than <code class="docutils literal notranslate"><span class="pre">2</span> <span class="pre">*</span> <span class="pre">chunksize</span></code> a single
worker thread will be used to avoid overhead.</p>
</dd>
<dt>chunksize</dt><dd><p>Number of objects to process per chunk. Each worker thread receives
data in chunks. Smaller chunks result in a more meaningful progressbar,
but more overhead.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a>, <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>] | <a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Optional" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.9)"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code></a>[<a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a>, <a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a>]]</span></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl>
<dt>clonotype<span class="classifier"><span class="annotation terse"><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a></span><span class="annotation full"><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a></span></span></dt><dd><p>A Series containing the clonotype id for each cell. Will be stored in
<code class="docutils literal notranslate"><span class="pre">adata.obs[key_added]</span></code> if <code class="docutils literal notranslate"><span class="pre">inplace</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code></p>
</dd>
<dt>clonotype_size<span class="classifier"><span class="annotation terse"><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a></span><span class="annotation full"><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v1.3.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a></span></span></dt><dd><p>A Series containing the number of cells in the respective clonotype
for each cell. Will be stored in <code class="docutils literal notranslate"><span class="pre">adata.obs[f&quot;{key_added}_size&quot;]</span></code> if <code class="docutils literal notranslate"><span class="pre">inplace</span></code>
is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
</dd>
<dt>distance_result<span class="classifier"><span class="annotation terse"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span><span class="annotation full"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code></a></span></span></dt><dd><dl class="simple">
<dt>A dictionary containing</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">distances</span></code>: A sparse, pairwise distance matrix between unique
receptor configurations</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cell_indices</span></code>: A dict of arrays, containing the adata.obs_names
(cell indices) for each row in the distance matrix.</p></li>
</ul>
</dd>
</dl>
<p>If <code class="docutils literal notranslate"><span class="pre">inplace</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, this is added to <code class="docutils literal notranslate"><span class="pre">adata.uns[key_added]</span></code>.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="scirpy.tl.clonotype_convergence.html" class="btn btn-neutral float-right" title="scirpy.tl.clonotype_convergence" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="scirpy.tl.define_clonotypes.html" class="btn btn-neutral float-left" title="scirpy.tl.define_clonotypes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2021, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>