<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Community Ecology Package</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Community Ecology Package
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;vegan&rsquo; version 2.5-5</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="# "> </a>
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#H">H</a>
<a href="#I">I</a>
<a href="#K">K</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#O">O</a>
<a href="#P">P</a>
<a href="#Q">Q</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#V">V</a>
<a href="#W">W</a>
</p>

<table width="100%">
<tr><td style="width: 25%;"><a href="vegan-package.html">vegan-package</a></td>
<td>Community Ecology Package: Ordination, Diversity and Dissimilarities</td></tr>
</table>

<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="add1.cca.html">add1.cca</a></td>
<td>Add or Drop Single Terms to a Constrained Ordination Model</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">adipart</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">adipart.default</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">adipart.formula</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="adonis.html">adonis</a></td>
<td>Permutational Multivariate Analysis of Variance Using Distance Matrices</td></tr>
<tr><td style="width: 25%;"><a href="adonis.html">adonis2</a></td>
<td>Permutational Multivariate Analysis of Variance Using Distance Matrices</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">AIC.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">AIC.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">AIC.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">alias.cca</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="anosim.html">anosim</a></td>
<td>Analysis of Similarities</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">anova.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="anova.cca.html">anova.cca</a></td>
<td>Permutation Test for Constrained Correspondence Analysis, Redundancy Analysis and Constrained Analysis of Principal Coordinates</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">as.fisher</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">as.hclust.spantree</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="oecosimu.html">as.mcmc.oecosimu</a></td>
<td>Evaluate Statistics with Null Models of Biological Communities</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">as.mcmc.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="vegan-deprecated.html">as.mlm</a></td>
<td>Deprecated Functions in vegan package</td></tr>
<tr><td style="width: 25%;"><a href="vegan-deprecated.html">as.mlm.cca</a></td>
<td>Deprecated Functions in vegan package</td></tr>
<tr><td style="width: 25%;"><a href="vegan-deprecated.html">as.mlm.rda</a></td>
<td>Deprecated Functions in vegan package</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">as.preston</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">as.rad</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="oecosimu.html">as.ts.oecosimu</a></td>
<td>Evaluate Statistics with Null Models of Biological Communities</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">as.ts.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="avgdist.html">avgdist</a></td>
<td>Averaged Subsampled Dissimilarity Matrices</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="BCI.html">BCI</a></td>
<td>Barro Colorado Island Tree Counts</td></tr>
<tr><td style="width: 25%;"><a href="BCI.html">BCI.env</a></td>
<td>Barro Colorado Island Tree Counts</td></tr>
<tr><td style="width: 25%;"><a href="beals.html">beals</a></td>
<td>Beals Smoothing and Degree of Absence</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="betadiver.html">betadiver</a></td>
<td>Indices of beta Diversity</td></tr>
<tr><td style="width: 25%;"><a href="bgdispersal.html">bgdispersal</a></td>
<td>Coefficients of Biogeographical Dispersal Direction</td></tr>
<tr><td style="width: 25%;"><a href="bioenv.html">bioenv</a></td>
<td>Best Subset of Environmental Variables with Maximum (Rank) Correlation with Community Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="bioenv.html">bioenv.default</a></td>
<td>Best Subset of Environmental Variables with Maximum (Rank) Correlation with Community Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="bioenv.html">bioenv.formula</a></td>
<td>Best Subset of Environmental Variables with Maximum (Rank) Correlation with Community Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="bioenv.html">bioenvdist</a></td>
<td>Best Subset of Environmental Variables with Maximum (Rank) Correlation with Community Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="biplot.rda.html">biplot.cca</a></td>
<td>PCA biplot</td></tr>
<tr><td style="width: 25%;"><a href="CCorA.html">biplot.CCorA</a></td>
<td>Canonical Correlation Analysis</td></tr>
<tr><td style="width: 25%;"><a href="biplot.rda.html">biplot.rda</a></td>
<td>PCA biplot</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">boxplot.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">boxplot.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">boxplot.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick.cca</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick.decorana</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick.default</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick.prcomp</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">bstick.princomp</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="permustats.html">c.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">calibrate</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">calibrate.cca</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="ordisurf.html">calibrate.ordisurf</a></td>
<td>Fit and Plot Smooth Surfaces of Variables on Ordination.</td></tr>
<tr><td style="width: 25%;"><a href="capscale.html">capscale</a></td>
<td>[Partial] Distance-based Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cascadeKM.html">cascadeKM</a></td>
<td>K-means partitioning using a range of values of K</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">cca</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">cca.default</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">cca.formula</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">cca.object</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="CCorA.html">CCorA</a></td>
<td>Canonical Correlation Analysis</td></tr>
<tr><td style="width: 25%;"><a href="designdist.html">chaodist</a></td>
<td>Design your own Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="cascadeKM.html">cIndexKM</a></td>
<td>K-means partitioning using a range of values of K</td></tr>
<tr><td style="width: 25%;"><a href="clamtest.html">clamtest</a></td>
<td>Multinomial Species Classification Method (CLAM)</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">coef.cca</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">coef.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">coef.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">coef.rda</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="commsim.html">commsim</a></td>
<td>Create an Object for Null Model Algorithms</td></tr>
<tr><td style="width: 25%;"><a href="MOStest.html">confint.MOStest</a></td>
<td>Mitchell-Olds &amp; Shaw Test for the Location of Quadratic Extreme</td></tr>
<tr><td style="width: 25%;"><a href="contribdiv.html">contribdiv</a></td>
<td>Contribution Diversity Approach</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">cooks.distance.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">cophenetic.spantree</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="vegemite.html">coverscale</a></td>
<td>Display Compact Ordered Community Tables</td></tr>
<tr><td style="width: 25%;"><a href="reorder.hclust.html">cutreeord</a></td>
<td>Reorder a Hierarchical Clustering Tree</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="capscale.html">dbrda</a></td>
<td>[Partial] Distance-based Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="decostand.html">decostand</a></td>
<td>Standardization Methods for Community Ecology</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">density.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">densityplot.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="designdist.html">designdist</a></td>
<td>Design your own Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="deviance.cca.html">deviance.cca</a></td>
<td>Statistics Resembling Deviance and AIC for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">deviance.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">deviance.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">deviance.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="deviance.cca.html">deviance.rda</a></td>
<td>Statistics Resembling Deviance and AIC for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">df.residual.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="dispindmorisita.html">dispindmorisita</a></td>
<td>Morisita index of intraspecific aggregation</td></tr>
<tr><td style="width: 25%;"><a href="dispweight.html">dispweight</a></td>
<td>Dispersion-based weighting of species counts</td></tr>
<tr><td style="width: 25%;"><a href="distconnected.html">distconnected</a></td>
<td>Connectedness of Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="diversity.html">diversity</a></td>
<td>Ecological Diversity Indices</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">downweight</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="rarefy.html">drarefy</a></td>
<td>Rarefaction Species Richness</td></tr>
<tr><td style="width: 25%;"><a href="add1.cca.html">drop1.cca</a></td>
<td>Add or Drop Single Terms to a Constrained Ordination Model</td></tr>
<tr><td style="width: 25%;"><a href="dune.html">dune</a></td>
<td>Vegetation and Environment in Dutch Dune Meadows.</td></tr>
<tr><td style="width: 25%;"><a href="dune.html">dune.env</a></td>
<td>Vegetation and Environment in Dutch Dune Meadows.</td></tr>
<tr><td style="width: 25%;"><a href="dune.taxon.html">dune.phylodis</a></td>
<td>Taxonomic Classification and Phylogeny of Dune Meadow Species</td></tr>
<tr><td style="width: 25%;"><a href="dune.taxon.html">dune.taxon</a></td>
<td>Taxonomic Classification and Phylogeny of Dune Meadow Species</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="wascores.html">eigengrad</a></td>
<td>Weighted Averages Scores for Species</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">eigenvals.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.cca</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.default</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.dudi</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.pca</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.pcnm</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.pco</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.prcomp</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.princomp</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">eigenvals.wcmdscale</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">envfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">envfit.default</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">envfit.formula</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">estaccumR</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">estimateR</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">estimateR.data.frame</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">estimateR.default</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">estimateR.matrix</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="eventstar.html">eventstar</a></td>
<td>Scale Parameter at the Minimum of the Tsallis Evenness Profile</td></tr>
<tr><td style="width: 25%;"><a href="deviance.cca.html">extractAIC.cca</a></td>
<td>Statistics Resembling Deviance and AIC for Constrained Ordination</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="envfit.html">factorfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="MOStest.html">fieller.MOStest</a></td>
<td>Mitchell-Olds &amp; Shaw Test for the Location of Quadratic Extreme</td></tr>
<tr><td style="width: 25%;"><a href="diversity.html">fisher.alpha</a></td>
<td>Ecological Diversity Indices</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">fisherfit</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">fitted.capscale</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">fitted.cca</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">fitted.dbrda</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">fitted.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">fitted.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">fitted.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">fitted.rda</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="dispweight.html">gdispweight</a></td>
<td>Dispersion-based weighting of species counts</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">goodness</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">goodness.cca</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="goodness.metaMDS.html">goodness.metaMDS</a></td>
<td>Goodness of Fit and Shepard Plot for Nonmetric Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="goodness.metaMDS.html">goodness.monoMDS</a></td>
<td>Goodness of Fit and Shepard Plot for Nonmetric Multidimensional Scaling</td></tr>
</table>

<h2><a name="H">-- H --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="influence.cca.html">hatvalues.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">hatvalues.rda</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">head.summary.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">hiersimu</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">hiersimu.default</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="adipart.html">hiersimu.formula</a></td>
<td>Additive Diversity Partitioning and Hierarchical Null Model Testing</td></tr>
<tr><td style="width: 25%;"><a href="vegan-deprecated.html">humpfit</a></td>
<td>Deprecated Functions in vegan package</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">humpfit-deprecated</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
</table>

<h2><a name="I">-- I --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ordiplot.html">identify.ordiplot</a></td>
<td>Alternative plot and identify Functions for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="indpower.html">indpower</a></td>
<td>Indicator Power of Species</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">inertcomp</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">initMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">intersetcor</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="isomap.html">isomap</a></td>
<td>Isometric Feature Mapping Ordination</td></tr>
<tr><td style="width: 25%;"><a href="isomap.html">isomapdist</a></td>
<td>Isometric Feature Mapping Ordination</td></tr>
</table>

<h2><a name="K">-- K --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="kendall.global.html">kendall.global</a></td>
<td>Kendall coefficient of concordance</td></tr>
<tr><td style="width: 25%;"><a href="kendall.global.html">kendall.post</a></td>
<td>Kendall coefficient of concordance</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="envfit.html">labels.envfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">lines.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">lines.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">lines.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">lines.preston</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">lines.prestonfit</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">lines.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">lines.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">lines.radline</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">lines.spantree</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">lines.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="linestack.html">linestack</a></td>
<td>Plots One-dimensional Diagrams without Overwriting Labels</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">logLik, radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">logLik, radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">logLik.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="make.cepnames.html">make.cepnames</a></td>
<td>Abbreviates a Botanical or Zoological Latin Name into an Eight-character Name</td></tr>
<tr><td style="width: 25%;"><a href="commsim.html">make.commsim</a></td>
<td>Create an Object for Null Model Algorithms</td></tr>
<tr><td style="width: 25%;"><a href="mantel.html">mantel</a></td>
<td>Mantel and Partial Mantel Tests for Dissimilarity Matrices</td></tr>
<tr><td style="width: 25%;"><a href="mantel.correlog.html">mantel.correlog</a></td>
<td>Mantel Correlogram</td></tr>
<tr><td style="width: 25%;"><a href="mantel.html">mantel.partial</a></td>
<td>Mantel and Partial Mantel Tests for Dissimilarity Matrices</td></tr>
<tr><td style="width: 25%;"><a href="MDSrotate.html">MDSrotate</a></td>
<td>Rotate First MDS Dimension Parallel to an External Variable</td></tr>
<tr><td style="width: 25%;"><a href="mrpp.html">meandist</a></td>
<td>Multi Response Permutation Procedure and Mean Dissimilarity Matrix</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">metaMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">metaMDSdist</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">metaMDSiter</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">metaMDSredist</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="mite.html">mite</a></td>
<td>Oribatid Mite Data with Explanatory Variables</td></tr>
<tr><td style="width: 25%;"><a href="mite.html">mite.env</a></td>
<td>Oribatid Mite Data with Explanatory Variables</td></tr>
<tr><td style="width: 25%;"><a href="mite.html">mite.pcnm</a></td>
<td>Oribatid Mite Data with Explanatory Variables</td></tr>
<tr><td style="width: 25%;"><a href="mite.html">mite.xy</a></td>
<td>Oribatid Mite Data with Explanatory Variables</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">model.frame.cca</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">model.matrix.cca</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">model.matrix.rda</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="monoMDS.html">monoMDS</a></td>
<td>Global and Local Non-metric Multidimensional Scaling and Linear and Hybrid Scaling</td></tr>
<tr><td style="width: 25%;"><a href="MOStest.html">MOStest</a></td>
<td>Mitchell-Olds &amp; Shaw Test for the Location of Quadratic Extreme</td></tr>
<tr><td style="width: 25%;"><a href="mrpp.html">mrpp</a></td>
<td>Multi Response Permutation Procedure and Mean Dissimilarity Matrix</td></tr>
<tr><td style="width: 25%;"><a href="mso.html">mso</a></td>
<td>Functions for performing and displaying a spatial partitioning of cca or rda results</td></tr>
<tr><td style="width: 25%;"><a href="mso.html">msoplot</a></td>
<td>Functions for performing and displaying a spatial partitioning of cca or rda results</td></tr>
<tr><td style="width: 25%;"><a href="multipart.html">multipart</a></td>
<td>Multiplicative Diversity Partitioning</td></tr>
<tr><td style="width: 25%;"><a href="multipart.html">multipart.default</a></td>
<td>Multiplicative Diversity Partitioning</td></tr>
<tr><td style="width: 25%;"><a href="multipart.html">multipart.formula</a></td>
<td>Multiplicative Diversity Partitioning</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestedbetajac</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestedbetasor</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestedchecker</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nesteddisc</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestedn0</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestednodf</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">nestedtemp</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="distconnected.html">no.shared</a></td>
<td>Connectedness of Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.adonis</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.betadisper</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.cca</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.CCorA</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.decorana</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">nobs.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.isomap</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.metaMDS</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.pcnm</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.procrustes</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.rad</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.varpart</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nobs.adonis.html">nobs.wcmdscale</a></td>
<td>Extract the Number of Observations from a vegan Fit.</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">nullmodel</a></td>
<td>Null Model and Simulation</td></tr>
</table>

<h2><a name="O">-- O --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="oecosimu.html">oecosimu</a></td>
<td>Evaluate Statistics with Null Models of Biological Communities</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">ordConstrained</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cascadeKM.html">orderingKM</a></td>
<td>K-means partitioning using a range of values of K</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordiareatest</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordiArrowTextXY.html">ordiArrowMul</a></td>
<td>Support Functions for Drawing Vectors</td></tr>
<tr><td style="width: 25%;"><a href="ordiarrows.html">ordiarrows</a></td>
<td>Add Arrows and Line Segments to Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordiArrowTextXY.html">ordiArrowTextXY</a></td>
<td>Support Functions for Drawing Vectors</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordibar</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">ordicloud</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordicluster</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordiellipse</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordiarrows.html">ordigrid</a></td>
<td>Add Arrows and Line Segments to Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordihull</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordilabel.html">ordilabel</a></td>
<td>Add Text on Non-transparent Label to an Ordination Plot.</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">ordilattice.getEnvfit</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">ordimedian</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="ordiplot.html">ordiplot</a></td>
<td>Alternative plot and identify Functions for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordipointlabel.html">ordipointlabel</a></td>
<td>Ordination Plots with Points and Optimized Locations for Text</td></tr>
<tr><td style="width: 25%;"><a href="ordistep.html">ordiR2step</a></td>
<td>Choose a Model by Permutation Tests in Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordiresids.html">ordiresids</a></td>
<td>Plots of Residuals and Fitted Values for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordiarrows.html">ordisegments</a></td>
<td>Add Arrows and Line Segments to Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">ordispider</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">ordisplom</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordistep.html">ordistep</a></td>
<td>Choose a Model by Permutation Tests in Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordisurf.html">ordisurf</a></td>
<td>Fit and Plot Smooth Surfaces of Variables on Ordination.</td></tr>
<tr><td style="width: 25%;"><a href="ordisurf.html">ordisurf.default</a></td>
<td>Fit and Plot Smooth Surfaces of Variables on Ordination.</td></tr>
<tr><td style="width: 25%;"><a href="ordisurf.html">ordisurf.formula</a></td>
<td>Fit and Plot Smooth Surfaces of Variables on Ordination.</td></tr>
<tr><td style="width: 25%;"><a href="orditkplot.html">orditkplot</a></td>
<td>Ordination Plot with Movable Labels</td></tr>
<tr><td style="width: 25%;"><a href="orditorp.html">orditorp</a></td>
<td>Add Text or Points to Ordination Plots</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">ordixyplot</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">ordiYbar</a></td>
<td>Result Object from Constrained Ordination</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ordixyplot.html">panel.ordi</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">panel.ordi3d</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">panel.ordiarrows</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="pcnm.html">pcnm</a></td>
<td>Principal Coordinates of Neighbourhood Matrix</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">permatfull</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">permatswap</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.adonis</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.anosim</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.anova.cca</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.CCorA</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.envfit</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.factorfit</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.mantel</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.mrpp</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.mso</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.oecosimu</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.ordiareatest</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.permutest.betadisper</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.permutest.cca</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.protest</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">permustats.vectorfit</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permutations.html">permutations</a></td>
<td>Permutation tests in Vegan</td></tr>
<tr><td style="width: 25%;"><a href="anova.cca.html">permutest</a></td>
<td>Permutation Test for Constrained Correspondence Analysis, Redundancy Analysis and Constrained Analysis of Principal Coordinates</td></tr>
<tr><td style="width: 25%;"><a href="permutest.betadisper.html">permutest.betadisper</a></td>
<td>Permutation test of multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="anova.cca.html">permutest.cca</a></td>
<td>Permutation Test for Constrained Correspondence Analysis, Redundancy Analysis and Constrained Analysis of Principal Coordinates</td></tr>
<tr><td style="width: 25%;"><a href="renyi.html">persp.renyiaccum</a></td>
<td>Renyi and Hill Diversities and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="tsallis.html">persp.tsallisaccum</a></td>
<td>Tsallis Diversity and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="anosim.html">plot.anosim</a></td>
<td>Analysis of Similarities</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">plot.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="betadiver.html">plot.betadiver</a></td>
<td>Indices of beta Diversity</td></tr>
<tr><td style="width: 25%;"><a href="cascadeKM.html">plot.cascadeKM</a></td>
<td>K-means partitioning using a range of values of K</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">plot.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="clamtest.html">plot.clamtest</a></td>
<td>Multinomial Species Classification Method (CLAM)</td></tr>
<tr><td style="width: 25%;"><a href="contribdiv.html">plot.contribdiv</a></td>
<td>Contribution Diversity Approach</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">plot.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">plot.envfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">plot.fisher</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">plot.fisherfit</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">plot.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">plot.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="isomap.html">plot.isomap</a></td>
<td>Isometric Feature Mapping Ordination</td></tr>
<tr><td style="width: 25%;"><a href="mantel.correlog.html">plot.mantel.correlog</a></td>
<td>Mantel Correlogram</td></tr>
<tr><td style="width: 25%;"><a href="mrpp.html">plot.meandist</a></td>
<td>Multi Response Permutation Procedure and Mean Dissimilarity Matrix</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">plot.metaMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="monoMDS.html">plot.monoMDS</a></td>
<td>Global and Local Non-metric Multidimensional Scaling and Linear and Hybrid Scaling</td></tr>
<tr><td style="width: 25%;"><a href="MOStest.html">plot.MOStest</a></td>
<td>Mitchell-Olds &amp; Shaw Test for the Location of Quadratic Extreme</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">plot.nestednodf</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="nestedtemp.html">plot.nestedtemp</a></td>
<td>Nestedness Indices for Communities of Islands or Patches</td></tr>
<tr><td style="width: 25%;"><a href="ordipointlabel.html">plot.ordipointlabel</a></td>
<td>Ordination Plots with Points and Optimized Locations for Text</td></tr>
<tr><td style="width: 25%;"><a href="ordisurf.html">plot.ordisurf</a></td>
<td>Fit and Plot Smooth Surfaces of Variables on Ordination.</td></tr>
<tr><td style="width: 25%;"><a href="orditkplot.html">plot.orditkplot</a></td>
<td>Ordination Plot with Movable Labels</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">plot.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">plot.poolaccum</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="prc.html">plot.prc</a></td>
<td>Principal Response Curves for Treatments with Repeated Observations</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">plot.preston</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">plot.prestonfit</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">plot.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">plot.rad</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">plot.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">plot.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">plot.radline</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="renyi.html">plot.renyi</a></td>
<td>Renyi and Hill Diversities and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="renyi.html">plot.renyiaccum</a></td>
<td>Renyi and Hill Diversities and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">plot.spantree</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">plot.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="taxondive.html">plot.taxondive</a></td>
<td>Indices of Taxonomic Diversity and Distinctness</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">plot.varpart</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">plot.varpart234</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="wcmdscale.html">plot.wcmdscale</a></td>
<td>Weighted Classical (Metric) Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">points.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">points.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">points.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">points.metaMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="monoMDS.html">points.monoMDS</a></td>
<td>Global and Local Non-metric Multidimensional Scaling and Linear and Hybrid Scaling</td></tr>
<tr><td style="width: 25%;"><a href="ordiplot.html">points.ordiplot</a></td>
<td>Alternative plot and identify Functions for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="orditkplot.html">points.orditkplot</a></td>
<td>Ordination Plot with Movable Labels</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">points.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">points.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">points.radline</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">poolaccum</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">postMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="prc.html">prc</a></td>
<td>Principal Response Curves for Treatments with Repeated Observations</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">predict.cca</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">predict.decorana</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">predict.fitspecaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">predict.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">predict.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">predict.radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">predict.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">predict.radline</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">predict.rda</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">predict.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="cascadeKM.html">pregraphKM</a></td>
<td>K-means partitioning using a range of values of K</td></tr>
<tr><td style="width: 25%;"><a href="ordixyplot.html">prepanel.ordi3d</a></td>
<td>Trellis (Lattice) Plots for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">prestondistr</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">prestonfit</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">print.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">print.cca</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="commsim.html">print.commsim</a></td>
<td>Create an Object for Null Model Algorithms</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">print.nullmodel</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">print.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">print.simmat</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">print.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">print.summary.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">print.summary.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">print.summary.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">profile.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="MOStest.html">profile.MOStest</a></td>
<td>Mitchell-Olds &amp; Shaw Test for the Location of Quadratic Extreme</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">protest</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="pyrifos.html">pyrifos</a></td>
<td>Response of Aquatic Invertebrates to Insecticide Treatment</td></tr>
</table>

<h2><a name="Q">-- Q --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="permustats.html">qqmath.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">qqnorm.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">qr.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="radfit.html">rad.lognormal</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">rad.null</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">rad.preempt</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">rad.zipf</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">rad.zipfbrot</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">radfit</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">radfit.data.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">radfit.default</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">radlattice</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="rankindex.html">rankindex</a></td>
<td>Compares Dissimilarity Indices for Gradient Detection</td></tr>
<tr><td style="width: 25%;"><a href="rarefy.html">rarecurve</a></td>
<td>Rarefaction Species Richness</td></tr>
<tr><td style="width: 25%;"><a href="rarefy.html">rarefy</a></td>
<td>Rarefaction Species Richness</td></tr>
<tr><td style="width: 25%;"><a href="rarefy.html">rareslope</a></td>
<td>Rarefaction Species Richness</td></tr>
<tr><td style="width: 25%;"><a href="raupcrick.html">raupcrick</a></td>
<td>Raup-Crick Dissimilarity with Unequal Sampling Densities of Species</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">rda</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">rda.default</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="cca.html">rda.formula</a></td>
<td>[Partial] [Constrained] Correspondence Analysis and Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="read.cep.html">read.cep</a></td>
<td>Reads a CEP (Canoco) data file</td></tr>
<tr><td style="width: 25%;"><a href="renyi.html">renyi</a></td>
<td>Renyi and Hill Diversities and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="renyi.html">renyiaccum</a></td>
<td>Renyi and Hill Diversities and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="reorder.hclust.html">reorder.hclust</a></td>
<td>Reorder a Hierarchical Clustering Tree</td></tr>
<tr><td style="width: 25%;"><a href="predict.cca.html">residuals.cca</a></td>
<td>Prediction Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">residuals.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="reorder.hclust.html">rev.hclust</a></td>
<td>Reorder a Hierarchical Clustering Tree</td></tr>
<tr><td style="width: 25%;"><a href="rarefy.html">rrarefy</a></td>
<td>Rarefaction Species Richness</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj.cca</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj.default</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj.glm</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj.lm</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="RsquareAdj.html">RsquareAdj.rda</a></td>
<td>Adjusted R-square</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">rstandard.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">rstudent.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="scores.html">scores</a></td>
<td>Get Species or Site Scores from an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">scores.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
<tr><td style="width: 25%;"><a href="betadiver.html">scores.betadiver</a></td>
<td>Indices of beta Diversity</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">scores.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">scores.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="scores.html">scores.default</a></td>
<td>Get Species or Site Scores from an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">scores.envfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="reorder.hclust.html">scores.hclust</a></td>
<td>Reorder a Hierarchical Clustering Tree</td></tr>
<tr><td style="width: 25%;"><a href="scores.html">scores.lda</a></td>
<td>Get Species or Site Scores from an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">scores.metaMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="monoMDS.html">scores.monoMDS</a></td>
<td>Global and Local Non-metric Multidimensional Scaling and Linear and Hybrid Scaling</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">scores.ordihull</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordiplot.html">scores.ordiplot</a></td>
<td>Alternative plot and identify Functions for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="orditkplot.html">scores.orditkplot</a></td>
<td>Ordination Plot with Movable Labels</td></tr>
<tr><td style="width: 25%;"><a href="pcnm.html">scores.pcnm</a></td>
<td>Principal Coordinates of Neighbourhood Matrix</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">scores.rda</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="wcmdscale.html">scores.wcmdscale</a></td>
<td>Weighted Classical (Metric) Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">screeplot.cca</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">screeplot.decorana</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">screeplot.prcomp</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="screeplot.cca.html">screeplot.princomp</a></td>
<td>Screeplots for Ordination Results and Broken Stick Distributions</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">showvarparts</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">sigma.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">simmat</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="simper.html">simper</a></td>
<td>Similarity Percentages</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">simpleDBRDA</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">simpleRDA2</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="simulate.rda.html">simulate.capscale</a></td>
<td>Simulate Responses with Gaussian Error or Permuted Residuals for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="simulate.rda.html">simulate.cca</a></td>
<td>Simulate Responses with Gaussian Error or Permuted Residuals for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">simulate.nullmodel</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="simulate.rda.html">simulate.rda</a></td>
<td>Simulate Responses with Gaussian Error or Permuted Residuals for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="sipoo.html">sipoo</a></td>
<td>Birds in the Archipelago of Sipoo (Sibbo and Borgå)</td></tr>
<tr><td style="width: 25%;"><a href="sipoo.html">sipoo.map</a></td>
<td>Birds in the Archipelago of Sipoo (Sibbo and Borgå)</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">smbind</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">spandepth</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="spantree.html">spantree</a></td>
<td>Minimum Spanning Tree</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="diversity.html">specnumber</a></td>
<td>Ecological Diversity Indices</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">specpool</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">specpool2vect</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">specslope</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">spenvcor</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
<tr><td style="width: 25%;"><a href="sppscores.html">sppscores</a></td>
<td>Add or Replace Species Scores in Distance-Based Ordination</td></tr>
<tr><td style="width: 25%;"><a href="sppscores.html">sppscores&lt;-</a></td>
<td>Add or Replace Species Scores in Distance-Based Ordination</td></tr>
<tr><td style="width: 25%;"><a href="sppscores.html">sppscores&lt;-.capscale</a></td>
<td>Add or Replace Species Scores in Distance-Based Ordination</td></tr>
<tr><td style="width: 25%;"><a href="sppscores.html">sppscores&lt;-.dbrda</a></td>
<td>Add or Replace Species Scores in Distance-Based Ordination</td></tr>
<tr><td style="width: 25%;"><a href="sppscores.html">sppscores&lt;-.metaMDS</a></td>
<td>Add or Replace Species Scores in Distance-Based Ordination</td></tr>
<tr><td style="width: 25%;"><a href="SSarrhenius.html">SSarrhenius</a></td>
<td>Self-Starting nls Species-Area Models</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">SSD.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="SSarrhenius.html">SSgitay</a></td>
<td>Self-Starting nls Species-Area Models</td></tr>
<tr><td style="width: 25%;"><a href="SSarrhenius.html">SSgleason</a></td>
<td>Self-Starting nls Species-Area Models</td></tr>
<tr><td style="width: 25%;"><a href="SSarrhenius.html">SSlomolino</a></td>
<td>Self-Starting nls Species-Area Models</td></tr>
<tr><td style="width: 25%;"><a href="stepacross.html">stepacross</a></td>
<td>Stepacross as Flexible Shortest Paths or Extended Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="nullmodel.html">str.nullmodel</a></td>
<td>Null Model and Simulation</td></tr>
<tr><td style="width: 25%;"><a href="goodness.metaMDS.html">stressplot</a></td>
<td>Goodness of Fit and Shepard Plot for Nonmetric Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.capscale</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.cca</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.dbrda</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="goodness.metaMDS.html">stressplot.default</a></td>
<td>Goodness of Fit and Shepard Plot for Nonmetric Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="goodness.metaMDS.html">stressplot.monoMDS</a></td>
<td>Goodness of Fit and Shepard Plot for Nonmetric Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.prcomp</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.princomp</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.rda</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="stressplot.wcmdscale.html">stressplot.wcmdscale</a></td>
<td>Display Ordination Distances Against Observed Distances in Eigenvector Ordinations</td></tr>
<tr><td style="width: 25%;"><a href="anosim.html">summary.anosim</a></td>
<td>Analysis of Similarities</td></tr>
<tr><td style="width: 25%;"><a href="bioenv.html">summary.bioenv</a></td>
<td>Best Subset of Environmental Variables with Maximum (Rank) Correlation with Community Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">summary.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="clamtest.html">summary.clamtest</a></td>
<td>Multinomial Species Classification Method (CLAM)</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">summary.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="dispweight.html">summary.dispweight</a></td>
<td>Dispersion-based weighting of species counts</td></tr>
<tr><td style="width: 25%;"><a href="eigenvals.html">summary.eigenvals</a></td>
<td>Extract Eigenvalues from an Ordination Object</td></tr>
<tr><td style="width: 25%;"><a href="humpfit.html">summary.humpfit</a></td>
<td>No-interaction Model for Hump-backed Species Richness vs. Biomass</td></tr>
<tr><td style="width: 25%;"><a href="isomap.html">summary.isomap</a></td>
<td>Isometric Feature Mapping Ordination</td></tr>
<tr><td style="width: 25%;"><a href="mrpp.html">summary.meandist</a></td>
<td>Multi Response Permutation Procedure and Mean Dissimilarity Matrix</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">summary.ordiellipse</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="ordihull.html">summary.ordihull</a></td>
<td>Display Groups or Factor Levels in Ordination Diagrams</td></tr>
<tr><td style="width: 25%;"><a href="permatfull.html">summary.permat</a></td>
<td>Matrix Permutation Algorithms for Presence-Absence and Count Data</td></tr>
<tr><td style="width: 25%;"><a href="permustats.html">summary.permustats</a></td>
<td>Extract, Analyse and Display Permutation Results</td></tr>
<tr><td style="width: 25%;"><a href="specpool.html">summary.poolaccum</a></td>
<td>Extrapolated Species Richness in a Species Pool</td></tr>
<tr><td style="width: 25%;"><a href="prc.html">summary.prc</a></td>
<td>Principal Response Curves for Treatments with Repeated Observations</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">summary.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="radfit.html">summary.radfit.frame</a></td>
<td>Rank - Abundance or Dominance / Diversity Models</td></tr>
<tr><td style="width: 25%;"><a href="simper.html">summary.simper</a></td>
<td>Similarity Percentages</td></tr>
<tr><td style="width: 25%;"><a href="specaccum.html">summary.specaccum</a></td>
<td>Species Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="taxondive.html">summary.taxondive</a></td>
<td>Indices of Taxonomic Diversity and Distinctness</td></tr>
<tr><td style="width: 25%;"><a href="beals.html">swan</a></td>
<td>Beals Smoothing and Degree of Absence</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="vegemite.html">tabasco</a></td>
<td>Display Compact Ordered Community Tables</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">tail.summary.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="taxondive.html">taxa2dist</a></td>
<td>Indices of Taxonomic Diversity and Distinctness</td></tr>
<tr><td style="width: 25%;"><a href="taxondive.html">taxondive</a></td>
<td>Indices of Taxonomic Diversity and Distinctness</td></tr>
<tr><td style="width: 25%;"><a href="plot.cca.html">text.cca</a></td>
<td>Plot or Extract Results of Constrained Correspondence Analysis or Redundancy Analysis</td></tr>
<tr><td style="width: 25%;"><a href="decorana.html">text.decorana</a></td>
<td>Detrended Correspondence Analysis and Basic Reciprocal Averaging</td></tr>
<tr><td style="width: 25%;"><a href="metaMDS.html">text.metaMDS</a></td>
<td>Nonmetric Multidimensional Scaling with Stable Solution from Random Starts, Axis Scaling and Species Scores</td></tr>
<tr><td style="width: 25%;"><a href="monoMDS.html">text.monoMDS</a></td>
<td>Global and Local Non-metric Multidimensional Scaling and Linear and Hybrid Scaling</td></tr>
<tr><td style="width: 25%;"><a href="ordiplot.html">text.ordiplot</a></td>
<td>Alternative plot and identify Functions for Ordination</td></tr>
<tr><td style="width: 25%;"><a href="orditkplot.html">text.orditkplot</a></td>
<td>Ordination Plot with Movable Labels</td></tr>
<tr><td style="width: 25%;"><a href="procrustes.html">text.procrustes</a></td>
<td>Procrustes Rotation of Two Configurations and PROTEST</td></tr>
<tr><td style="width: 25%;"><a href="tolerance.html">tolerance</a></td>
<td>Species tolerances and sample heterogeneities</td></tr>
<tr><td style="width: 25%;"><a href="tolerance.html">tolerance.cca</a></td>
<td>Species tolerances and sample heterogeneities</td></tr>
<tr><td style="width: 25%;"><a href="treedive.html">treedist</a></td>
<td>Functional Diversity and Community Distances from Species Trees</td></tr>
<tr><td style="width: 25%;"><a href="treedive.html">treedive</a></td>
<td>Functional Diversity and Community Distances from Species Trees</td></tr>
<tr><td style="width: 25%;"><a href="treedive.html">treeheight</a></td>
<td>Functional Diversity and Community Distances from Species Trees</td></tr>
<tr><td style="width: 25%;"><a href="tsallis.html">tsallis</a></td>
<td>Tsallis Diversity and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="tsallis.html">tsallisaccum</a></td>
<td>Tsallis Diversity and Corresponding Accumulation Curves</td></tr>
<tr><td style="width: 25%;"><a href="betadisper.html">TukeyHSD.betadisper</a></td>
<td>Multivariate homogeneity of groups dispersions (variances)</td></tr>
</table>

<h2><a name="U">-- U --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="nullmodel.html">update.nullmodel</a></td>
<td>Null Model and Simulation</td></tr>
</table>

<h2><a name="V">-- V --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="varechem.html">varechem</a></td>
<td>Vegetation and environment in lichen pastures</td></tr>
<tr><td style="width: 25%;"><a href="varechem.html">varespec</a></td>
<td>Vegetation and environment in lichen pastures</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">varpart</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">varpart2</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">varpart3</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="varpart.html">varpart4</a></td>
<td>Partition the Variation of Community Matrix by 2, 3, or 4 Explanatory Matrices</td></tr>
<tr><td style="width: 25%;"><a href="influence.cca.html">vcov.cca</a></td>
<td>Linear Model Diagnostics for Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="envfit.html">vectorfit</a></td>
<td>Fits an Environmental Vector or Factor onto an Ordination</td></tr>
<tr><td style="width: 25%;"><a href="vegan-package.html">vegan</a></td>
<td>Community Ecology Package: Ordination, Diversity and Dissimilarities</td></tr>
<tr><td style="width: 25%;"><a href="vegan-deprecated.html">vegan-deprecated</a></td>
<td>Deprecated Functions in vegan package</td></tr>
<tr><td style="width: 25%;"><a href="vegandocs.html">vegandocs</a></td>
<td>Display vegan Documentation</td></tr>
<tr><td style="width: 25%;"><a href="vegdist.html">vegdist</a></td>
<td>Dissimilarity Indices for Community Ecologists</td></tr>
<tr><td style="width: 25%;"><a href="vegemite.html">vegemite</a></td>
<td>Display Compact Ordered Community Tables</td></tr>
<tr><td style="width: 25%;"><a href="fisherfit.html">veiledspec</a></td>
<td>Fit Fisher's Logseries and Preston's Lognormal Model to Abundance Data</td></tr>
<tr><td style="width: 25%;"><a href="goodness.cca.html">vif.cca</a></td>
<td>Diagnostic Tools for [Constrained] Ordination (CCA, RDA, DCA, CA, PCA)</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="wascores.html">wascores</a></td>
<td>Weighted Averages Scores for Species</td></tr>
<tr><td style="width: 25%;"><a href="wcmdscale.html">wcmdscale</a></td>
<td>Weighted Classical (Metric) Multidimensional Scaling</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">weights.cca</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">weights.decorana</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="cca.object.html">weights.rda</a></td>
<td>Result Object from Constrained Ordination</td></tr>
<tr><td style="width: 25%;"><a href="decostand.html">wisconsin</a></td>
<td>Standardization Methods for Community Ecology</td></tr>
</table>
</body></html>
