

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  

  <title>Glossary &mdash; scirpy  documentation</title>
  


  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/scanpy.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Other versions" href="versions.html" />
    <link rel="prev" title="Immune receptor (IR) model" href="ir-biology.html" />
 


</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.9.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage-principles.html">Usage principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Glossary</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/icbi-lab/scirpy/blob/master/docs/glossary.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="glossary">
<h1>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline">¶</a></h1>
<dl class="glossary">
<dt id="term-AIRR">AIRR</dt><dd><p>Adaptive Immune Receptor Repertoire. Within the Scirpy documentation, we simply
speak of <a class="reference internal" href="#term-IR"><span class="xref std std-term">immune receptors (IR)</span></a>.</p>
<p>The <a class="reference external" href="https://www.antibodysociety.org/the-airr-community/">AIRR community</a>
defines standards around AIRR data. Scirpy supports the <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR Rearrangement</a>
schema and complies with the <a class="reference external" href="https://docs.airr-community.org/en/latest/swtools/airr_swtools_standard.html">AIRR Software Guidelines</a>.</p>
</dd>
<dt id="term-Alellically-included-B-cells">Alellically included B-cells</dt><dd><p>A B cell with two pairs of <a class="reference internal" href="#term-IG"><span class="xref std std-term">IG</span></a> chains. See <a class="reference internal" href="#term-Dual-IR"><span class="xref std std-term">Dual IR</span></a>.</p>
</dd>
<dt id="term-BCR">BCR</dt><dd><p>B-cell receptor. A BCR consiste of two Immunoglobulin (IG) heavy chains and
two IG light chains. The two light chains contain a variable region, which is
responsible for antigen recognition.</p>
<figure class="align-default" id="id13">
<a class="reference internal image-reference" href="_images/bcr.jpg"><img alt="_images/bcr.jpg" src="_images/bcr.jpg" style="width: 310px;" /></a>
<figcaption>
<p><span class="caption-text">Image By CNX <a class="reference external" href="http://cnx.org/contents/GFy_h8cu&#64;10.53:rZudN6XP&#64;2/Introduction">OpenStax</a>
under the <a class="reference external" href="https://creativecommons.org/licenses/by/4.0/deed.en">CC BY-4.0</a> license,
obtained from <a class="reference external" href="https://commons.wikimedia.org/w/index.php?curid=49935883">wikimedia commons</a></span><a class="headerlink" href="#id13" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
<dt id="term-CDR">CDR</dt><dd><p>Complementary-determining region. The diversity and, therefore, antigen-specificity
of <a class="reference internal" href="#term-IR"><span class="xref std std-term">IRs</span></a> is predominanly determined by three hypervariable loops
(CDR1, CDR2, and CDR3) on each of the α- and β receptor arms.</p>
<p>CDR1 and CDR2 are fully encoded in germline V genes. In contrast,
the CDR3 loops are assembled from <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">V, (D), and J segments</span></a> and
comprise random additions and deletions at the junction
sites. Thus, CDR3 regions make up a large part of the
adpative immune receptor variability and are therefore thought to
be particularly important for antigen specificity
(reviewed in <span id="id3">[<a class="reference internal" href="zzz_bibliography.html#id14" title="Meriem Attaf, Eric Huseby, and Andrew K Sewell. $\upalpha $$\upbeta $ t cell receptors as predictors of health and disease. Cellular &amp; Molecular Immunology, 12(4):391–399, January 2015. URL: https://doi.org/10.1038/cmi.2014.134, doi:10.1038/cmi.2014.134.">AHS15</a>]</span>).</p>
<figure class="align-default" id="id14">
<a class="reference internal image-reference" href="_images/tcr_cdr3.png"><img alt="_images/tcr_cdr3.png" src="_images/tcr_cdr3.png" style="width: 310px;" /></a>
<figcaption>
<p><span class="caption-text">Image from <span id="id4">[<a class="reference internal" href="zzz_bibliography.html#id14" title="Meriem Attaf, Eric Huseby, and Andrew K Sewell. $\upalpha $$\upbeta $ t cell receptors as predictors of health and disease. Cellular &amp; Molecular Immunology, 12(4):391–399, January 2015. URL: https://doi.org/10.1038/cmi.2014.134, doi:10.1038/cmi.2014.134.">AHS15</a>]</span> under the <a class="reference external" href="https://creativecommons.org/licenses/by-nc-sa/3.0/">CC BY-NC-SA-3.0</a> license.</span><a class="headerlink" href="#id14" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
<dt id="term-CDR3">CDR3</dt><dd><p>Complementary-determining region 3. See <a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR</span></a>.</p>
</dd>
<dt id="term-Chain-locus">Chain locus</dt><dd><p>Scirpy supports all valid <a class="reference external" href="http://www.imgt.org/IMGTScientificChart/Nomenclature/IMGTnomenclature.html">IGMT locus names</a>:</p>
<dl class="simple">
<dt>Loci with a <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VJ</span></a> junction:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">TRA</span></code> (T-cell receptor alpha)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TRG</span></code> (T-cell receptor gamma)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">IGL</span></code> (Immunoglobulin lambda)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">IGK</span></code> (Immunoglobulin kappa)</p></li>
</ul>
</dd>
<dt>Loci with a <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VDJ</span></a> junction:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">TRB</span></code> (T-cell receptor beta)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TRD</span></code> (T cell receptor delta)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">IGH</span></code> (Immunoglobulin heavy chain)</p></li>
</ul>
</dd>
</dl>
</dd>
<dt id="term-Clonotype">Clonotype</dt><dd><p>A clonotype designates a collection of T or B cells that descend from a
common, antecedent cell, and therefore, bear the same adaptive
immune receptors and recognize the same <a class="reference internal" href="#term-Epitope"><span class="xref std std-term">epitopes</span></a>.</p>
<p>In single-cell RNA-sequencing (scRNA-seq) data, T or B cells sharing identical
complementarity-determining regions 3 (<a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a>) nucleotide sequences of both
<a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VJ</span></a> and <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VDJ</span></a> chains (e.g. both α and β <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a> chains)
make up a clonotype..</p>
<p><em>Scirpy</em> provides a flexible approach to clonotype definition based on
<a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR3</span></a> sequence identity or similarity. Additionally, it is possible
to require clonotypes to have the same <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">V-gene</span></a>, enforcing the CDR 1
and 2 regions to be the same.</p>
<p>For more details, see the page about our <a class="reference internal" href="ir-biology.html#receptor-model"><span class="std std-ref">IR model</span></a> and
the API documentation of <a class="reference internal" href="generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotypes()</span></code></a>.</p>
</dd>
<dt id="term-Clonotype-cluster">Clonotype cluster</dt><dd><p>A higher-order aggregation of <a class="reference internal" href="#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> that have different
<a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR3</span></a> nucleotide sequences, but might recognize the same antigen
because they have the same or similar CDR3 amino acid sequence.</p>
<p>See also: <a class="reference internal" href="generated/scirpy.tl.define_clonotype_clusters.html#scirpy.tl.define_clonotype_clusters" title="scirpy.tl.define_clonotype_clusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotype_clusters()</span></code></a>.</p>
</dd>
<dt id="term-Clonotype-modularity">Clonotype modularity</dt><dd><p>The clonotype modularity measures how densly connected the transcriptomics
neighborhood graph underlying the cells in a clonotype is. Clonotypes with
a high modularity consist of cells that are transcriptionally more similar
than that of a clonotype with a low modularity.
See also <a class="reference internal" href="generated/scirpy.tl.clonotype_modularity.html#scirpy.tl.clonotype_modularity" title="scirpy.tl.clonotype_modularity"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.clonotype_modularity()</span></code></a></p>
</dd>
<dt id="term-Convergent-evolution-of-clonotypes">Convergent evolution of clonotypes</dt><dd><p>It has been proposed that <a class="reference internal" href="#term-IR"><span class="xref std std-term">IRs</span></a> are subject to convergent evolution,
i.e. a selection pressure that leads to IRs recognizing the same antigen
(<span id="id2">[<a class="reference internal" href="zzz_bibliography.html#id21" title="V. Venturi, K. Kedzierska, D. A. Price, P. C. Doherty, D. C. Douek, S. J. Turner, and M. P. Davenport. Sharing of t cell receptors in antigen-specific responses is driven by convergent recombination. Proceedings of the National Academy of Sciences, 103(49):18691–18696, November 2006. URL: https://doi.org/10.1073/pnas.0608907103, doi:10.1073/pnas.0608907103.">VKP+06</a>]</span>).</p>
<p>Evidence of convergent evolution could be clonotypes with the same <a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a>
amino acid sequence, but different CDR3 nucleotide sequences (due to synonymous
codons) or clonotypes with highly similar CDR3 amino acid sequences that
recognize the same antigen.</p>
</dd>
<dt id="term-Dual-IR">Dual IR</dt><dd><p><a class="reference internal" href="#term-IR"><span class="xref std std-term">IRs</span></a> with more than one pair of <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VJ</span></a> and
<a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VDJ</span></a> sequences. While this was
previously thought to be impossible due to the mechanism of allelic exclusion
(<span id="id9">[<a class="reference internal" href="zzz_bibliography.html#id7" title="Brenna L Brady, Natalie C Steinel, and Craig H Bassing. Antigen receptor allelic exclusion: an update and reappraisal. J. Immunol., 185(7):3801–3808, October 2010.">BSB10</a>]</span>), there is an increasing amound of evidence for a <em>bona fide</em>
dual-IR population (<span id="id10">[<a class="reference internal" href="zzz_bibliography.html#id5" title="Nathaniel J. Schuldt and Bryce A. Binstadt. Dual TCR t cells: identity crisis or multitaskers? The Journal of Immunology, 202(3):637–644, January 2019. URL: https://doi.org/10.4049/jimmunol.1800904, doi:10.4049/jimmunol.1800904.">SB19</a>]</span>, <span id="id11">[<a class="reference internal" href="zzz_bibliography.html#id11" title="Qingyong Ji, Antoine Perchellet, and Joan M Goverman. Viral infection triggers central nervous system autoimmunity via activation of CD8+ T cells expressing dual TCRs. Nat. Immunol., 11(7):628–634, July 2010.">JPG10</a>]</span>, <span id="id12">[<a class="reference internal" href="zzz_bibliography.html#id3" title="Christian Vettermann and Mark S. Schlissel. Allelic exclusion of immunoglobulin genes: models and mechanisms. Immunological Reviews, 237(1):22–42, August 2010. URL: https://doi.org/10.1111/j.1600-065x.2010.00935.x, doi:10.1111/j.1600-065x.2010.00935.x.">VS10</a>]</span>).</p>
<p>For more information on how <em>Scirpy</em> handles dual IRs, see the
page about our <a class="reference internal" href="ir-biology.html#receptor-model"><span class="std std-ref">IR model</span></a>.</p>
</dd>
<dt id="term-Dual-TCR">Dual TCR</dt><dd><p><a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCRs</span></a> with more than one pair of α- and β (or γ- and δ) chains.
See <a class="reference internal" href="#term-Dual-IR"><span class="xref std std-term">Dual IR</span></a>.</p>
</dd>
<dt id="term-Epitope">Epitope</dt><dd><p>The part of an antigen that is recognized by the <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a>, <a class="reference internal" href="#term-BCR"><span class="xref std std-term">BCR</span></a>,
or antibody.</p>
</dd>
<dt id="term-IG">IG</dt><dd><p>Immunoglobulin</p>
</dd>
<dt id="term-IR">IR</dt><dd><p>Immune receptor.</p>
</dd>
<dt id="term-Multi-tissue-clonotype">Multi-tissue clonotype</dt><dd><p>A clonotype that occurs in multiple tissues of the same patient.</p>
</dd>
<dt id="term-Multichain-cell">Multichain-cell</dt><dd><p>Cells with more than two pairs of <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VJ</span></a> and
<a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VDJ</span></a> sequences that do not fit into the <a class="reference internal" href="#term-Dual-IR"><span class="xref std std-term">Dual IR</span></a>
model. These are usually rare and could be explained by
doublets/multiplets, i.e. two ore more cells that were captured in
the same droplet.</p>
<figure class="align-default" id="id15">
<a class="reference internal image-reference" href="_images/multichain.png"><img alt="_images/multichain.png" src="_images/multichain.png" style="width: 450px;" /></a>
<figcaption>
<p><span class="caption-text">(a) UMAP plot of 96,000 cells from <span id="id7">[<a class="reference internal" href="zzz_bibliography.html#id4" title="Thomas D. Wu, Shravan Madireddi, Patricia E. de Almeida, Romain Banchereau, Ying-Jiun J. Chen, Avantika S. Chitre, Eugene Y. Chiang, Hina Iftikhar, William E. O'Gorman, Amelia Au-Yeung, Chikara Takahashi, Leonard D. Goldstein, Chungkee Poon, Shilpa Keerthivasan, Denise E. de Almeida Nagata, Xiangnan Du, Hyang-Mi Lee, Karl L. Banta, Sanjeev Mariathasan, Meghna Das Thakur, Mahrukh A. Huseni, Marcus Ballinger, Ivette Estay, Patrick Caplazi, Zora Modrusan, Lélia Delamarre, Ira Mellman, Richard Bourgon, and Jane L. Grogan. Peripheral t cell expansion predicts tumour infiltration and clinical response. Nature, 579(7798):274–278, February 2020. URL: https://doi.org/10.1038/s41586-020-2056-8, doi:10.1038/s41586-020-2056-8.">WMdA+20</a>]</span> with at least one detected
<a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequence with multichain-cells (n=474) highlighted in green.
(b)  Comparison of detected reads per cell in multichain-cells and other cells.
Multichain cells comprised significantly more reads per cell
(p = 9.45 × 10−251, Wilcoxon-Mann-Whitney-test), supporting the hypothesis that
(most of) multichain cells are technical artifacts arising from
cell-multiplets (<span id="id8">[<a class="reference internal" href="zzz_bibliography.html#id15" title="Tomislav Ilicic, Jong Kyoung Kim, Aleksandra A. Kolodziejczyk, Frederik Otzen Bagger, Davis James McCarthy, John C. Marioni, and Sarah A. Teichmann. Classification of low quality cells from single-cell RNA-seq data. Genome Biology, February 2016. URL: https://doi.org/10.1186/s13059-016-0888-1, doi:10.1186/s13059-016-0888-1.">IKK+16</a>]</span>).</span><a class="headerlink" href="#id15" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
<dt id="term-Orphan-chain">Orphan chain</dt><dd><p>A <a class="reference internal" href="#term-IR"><span class="xref std std-term">IR</span></a> chain is called <em>orphan</em>, if its corresponding counterpart
has not been detected. For instance, if a cell has only a <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VJ</span></a> chain,
(e.g. TCR-alpha), but no <a class="reference internal" href="#term-V-D-J"><span class="xref std std-term">VDJ</span></a> chain (e.g. TCR-beta),
the cell will be flagged as “Orphan VJ”.</p>
<p>Orphan chains are most likely the effect of stochastic dropouts due to
sequencing inefficiencies.</p>
<p>See also <a class="reference internal" href="generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a>.</p>
</dd>
<dt id="term-Private-clonotype">Private clonotype</dt><dd><p>A clonotype that is specific for a certain patient.</p>
</dd>
<dt id="term-Productive-chain">Productive chain</dt><dd><p>Productive chains are <a class="reference internal" href="#term-IR"><span class="xref std std-term">IR</span></a> chains with a <a class="reference internal" href="#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequence that produces
a functional peptide. Scirpy relies on the preprocessing tools (e.g.
CellRanger or TraCeR) for flagging non-productive chains.
Typically chains are flagged as non-productive if they contain
a stop codon or are not within the reading frame.</p>
</dd>
<dt id="term-Public-clonotype">Public clonotype</dt><dd><p>A clonotype that is shared across multiple
patients, e.g. a clonotype recognizing common viral epitope.</p>
<figure class="align-default" id="id16">
<a class="reference internal image-reference" href="_images/public-private.jpg"><img alt="_images/public-private.jpg" src="_images/public-private.jpg" style="width: 420px;" /></a>
<figcaption>
<p><span class="caption-text">Image from <span id="id1">[<a class="reference internal" href="zzz_bibliography.html#id16" title="Ian Setliff, Wyatt J. McDonnell, Nagarajan Raju, Robin G. Bombardi, Amyn A. Murji, Cathrine Scheepers, Rutendo Ziki, Charissa Mynhardt, Bryan E. Shepherd, Alusha A. Mamchak, Nigel Garrett, Salim Abdool Karim, Simon A. Mallal, James E. Crowe, Lynn Morris, and Ivelin S. Georgiev. Multi-donor longitudinal antibody repertoire sequencing reveals the existence of public antibody clonotypes in HIV-1 infection. Cell Host &amp; Microbe, 23(6):845–854.e6, June 2018. URL: https://doi.org/10.1016/j.chom.2018.05.001, doi:10.1016/j.chom.2018.05.001.">SMR+18</a>]</span> under the <a class="reference external" href="https://creativecommons.org/licenses/by/4.0/">CC BY-4.0</a> license.</span><a class="headerlink" href="#id16" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
<dt id="term-Receptor-subtype">Receptor subtype</dt><dd><p>More fine-grained classification of the <a class="reference internal" href="#term-Receptor-type"><span class="xref std std-term">receptor type</span></a>
into</p>
<ul class="simple">
<li><p>α/β T cells</p></li>
<li><p>γ/δ T cells</p></li>
<li><p>IG-heavy/IG-κ B cells</p></li>
<li><p>IG-heavy/IG-λ B cells</p></li>
</ul>
<p>See also <a class="reference internal" href="generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a>.</p>
</dd>
<dt id="term-Receptor-type">Receptor type</dt><dd><p>Classification of immune receptors into <a class="reference internal" href="#term-BCR"><span class="xref std std-term">BCR</span></a> and <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a>.</p>
<p>See also <a class="reference internal" href="generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a>.</p>
</dd>
<dt id="term-TCR">TCR</dt><dd><p>T-cell receptor. A TCR consists of one α and one β chain
(or, alternatively, one γ and one δ chain). Each chain consists of a
constant and a variable region. The variable region is responsible for antigen
recognition, mediated by <a class="reference internal" href="#term-CDR"><span class="xref std std-term">CDR</span></a> regions.</p>
<p>For more information on how Scirpy represents TCRs, see the page about our
<a class="reference internal" href="ir-biology.html#receptor-model"><span class="std std-ref">receptor model</span></a>.</p>
<figure class="align-default" id="id17">
<a class="reference internal image-reference" href="_images/tcr.jpg"><img alt="_images/tcr.jpg" src="_images/tcr.jpg" style="width: 310px;" /></a>
<figcaption>
<p><span class="caption-text">Image from <a class="reference external" href="https://commons.wikimedia.org/wiki/File:2215_Alpha-Beta_T_Cell_Receptor.jpg">Wikimedia commons</a>
under the <a class="reference external" href="https://creativecommons.org/licenses/by/3.0/deed.en">CC BY-3.0</a> license.</span><a class="headerlink" href="#id17" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
<dt id="term-Tissue-specific-clonotype">Tissue-specific clonotype</dt><dd><p>A clonotype that only occurs in a certain tissue of a certain patient.</p>
</dd>
<dt id="term-UMI">UMI</dt><dd><p>Unique molecular identifier. Some single-cell RNA-seq protocols
label each RNA with a unique barcode prior to PCR-amplification to mitigate
PCR bias. With these protocols, UMI-counts replace the read-counts
generally used with RNA-seq.</p>
</dd>
<dt id="term-V-D-J">V(D)J</dt><dd><p>The variability of <a class="reference internal" href="#term-IR"><span class="xref std std-term">IR</span></a> chain sequences originates from the genetic recombination
of <strong>V</strong>ariable, <strong>D</strong>iversity and <strong>J</strong>oining gene segments. The <a class="reference internal" href="#term-TCR"><span class="xref std std-term">TCR</span></a>-α,
TCR-ɣ, <a class="reference internal" href="#term-IG"><span class="xref std std-term">IG</span></a>-κ, and IG-λ chains get assembled from V and J loci only. We refer
to these chains as <code class="docutils literal notranslate"><span class="pre">VJ</span></code> chains in Scirpy. The TCR-β, TCR-δ, and IG-heavy chains
get assembled from all three segments. We refer to these chains as <code class="docutils literal notranslate"><span class="pre">VDJ</span></code>-chains
in Scirpy.</p>
<p>As an example, the figure below shows how a TCR-α chain is assembed from
the <em>tra</em> locus. V to J recombination joins one of many <code class="docutils literal notranslate"><span class="pre">TRAV</span></code> segments to one of
many <code class="docutils literal notranslate"><span class="pre">TRAJ</span></code> segments. Next, introns are spliced out, resulting in a TCR-α chain
transcript with V, J and C segments directly next to each other (reviewed in <span id="id5">[<a class="reference internal" href="zzz_bibliography.html#id14" title="Meriem Attaf, Eric Huseby, and Andrew K Sewell. $\upalpha $$\upbeta $ t cell receptors as predictors of health and disease. Cellular &amp; Molecular Immunology, 12(4):391–399, January 2015. URL: https://doi.org/10.1038/cmi.2014.134, doi:10.1038/cmi.2014.134.">AHS15</a>]</span>).</p>
<figure class="align-default" id="id18">
<a class="reference internal image-reference" href="_images/vdj.png"><img alt="_images/vdj.png" src="_images/vdj.png" style="width: 600px;" /></a>
<figcaption>
<p><span class="caption-text">Image from <span id="id6">[<a class="reference internal" href="zzz_bibliography.html#id14" title="Meriem Attaf, Eric Huseby, and Andrew K Sewell. $\upalpha $$\upbeta $ t cell receptors as predictors of health and disease. Cellular &amp; Molecular Immunology, 12(4):391–399, January 2015. URL: https://doi.org/10.1038/cmi.2014.134, doi:10.1038/cmi.2014.134.">AHS15</a>]</span> under the <a class="reference external" href="https://creativecommons.org/licenses/by-nc-sa/3.0/">CC BY-NC-SA-3.0</a> license.</span><a class="headerlink" href="#id18" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
</dd>
</dl>
</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="versions.html" class="btn btn-neutral float-right" title="Other versions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="ir-biology.html" class="btn btn-neutral float-left" title="Immune receptor (IR) model" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2021, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>