

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  

  <title>Bibliography &mdash; scirpy  documentation</title>
  


  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "tex2jax_ignore|mathjax_ignore|document", "processClass": "tex2jax_process|mathjax_process|math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/scanpy.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Other versions" href="versions.html" />
 


</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.9.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage-principles.html">Usage principles</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="versions.html">Other versions</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Bibliography</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/icbi-lab/scirpy/blob/master/docs/zzz_bibliography.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="bibliography">
<h1>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline">¶</a></h1>
<p id="id1"><dl class="citation">
<dt class="label" id="id8"><span class="brackets">AYB+17</span></dt>
<dd><p>Shaked Afik, Kathleen B Yates, Kevin Bi, Samuel Darko, Jernej Godec, Ulrike Gerdemann, Leo Swadling, Daniel C Douek, Paul Klenerman, Eleanor J Barnes, Arlene H Sharpe, W Nicholas Haining, and Nir Yosef. Targeted reconstruction of T cell receptor sequence from single cell RNA-seq links CDR3 length to T cell differentiation state. <em>Nucleic Acids Res.</em>, 45(16):e148, September 2017.</p>
</dd>
<dt class="label" id="id14"><span class="brackets">AHS15</span></dt>
<dd><p>Meriem Attaf, Eric Huseby, and Andrew K Sewell. $\upalpha $$\upbeta $ t cell receptors as predictors of health and disease. <em>Cellular &amp; Molecular Immunology</em>, 12(4):391–399, January 2015. URL: <a class="reference external" href="https://doi.org/10.1038/cmi.2014.134">https://doi.org/10.1038/cmi.2014.134</a>, <a class="reference external" href="https://doi.org/10.1038/cmi.2014.134">doi:10.1038/cmi.2014.134</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">BSB10</span></dt>
<dd><p>Brenna L Brady, Natalie C Steinel, and Craig H Bassing. Antigen receptor allelic exclusion: an update and reappraisal. <em>J. Immunol.</em>, 185(7):3801–3808, October 2010.</p>
</dd>
<dt class="label" id="id19"><span class="brackets">CEL+17</span></dt>
<dd><p>Woosung Chung, Hye Hyeon Eum, Hae-Ock Lee, Kyung-Min Lee, Han-Byoel Lee, Kyu-Tae Kim, Han Suk Ryu, Sangmin Kim, Jeong Eon Lee, Yeon Hee Park, Zhengyan Kan, Wonshik Han, and Woong-Yang Park. Single-cell RNA-seq enables comprehensive tumour and immune cell profiling in primary breast cancer. <em>Nature Communications</em>, May 2017. URL: <a class="reference external" href="https://doi.org/10.1038/ncomms15081">https://doi.org/10.1038/ncomms15081</a>, <a class="reference external" href="https://doi.org/10.1038/ncomms15081">doi:10.1038/ncomms15081</a>.</p>
</dd>
<dt class="label" id="id17"><span class="brackets">Dai16</span></dt>
<dd><p>Jeff Daily. Parasail: SIMD c library for global, semi-global, and local pairwise sequence alignments. <em>BMC Bioinformatics</em>, February 2016. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-016-0930-z">https://doi.org/10.1186/s12859-016-0930-z</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-016-0930-z">doi:10.1186/s12859-016-0930-z</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">DFGH+17</span></dt>
<dd><p>Pradyot Dash, Andrew J. Fiore-Gartland, Tomer Hertz, George C. Wang, Shalini Sharma, Aisha Souquette, Jeremy Chase Crawford, E. Bridie Clemens, Thi H. O. Nguyen, Katherine Kedzierska, Nicole L. La Gruta, Philip Bradley, and Paul G. Thomas. Quantifiable predictive features define epitope-specific t cell receptor repertoires. <em>Nature</em>, 547(7661):89–93, June 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nature22383">https://doi.org/10.1038/nature22383</a>, <a class="reference external" href="https://doi.org/10.1038/nature22383">doi:10.1038/nature22383</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">FWST19</span></dt>
<dd><p>David S. Fischer, Yihan Wu, Benjamin Schubert, and Fabian J. Theis. Predicting antigen-specificity of single t-cells based on TCR CDR3 regions. <em>BioRxiv</em>, August 2019. URL: <a class="reference external" href="https://doi.org/10.1101/734053">https://doi.org/10.1101/734053</a>, <a class="reference external" href="https://doi.org/10.1101/734053">doi:10.1101/734053</a>.</p>
</dd>
<dt class="label" id="id13"><span class="brackets">GHN+17</span></dt>
<dd><p>Jacob Glanville, Huang Huang, Allison Nau, Olivia Hatton, Lisa E Wagar, Florian Rubelt, Xuhuai Ji, Arnold Han, Sheri M Krams, Christina Pettus, Nikhil Haas, Cecilia S Lindestam Arlehamn, Alessandro Sette, Scott D Boyd, Thomas J Scriba, Olivia M Martinez, and Mark M Davis. Identifying specificity groups in the T cell receptor repertoire. <em>Nature</em>, 547(7661):94–98, July 2017.</p>
</dd>
<dt class="label" id="id15"><span class="brackets">IKK+16</span></dt>
<dd><p>Tomislav Ilicic, Jong Kyoung Kim, Aleksandra A. Kolodziejczyk, Frederik Otzen Bagger, Davis James McCarthy, John C. Marioni, and Sarah A. Teichmann. Classification of low quality cells from single-cell RNA-seq data. <em>Genome Biology</em>, February 2016. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-016-0888-1">https://doi.org/10.1186/s13059-016-0888-1</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-016-0888-1">doi:10.1186/s13059-016-0888-1</a>.</p>
</dd>
<dt class="label" id="id11"><span class="brackets">JPG10</span></dt>
<dd><p>Qingyong Ji, Antoine Perchellet, and Joan M Goverman. Viral infection triggers central nervous system autoimmunity via activation of CD8+ T cells expressing dual TCRs. <em>Nat. Immunol.</em>, 11(7):628–634, July 2010.</p>
</dd>
<dt class="label" id="id18"><span class="brackets">LLY+20</span></dt>
<dd><p>Mingfeng Liao, Yang Liu, Jing Yuan, Yanling Wen, Gang Xu, Juanjuan Zhao, Lin Cheng, Jinxiu Li, Xin Wang, Fuxiang Wang, Lei Liu, Ido Amit, Shuye Zhang, and Zheng Zhang. Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19. <em>Nature Medicine</em>, May 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41591-020-0901-9">https://doi.org/10.1038/s41591-020-0901-9</a>, <a class="reference external" href="https://doi.org/10.1038/s41591-020-0901-9">doi:10.1038/s41591-020-0901-9</a>.</p>
</dd>
<dt class="label" id="id22"><span class="brackets">LEM+18</span></dt>
<dd><p>Ida Lindeman, Guy Emerton, Lira Mamanova, Omri Snir, Krzysztof Polanski, Shuo-Wang Qiao, Ludvig M. Sollid, Sarah A. Teichmann, and Michael J. T. Stubbington. BraCeR: b-cell-receptor reconstruction and clonality inference from single-cell RNA-seq. <em>Nature Methods</em>, 15(8):563–565, July 2018. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-018-0082-3">https://doi.org/10.1038/s41592-018-0082-3</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-018-0082-3">doi:10.1038/s41592-018-0082-3</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">MMR+20</span></dt>
<dd><p>Ashley Maynard, Caroline E. McCoach, Julia K. Rotow, Lincoln Harris, Franziska Haderk, D. Lucas Kerr, Elizabeth A. Yu, Erin L. Schenk, Weilun Tan, Alexander Zee, Michelle Tan, Philippe Gui, Tasha Lea, Wei Wu, Anatoly Urisman, Kirk Jones, Rene Sit, Pallav K. Kolli, Eric Seeley, Yaron Gesthalter, Daniel D. Le, Kevin A. Yamauchi, David M. Naeger, Sourav Bandyopadhyay, Khyati Shah, Lauren Cech, Nicholas J. Thomas, Anshal Gupta, Mayra Gonzalez, Hien Do, Lisa Tan, Bianca Bacaltos, Rafael Gomez-Sjoberg, Matthew Gubens, Thierry Jahan, Johannes R. Kratz, David Jablons, Norma Neff, Robert C. Doebele, Jonathan Weissman, Collin M. Blakely, Spyros Darmanis, and Trever G. Bivona. Therapy-induced evolution of human lung cancer revealed by single-cell RNA sequencing. <em>Cell</em>, 182(5):1232–1251.e22, September 2020. URL: <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.07.017">https://doi.org/10.1016/j.cell.2020.07.017</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2020.07.017">doi:10.1016/j.cell.2020.07.017</a>.</p>
</dd>
<dt class="label" id="id24"><span class="brackets">SGC+21</span></dt>
<dd><p>Stefan A. Schattgen, Kate Guion, Jeremy Chase Crawford, Aisha Souquette, Alvaro Martinez Barrio, Michael J. T. Stubbington, Paul G. Thomas, and Philip Bradley. Integrating t cell receptor sequences and transcriptional profiles by clonotype neighbor graph analysis (conga). <em>Nature Biotechnology</em>, Aug 2021. URL: <a class="reference external" href="https://doi.org/10.1038/s41587-021-00989-2">https://doi.org/10.1038/s41587-021-00989-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41587-021-00989-2">doi:10.1038/s41587-021-00989-2</a>.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">SB19</span></dt>
<dd><p>Nathaniel J. Schuldt and Bryce A. Binstadt. Dual TCR t cells: identity crisis or multitaskers? <em>The Journal of Immunology</em>, 202(3):637–644, January 2019. URL: <a class="reference external" href="https://doi.org/10.4049/jimmunol.1800904">https://doi.org/10.4049/jimmunol.1800904</a>, <a class="reference external" href="https://doi.org/10.4049/jimmunol.1800904">doi:10.4049/jimmunol.1800904</a>.</p>
</dd>
<dt class="label" id="id16"><span class="brackets">SMR+18</span></dt>
<dd><p>Ian Setliff, Wyatt J. McDonnell, Nagarajan Raju, Robin G. Bombardi, Amyn A. Murji, Cathrine Scheepers, Rutendo Ziki, Charissa Mynhardt, Bryan E. Shepherd, Alusha A. Mamchak, Nigel Garrett, Salim Abdool Karim, Simon A. Mallal, James E. Crowe, Lynn Morris, and Ivelin S. Georgiev. Multi-donor longitudinal antibody repertoire sequencing reveals the existence of public antibody clonotypes in HIV-1 infection. <em>Cell Host &amp; Microbe</em>, 23(6):845–854.e6, June 2018. URL: <a class="reference external" href="https://doi.org/10.1016/j.chom.2018.05.001">https://doi.org/10.1016/j.chom.2018.05.001</a>, <a class="reference external" href="https://doi.org/10.1016/j.chom.2018.05.001">doi:10.1016/j.chom.2018.05.001</a>.</p>
</dd>
<dt class="label" id="id23"><span class="brackets">SRB+21</span></dt>
<dd><p>Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, Natsuhiko Kumasaka, Katarzyna Kania, Justin Engelbert, Bayanne Olabi, Jarmila Stremenova Spegarova, Nicola K Wilson, Nicole Mende, Laura Jardine, Louis CS Gardner, Issac Goh, Dave Horsfall, Jim McGrath, Simone Webb, Michael W. Mather, Rik GH Lindeboom, Emma Dann, Ni Huang, Krzysztof Polanski, Elena Prigmore, Florian Gothe, Jonathan Scott, Rebecca P Payne, Kenneth F Baker, Aidan T Hanrath, Ina CD Schim van der Loeff, Andrew S Barr, Amada Sanchez-Gonzalez, Laura Bergamaschi, Federica Mescia, Josephine L Barnes, Eliz Kilich, Angus de Wilton, Anita Saigal, Aarash Saleh, Sam M Janes, Claire M Smith, Nusayhah Gopee, Caroline Wilson, Paul Coupland, Jonathan M Coxhead, Vladimir Y Kiselev, Stijn van Dongen, Jaume Bacardit, Hamish W King, Anthony J Rostron, A John Simpson, Sophie Hambleton, Elisa Laurenti, Paul A Lyons, Kerstin B Meyer, Marko Z Nikolic, Christopher JA Duncan, Ken Smith, Sarah A Teichmann, Menna R Clatworthy, John C Marioni, Berthold Gottgens, and Muzlifah Haniffa and. Single-cell multi-omics analysis of the immune response in COVID-19. <em>Nature Medicine</em>, April 2021. URL: <a class="reference external" href="https://www.nature.com/articles/s41591-021-01329-2">https://www.nature.com/articles/s41591-021-01329-2</a>, <a class="reference external" href="https://doi.org/10.1038/s41591-021-01329-2">doi:10.1038/s41591-021-01329-2</a>.</p>
</dd>
<dt class="label" id="id12"><span class="brackets">SLonnbergP+16</span></dt>
<dd><p>Michael J T Stubbington, Tapio Lönnberg, Valentina Proserpio, Simon Clare, Anneliese O Speak, Gordon Dougan, and Sarah A Teichmann. T cell fate and clonality inference from single-cell transcriptomes. <em>Nat. Methods</em>, 13(4):329–332, April 2016.</p>
</dd>
<dt class="label" id="id21"><span class="brackets">VKP+06</span></dt>
<dd><p>V. Venturi, K. Kedzierska, D. A. Price, P. C. Doherty, D. C. Douek, S. J. Turner, and M. P. Davenport. Sharing of t cell receptors in antigen-specific responses is driven by convergent recombination. <em>Proceedings of the National Academy of Sciences</em>, 103(49):18691–18696, November 2006. URL: <a class="reference external" href="https://doi.org/10.1073/pnas.0608907103">https://doi.org/10.1073/pnas.0608907103</a>, <a class="reference external" href="https://doi.org/10.1073/pnas.0608907103">doi:10.1073/pnas.0608907103</a>.</p>
</dd>
<dt class="label" id="id3"><span class="brackets">VS10</span></dt>
<dd><p>Christian Vettermann and Mark S. Schlissel. Allelic exclusion of immunoglobulin genes: models and mechanisms. <em>Immunological Reviews</em>, 237(1):22–42, August 2010. URL: <a class="reference external" href="https://doi.org/10.1111/j.1600-065x.2010.00935.x">https://doi.org/10.1111/j.1600-065x.2010.00935.x</a>, <a class="reference external" href="https://doi.org/10.1111/j.1600-065x.2010.00935.x">doi:10.1111/j.1600-065x.2010.00935.x</a>.</p>
</dd>
<dt class="label" id="id20"><span class="brackets">WAY+20</span></dt>
<dd><p>Cédric R Weber, Rahmad Akbar, Alexander Yermanos, Milena Pavlović, Igor Snapkov, Geir K Sandve, Sai T Reddy, and Victor Greiff. immuneSIM: tunable multi-feature simulation of b- and t-cell receptor repertoires for immunoinformatics benchmarking. <em>Bioinformatics</em>, April 2020. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa158">https://doi.org/10.1093/bioinformatics/btaa158</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btaa158">doi:10.1093/bioinformatics/btaa158</a>.</p>
</dd>
<dt class="label" id="id4"><span class="brackets">WMdA+20</span></dt>
<dd><p>Thomas D. Wu, Shravan Madireddi, Patricia E. de Almeida, Romain Banchereau, Ying-Jiun J. Chen, Avantika S. Chitre, Eugene Y. Chiang, Hina Iftikhar, William E. O'Gorman, Amelia Au-Yeung, Chikara Takahashi, Leonard D. Goldstein, Chungkee Poon, Shilpa Keerthivasan, Denise E. de Almeida Nagata, Xiangnan Du, Hyang-Mi Lee, Karl L. Banta, Sanjeev Mariathasan, Meghna Das Thakur, Mahrukh A. Huseni, Marcus Ballinger, Ivette Estay, Patrick Caplazi, Zora Modrusan, Lélia Delamarre, Ira Mellman, Richard Bourgon, and Jane L. Grogan. Peripheral t cell expansion predicts tumour infiltration and clinical response. <em>Nature</em>, 579(7798):274–278, February 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41586-020-2056-8">https://doi.org/10.1038/s41586-020-2056-8</a>, <a class="reference external" href="https://doi.org/10.1038/s41586-020-2056-8">doi:10.1038/s41586-020-2056-8</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets">ZYZ+18</span></dt>
<dd><p>Lei Zhang, Xin Yu, Liangtao Zheng, Yuanyuan Zhang, Yansen Li, Qiao Fang, Ranran Gao, Boxi Kang, Qiming Zhang, Julie Y Huang, Hiroyasu Konno, Xinyi Guo, Yingjiang Ye, Songyuan Gao, Shan Wang, Xueda Hu, Xianwen Ren, Zhanlong Shen, Wenjun Ouyang, and Zemin Zhang. Lineage tracking reveals dynamic relationships of T cells in colorectal cancer. <em>Nature</em>, 564(7735):268–272, December 2018.</p>
</dd>
</dl>
</p>
</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="versions.html" class="btn btn-neutral float-left" title="Other versions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2021, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>