<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <title>Analysis of 3k T cells from cancer &mdash; scirpy  documentation</title>

      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/scanpy.css" type="text/css" />
      <link rel="stylesheet" href="../_static/typehints.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />

  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="API" href="../api.html" />
    <link rel="prev" title="scirpy.io.from_airr_cells" href="../generated/scirpy.io.from_airr_cells.html" />
 


</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.13.0rc2.dev5+gf6ef57e
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../data-structure.html">Data structure and usage principles</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_io.html">Loading adaptive immune receptor repertoire (<span class="xref std std-term">AIRR</span>)-sequencing data with Scirpy</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Analysis of 3k T cells from cancer</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Importing-data">Importing data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preprocessing-Transcriptomics-data">Preprocessing Transcriptomics data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Creating-chain-indices">Creating chain indices</a></li>
<li class="toctree-l3"><a class="reference internal" href="#TCR-Quality-Control">TCR Quality Control</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Define-clonotypes-and-clonotype-clusters">Define clonotypes and clonotype clusters</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Compute-CDR3-neighborhood-graph-and-define-clonotypes">Compute CDR3 neighborhood graph and define clonotypes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Re-compute-CDR3-neighborhood-graph-and-define-clonotype-clusters">Re-compute CDR3 neighborhood graph and define clonotype clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Including-the-V-gene-in-clonotype-definition">Including the V-gene in clonotype definition</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Clonotype-analysis">Clonotype analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Clonal-expansion">Clonal expansion</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Clonotype-abundance">Clonotype abundance</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Convergent-evolution">Convergent evolution</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Gene-usage">Gene usage</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Spectratype-plots">Spectratype plots</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Comparing-repertoires">Comparing repertoires</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Repertoire-simlarity-and-overlaps">Repertoire simlarity and overlaps</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Integrating-gene-expression-data">Integrating gene expression data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Clonotype-modularity">Clonotype modularity</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Clonotype-imbalance-among-cell-clusters">Clonotype imbalance among cell clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Repertoire-overlap-of-cell-types">Repertoire overlap of cell types</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Marker-genes-in-top-clonotypes">Marker genes in top clonotypes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Query-epitope-databases">Query epitope databases</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_docs.html">Developer Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zzz_bibliography.html">Bibliography</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scirpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Analysis of 3k T cells from cancer</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/scverse/scirpy/blob/master/docs/tutorials/tutorial_3k_tcr.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Analysis-of-3k-T-cells-from-cancer">
<h1>Analysis of 3k T cells from cancer<a class="headerlink" href="#Analysis-of-3k-T-cells-from-cancer" title="Permalink to this headline"></a></h1>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This cell is for development only. Don&#39;t copy this to your notebook.</span>
<span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2

<span class="c1"># Temporarily suppress FutureWarnings</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">muon</span> <span class="k">as</span> <span class="nn">mu</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">ir</span>
<span class="kn">from</span> <span class="nn">cycler</span> <span class="kn">import</span> <span class="n">cycler</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">cm</span> <span class="k">as</span> <span class="n">mpl_cm</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.9.3 anndata==0.9.0 umap==0.5.3 numpy==1.23.5 scipy==1.10.1 pandas==1.5.3 scikit-learn==1.2.2 statsmodels==0.13.5 python-igraph==0.10.4 pynndescent==0.5.8
</pre></div></div>
</div>
<p>In this tutorial, we re-analyze single-cell TCR/RNA-seq data from Wu et al. (<span id="id1">[<a class="reference internal" href="../zzz_bibliography.html#id4" title="Thomas D. Wu, Shravan Madireddi, Patricia E. de Almeida, Romain Banchereau, Ying-Jiun J. Chen, Avantika S. Chitre, Eugene Y. Chiang, Hina Iftikhar, William E. O'Gorman, Amelia Au-Yeung, Chikara Takahashi, Leonard D. Goldstein, Chungkee Poon, Shilpa Keerthivasan, Denise E. de Almeida Nagata, Xiangnan Du, Hyang-Mi Lee, Karl L. Banta, Sanjeev Mariathasan, Meghna Das Thakur, Mahrukh A. Huseni, Marcus Ballinger, Ivette Estay, Patrick Caplazi, Zora Modrusan, Lélia Delamarre, Ira Mellman, Richard Bourgon, and Jane L. Grogan. Peripheral t cell expansion predicts tumour infiltration and clinical response. Nature, 579(7798):274–278, February 2020. URL: https://doi.org/10.1038/s41586-020-2056-8, doi:10.1038/s41586-020-2056-8.">WMdA+20</a>]</span>)
generated on the 10x Genomics platform. The original dataset consists of &gt;140k T cells
from 14 treatment-naive patients across four different types of cancer.
For this tutorial, to speed up computations, we use a downsampled version of 3k cells.</p>
<section id="Importing-data">
<h2>Importing data<a class="headerlink" href="#Importing-data" title="Permalink to this headline"></a></h2>
<p><cite>Scirpy</cite> natively supports reading <a class="reference internal" href="../glossary.html#term-IR"><span class="xref std std-term">IR</span></a> data from <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cellranger</a> (10x), <a class="reference external" href="https://github.com/Teichlab/tracer">TraCeR</a> (Smart-seq2)
or the <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR rearrangement schema</a> and provides helper functions to import other data types. We provide a <a class="reference internal" href="tutorial_io.html#importing-data"><span class="std std-ref">dedicated tutorial on data loading</span></a> with more details.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p><strong>scirpy’s data structure has been updated in v0.13.0.</strong></p>
<p>Previously, receptor data was expanded into columns of <cite>adata.obs</cite>, now they are stored as an <a class="reference internal" href="../glossary.html#term-awkward-array"><span class="xref std std-term">awkward array</span></a> in <cite>adata.obsm[“airr”]</cite>.
Moreover, we now use <a class="reference external" href="https://mudata.readthedocs.io/en/latest/api/generated/mudata.MuData.html#mudata.MuData" title="(in mudata)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a> to handle paired transcriptomics and <a class="reference internal" href="../glossary.html#term-AIRR"><span class="xref std std-term">AIRR</span></a> data.</p>
<p><cite>AnnData</cite> objects created with older versions of scirpy can be upgraded with <a class="reference internal" href="../generated/scirpy.io.upgrade_schema.html#scirpy.io.upgrade_schema" title="scirpy.io.upgrade_schema"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.io.upgrade_schema()</span></code></a> to be compatible with the latest version of scirpy.</p>
<p>Please check out</p>
<blockquote>
<div><ul class="simple">
<li><p>The <a class="reference external" href="https://github.com/scverse/scirpy/releases/tag/v0.13.0">release notes</a> for details about the changes</p></li>
<li><p>The documentation about <a class="reference internal" href="../data-structure.html#data-structure"><span class="std std-ref">Scirpy’s data structure</span></a> and</p></li>
<li><p><a class="reference internal" href="../ir-biology.html#receptor-model"><span class="std std-ref">Scirpy’s receptor model</span></a>.</p></li>
</ul>
</div></blockquote>
</div>
<p>The example dataset for this tutorial ships with the <cite>scirpy</cite> package. We can conveniently load it from the <a class="reference internal" href="../api.html#module-scirpy.datasets" title="scirpy.datasets"><code class="xref py py-mod docutils literal notranslate"><span class="pre">datasets</span></code></a> module:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">wu2020_3k</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Downloading file &#39;wu2020_3k.h5mu&#39; from &#39;doi:10.6084/m9.figshare.22249894.v1/wu2020_3k.h5mu&#39; to &#39;C:\Users\runneradmin\AppData\Local\scirpy\scirpy\Cache\main&#39;.
100%|#############################################| 17.3M/17.3M [00:00&lt;?, ?B/s]
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\aligned_mapping.py:54: ExperimentalFeatureWarning: Support for Awkward Arrays is currently experimental. Behavior may change in the future. Please report any issues you may encounter!
  warnings.warn(
</pre></div></div>
</div>
<p><cite>mdata</cite> is a <a class="reference external" href="https://mudata.readthedocs.io/en/latest/api/generated/mudata.MuData.html#mudata.MuData" title="(in mudata)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a> object with two modalities: <cite>gex</cite> for gene expression and <cite>airr</cite> for immune-receptor data.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 3000 × 30727
  2 modalities
    gex:    3000 x 30727
      obs:  &#x27;cluster_orig&#x27;, &#x27;patient&#x27;, &#x27;sample&#x27;, &#x27;source&#x27;
      uns:  &#x27;cluster_orig_colors&#x27;
      obsm: &#x27;X_umap_orig&#x27;
    airr:   3000 x 0
      obs:  &#x27;high_confidence&#x27;, &#x27;is_cell&#x27;, &#x27;clonotype_orig&#x27;
      obsm: &#x27;airr&#x27;</pre></div>
</div>
</section>
<section id="Preprocessing-Transcriptomics-data">
<h2>Preprocessing Transcriptomics data<a class="headerlink" href="#Preprocessing-Transcriptomics-data" title="Permalink to this headline"></a></h2>
<p>Gene expression (GEX) data needs to be filtered and preprocessed as with any other single-cell dataset. For this tutorial, we perform minimal preprocessing for demonstration purposes. For a real dataset, we recommend following the <a class="reference external" href="https://scverse.org/learn/">scverse tutorials</a> and the <a class="reference external" href="https://sc-best-practices.org/">single-cell best practice book</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
filtered out 18877 genes that are detected in less than 10 cells
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_per_cell</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
normalizing by total count per cell
    finished (0:00:00): normalized adata.X and added    &#39;n_counts&#39;, counts per cell before normalization (adata.obs)
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:01)
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:01)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
    finished (0:00:03)
</pre></div></div>
</div>
<p>For the <em>Wu2020</em> dataset, the authors already provide clusters and UMAP coordinates. Instead of performing clustering and cluster annotation ourselves, we will just use the provided data. The clustering and annotation methodology is described in <a class="reference external" href="https://doi.org/10.1038/s41586-020-2056-8">their paper</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap_orig&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mapping</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;nan&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
    <span class="s2">&quot;3.1-MT&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.1-Trm&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_Trm&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.2-RPL32&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_RPL32&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.3-TCF7&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_TCF7&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.4-FOS&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_FOSS&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.5-IL6ST&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_IL6ST&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.6a-Treg&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_Treg&quot;</span><span class="p">,</span>
    <span class="s2">&quot;4.6b-Treg&quot;</span><span class="p">:</span> <span class="s2">&quot;CD4_Treg&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.1-Teff&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8_Teff&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.2-Tem&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8_Tem&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.3a-Trm&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8_Trm&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.3b-Trm&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8_Trm&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.3c-Trm&quot;</span><span class="p">:</span> <span class="s2">&quot;CD8_Trm&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.4-Chrom&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.5-Mitosis&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
    <span class="s2">&quot;8.6-KLRB1&quot;</span><span class="p">:</span> <span class="s2">&quot;other&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cluster&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cluster_orig&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">mapping</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now that we filtered obs and var of the GEX modality, we need to propagate those changes back to the <code class="docutils literal notranslate"><span class="pre">MuData</span></code> object.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">update</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Let’s inspect the UMAP plots. The first three panels show the UMAP plot colored by sample, patient and cluster. We don’t observe any obvious clustering of samples or patients that could hint at batch effects.</p>
<p>The last three panels show the UMAP colored by the T cell markers <em>CD8</em>, <em>CD4</em>, and <em>FOXP3</em>. We can confirm that the markers correspond to their respective cluster labels.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span> <span class="s2">&quot;gex:patient&quot;</span><span class="p">,</span> <span class="s2">&quot;gex:cluster&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD8A&quot;</span><span class="p">,</span> <span class="s2">&quot;CD4&quot;</span><span class="p">,</span> <span class="s2">&quot;FOXP3&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span>
    <span class="n">wspace</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_19_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_19_1.png" style="width: 1420px; height: 316px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_19_2.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_19_2.png" style="width: 1349px; height: 299px;" />
</div>
</div>
</section>
<section id="Creating-chain-indices">
<h2>Creating chain indices<a class="headerlink" href="#Creating-chain-indices" title="Permalink to this headline"></a></h2>
<p>As a first step, we need to create <a class="reference internal" href="../data-structure.html#chain-indices"><span class="std std-ref">chain indices</span></a> for the receptor data.</p>
<p>By importing <a class="reference internal" href="../glossary.html#term-IR"><span class="xref std std-term">IR</span></a> data, essentially, for each cell a list of receptor chains is stored in <cite>adata.obsm[“airr”]</cite>.
The <a class="reference internal" href="../ir-biology.html#receptor-model"><span class="std std-ref">scirpy receptor model</span></a> allows up to two pairs of chains per cell. This representation
requires separation of chains by <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">locus</span></a> into <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ</span></a> and <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VDJ</span></a> chains,
and (optionally) filtering non-productive chains.</p>
<p>The <a class="reference internal" href="../generated/scirpy.pp.index_chains.html#scirpy.pp.index_chains" title="scirpy.pp.index_chains"><code class="xref py py-func docutils literal notranslate"><span class="pre">index_chains()</span></code></a> function applies the scirpy receptor model by storing references to the
selected chains in <cite>adata.obsm[“chain_indices”]</cite>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3000/3000 [00:04&lt;00:00, 732.47it/s]
</pre></div></div>
</div>
</section>
<section id="TCR-Quality-Control">
<h2>TCR Quality Control<a class="headerlink" href="#TCR-Quality-Control" title="Permalink to this headline"></a></h2>
<p>While most of T cell receptors have exactly one pair of α and β chains, up to one third of
T cells can have <a class="reference internal" href="../glossary.html#term-Dual-IR"><span class="xref std std-term">dual TCRs</span></a>, i.e. two pairs of receptors originating from different alleles (<span id="id2">[<a class="reference internal" href="../zzz_bibliography.html#id5" title="Nathaniel J. Schuldt and Bryce A. Binstadt. Dual TCR t cells: identity crisis or multitaskers? The Journal of Immunology, 202(3):637–644, January 2019. URL: https://doi.org/10.4049/jimmunol.1800904, doi:10.4049/jimmunol.1800904.">SB19</a>]</span>).</p>
<p>Using the <a class="reference internal" href="../generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a> function, we can add a summary
about the Immune cell-receptor compositions to <cite>adata.obs</cite>. We can visualize it using <a class="reference internal" href="../generated/scirpy.pl.group_abundance.html#scirpy.pl.group_abundance" title="scirpy.pl.group_abundance"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.group_abundance()</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>chain pairing</strong></p>
<ul class="simple">
<li><p><em>Orphan chain</em> refers to cells that have either a single alpha or beta receptor chain.</p></li>
<li><p><em>Extra chain</em> refers to cells that have a full alpha/beta receptor pair, and an additional chain.</p></li>
<li><p><a class="reference internal" href="../glossary.html#term-Multichain-cell"><span class="xref std std-term">Multichain</span></a> refers to cells with more than two receptor pairs detected.
These cells are likely doublets.</p></li>
</ul>
<p><strong>receptor type and receptor subtype</strong></p>
<ul class="simple">
<li><p><cite>receptor_type</cite> refers to a coarse classification into <cite>BCR</cite> and <cite>TCR</cite>. Cells with both <cite>BCR</cite> and <cite>TCR</cite> chains
are labelled as <cite>ambiguous</cite>.</p></li>
<li><p><cite>receptor_subtype</cite> refers to a more specific classification into α/β, ɣ/δ, IG-λ, and IG-κ chain configurations.</p></li>
</ul>
<p>For more details, see <a class="reference internal" href="../generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a>.</p>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_type&#34;]`.
Stored result in `mdata.obs[&#34;airr:receptor_subtype&#34;]`.
Stored result in `mdata.obs[&#34;airr:chain_pairing&#34;]`.
</pre></div></div>
</div>
<p>As expected, the dataset contains only α/β T-cell receptors:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:receptor_subtype&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:source&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_27_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_27_0.png" style="width: 537px; height: 346px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:source&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_28_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_28_0.png" style="width: 537px; height: 358px;" />
</div>
</div>
<p>Indeed, in this dataset, ~7% of cells have more than one pair of productive T-cell receptors:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span>
    <span class="s2">&quot;Fraction of cells with more than one pair of TCRs: </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
            <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="p">[</span><span class="s2">&quot;extra VJ&quot;</span><span class="p">,</span> <span class="s2">&quot;extra VDJ&quot;</span><span class="p">,</span> <span class="s2">&quot;two full chains&quot;</span><span class="p">,</span> <span class="s2">&quot;multichain&quot;</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="o">/</span> <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;airr&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">n_obs</span>
    <span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Fraction of cells with more than one pair of TCRs: 0.07
</pre></div></div>
</div>
<p>Next, we visualize the <a class="reference internal" href="../glossary.html#term-Multichain-cell"><span class="xref std std-term">Multichain-cells</span></a> on the UMAP plot and exclude them from downstream analysis:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s2">&quot;multichain&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_32_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_32_1.png" style="width: 380px; height: 296px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">!=</span> <span class="s2">&quot;multichain&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Similarly, we can use the <code class="docutils literal notranslate"><span class="pre">chain_pairing</span></code> information to exclude all cells that don’t have at least one full pair of receptor sequences:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_obs</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;orphan VDJ&quot;</span><span class="p">,</span> <span class="s2">&quot;orphan VJ&quot;</span><span class="p">])</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 1845 × 11850
  2 modalities
    gex:    1845 x 11850
      obs:  &#x27;cluster_orig&#x27;, &#x27;patient&#x27;, &#x27;sample&#x27;, &#x27;source&#x27;, &#x27;n_genes&#x27;, &#x27;n_counts&#x27;, &#x27;cluster&#x27;
      var:  &#x27;n_cells&#x27;, &#x27;highly_variable&#x27;, &#x27;means&#x27;, &#x27;dispersions&#x27;, &#x27;dispersions_norm&#x27;
      uns:  &#x27;cluster_orig_colors&#x27;, &#x27;log1p&#x27;, &#x27;hvg&#x27;, &#x27;pca&#x27;, &#x27;neighbors&#x27;, &#x27;gex:sample_colors&#x27;, &#x27;gex:patient_colors&#x27;, &#x27;gex:cluster_colors&#x27;, &#x27;airr:chain_pairing_colors&#x27;
      obsm: &#x27;X_umap_orig&#x27;, &#x27;X_pca&#x27;, &#x27;X_umap&#x27;
      varm: &#x27;PCs&#x27;
      obsp: &#x27;distances&#x27;, &#x27;connectivities&#x27;
    airr:   1845 x 0
      obs:  &#x27;high_confidence&#x27;, &#x27;is_cell&#x27;, &#x27;clonotype_orig&#x27;, &#x27;receptor_type&#x27;, &#x27;receptor_subtype&#x27;, &#x27;chain_pairing&#x27;
      uns:  &#x27;chain_indices&#x27;
      obsm: &#x27;airr&#x27;, &#x27;chain_indices&#x27;</pre></div>
</div>
<p>Finally, we re-create the chain-pairing plot to ensure that the filtering worked as expected:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ax</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:chain_pairing&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:source&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_38_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_38_0.png" style="width: 537px; height: 358px;" />
</div>
</div>
</section>
<section id="Define-clonotypes-and-clonotype-clusters">
<h2>Define clonotypes and clonotype clusters<a class="headerlink" href="#Define-clonotypes-and-clonotype-clusters" title="Permalink to this headline"></a></h2>
<p id="define-clonotypes-tutorial">In this section, we will define and visualize <a class="reference internal" href="../glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> and <a class="reference internal" href="../glossary.html#term-Clonotype-cluster"><span class="xref std std-term">clonotype clusters</span></a>.</p>
<p><em>Scirpy</em> implements a network-based approach for clonotype definition. The steps to create and visualize the clonotype-network are analogous to the construction of a neighborhood graph from transcriptomics data with <em>Scanpy</em>.</p>
<table class="colwidths-given docutils align-default" id="id4">
<caption><span class="caption-text">Analysis steps on IR data</span><a class="headerlink" href="#id4" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 40%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>scirpy function</p></th>
<th class="head"><p>objective</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pp.ir_dist()</span></code></a></p></td>
<td><p>Compute sequence-based distance matrices for all <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">VJ</span></a> and
<a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">VDJ</span></a> sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotypes()</span></code></a></p></td>
<td><p>Define <a class="reference internal" href="../glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> by nucleotide
sequence identity.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.tl.define_clonotype_clusters.html#scirpy.tl.define_clonotype_clusters" title="scirpy.tl.define_clonotype_clusters"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotype_clusters()</span></code></a></p></td>
<td><p>Cluster cells by the similarity of their CDR3-sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.tl.clonotype_network.html#scirpy.tl.clonotype_network" title="scirpy.tl.clonotype_network"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.clonotype_network()</span></code></a></p></td>
<td><p>Compute layout of the clonotype network.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.pl.clonotype_network.html#scirpy.pl.clonotype_network" title="scirpy.pl.clonotype_network"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.clonotype_network()</span></code></a></p></td>
<td><p>Plot clonotype network colored by different parameters.</p></td>
</tr>
</tbody>
</table>
<section id="Compute-CDR3-neighborhood-graph-and-define-clonotypes">
<h3>Compute CDR3 neighborhood graph and define clonotypes<a class="headerlink" href="#Compute-CDR3-neighborhood-graph-and-define-clonotypes" title="Permalink to this headline"></a></h3>
<p><a class="reference internal" href="../generated/scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pp.ir_dist()</span></code></a> computes distances between <a class="reference internal" href="../glossary.html#term-CDR"><span class="xref std std-term">CDR3</span></a> nucleotide (<cite>nt</cite>) or amino acid (<cite>aa</cite>) sequences, either based on sequence identity or similarity. It creates two distance matrices: one for all unique <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">VJ</span></a> sequences and one for all unique <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">VDJ</span></a> sequences. The distance matrices are added to <cite>adata.uns</cite>.</p>
<p>The function <a class="reference internal" href="../generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotypes()</span></code></a> matches cells based on the distances of their
<cite>VJ</cite> and <cite>VDJ</cite> CDR3-sequences and value of the function parameters <cite>dual_ir</cite> and <cite>receptor_arms</cite>. Finally, it
detects connected modules in the graph and annotates them as clonotypes. This will add a <cite>clone_id</cite> and
<cite>clone_id_size</cite> column to <cite>adata.obs</cite>.</p>
<p>The <cite>dual_ir</cite> parameter defines how scirpy handles cells with <a class="reference internal" href="../glossary.html#term-Dual-IR"><span class="xref std std-term">more than one pair of receptors</span></a>. The default value is <cite>any</cite> which implies that cells with any of their primary or secondary receptor chain matching will be considered to be of the same clonotype.</p>
<p>Here, we define <a class="reference internal" href="../glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> based on nt-sequence identity.
In a later step, we will define <a class="reference internal" href="../glossary.html#term-Clonotype-cluster"><span class="xref std std-term">clonotype clusters</span></a> based on
amino-acid similarity.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># using default parameters, `ir_dist` will compute nucleotide sequence identity</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotypes</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VJ sequences.
Computing sequence x sequence distance matrix for VDJ sequences.
Initializing lookup tables.
Computing clonotype x clonotype distances.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 1526/1526 [00:02&lt;00:00, 659.46it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:clone_id&#34;]`.
Stored result in `mdata.obs[&#34;airr:clone_id_size&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p>To visualize the network we first call <a class="reference internal" href="../generated/scirpy.tl.clonotype_network.html#scirpy.tl.clonotype_network" title="scirpy.tl.clonotype_network"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.clonotype_network()</span></code></a> to compute the layout.
We can then visualize it using <a class="reference internal" href="../generated/scirpy.pl.clonotype_network.html#scirpy.pl.clonotype_network" title="scirpy.pl.clonotype_network"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.clonotype_network()</span></code></a>. We recommend setting the
<cite>min_cells</cite> parameter to <cite>&gt;=2</cite>, to prevent the singleton clonotypes from cluttering the network.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The resulting plot is a network, where each dot represents cells with identical
receptor configurations. As we define <a class="reference internal" href="../glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> as cells with identical CDR3-sequences, each
dot is also a clonotype. For each clonotype, the numeric clonotype id is shown in the graph.
The size of each dot refers to the number of cells with the same receptor configurations.
Categorical variables can be visualized as pie charts.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;gex:source&quot;</span><span class="p">,</span> <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
gex:source
Blood    107
NAT      756
Tumor    982
dtype: int64
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gex:source&quot;</span><span class="p">,</span> <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">label_fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_48_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_48_0.png" style="width: 642px; height: 482px;" />
</div>
</div>
</section>
<section id="Re-compute-CDR3-neighborhood-graph-and-define-clonotype-clusters">
<h3>Re-compute CDR3 neighborhood graph and define clonotype clusters<a class="headerlink" href="#Re-compute-CDR3-neighborhood-graph-and-define-clonotype-clusters" title="Permalink to this headline"></a></h3>
<p>We can now re-compute the clonotype network based on amino-acid sequence similarity
and define <a class="reference internal" href="../glossary.html#term-Clonotype-cluster"><span class="xref std std-term">clonotype clusters</span></a>.</p>
<p>To this end, we need to change set <cite>metric=”alignment”</cite> and specify a <cite>cutoff</cite> parameter.
The distance is based on the <a class="reference external" href="https://en.wikipedia.org/wiki/BLOSUM">BLOSUM62</a> matrix.
For instance, a distance of <cite>10</cite> is equivalent to 2 Rs mutating into N.
This appoach was initially proposed as <em>TCRdist</em> by Dash et al. (<span id="id3">[<a class="reference internal" href="../zzz_bibliography.html#id6" title="Pradyot Dash, Andrew J. Fiore-Gartland, Tomer Hertz, George C. Wang, Shalini Sharma, Aisha Souquette, Jeremy Chase Crawford, E. Bridie Clemens, Thi H. O. Nguyen, Katherine Kedzierska, Nicole L. La Gruta, Philip Bradley, and Paul G. Thomas. Quantifiable predictive features define epitope-specific t cell receptor repertoires. Nature, 547(7661):89–93, June 2017. URL: https://doi.org/10.1038/nature22383, doi:10.1038/nature22383.">DFGH+17</a>]</span>).</p>
<p>All cells with a distance between their CDR3 sequences lower than <cite>cutoff</cite> will be connected in the network.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">cutoff</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VJ sequences.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 496/496 [00:35&lt;00:00, 14.03it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VDJ sequences.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 496/496 [00:35&lt;00:00, 13.91it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotype_clusters</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span> <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;any&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing lookup tables.
Computing clonotype x clonotype distances.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 1549/1549 [00:06&lt;00:00, 225.55it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:cc_aa_alignment&#34;]`.
Stored result in `mdata.obs[&#34;airr:cc_aa_alignment_size&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Compared to the previous plot, we observere several connected dots. Each fully connected subnetwork represents a “clonotype cluster”, each dot still represents cells with identical receptor configurations.</p>
<p>The dots are colored by patient. We observe, that for instance, clonotypes <code class="docutils literal notranslate"><span class="pre">101</span></code> and <code class="docutils literal notranslate"><span class="pre">68</span></code> (left top and bottom) are <em>private</em>, i.e. they contain cells from a single patient only. On the other hand, clonotype <code class="docutils literal notranslate"><span class="pre">159</span></code> (left middle) is <em>public</em>, i.e. it is shared across patients <em>Lung1</em> and <em>Lung3</em>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gex:patient&quot;</span><span class="p">,</span> <span class="n">label_fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_55_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_55_0.png" style="width: 652px; height: 479px;" />
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since v0.13, AIRR data is not by default included in <cite>obs</cite>. We can retrieve a per-cell data frame with AIRR information
using <a class="reference internal" href="../generated/scirpy.get.airr.html#scirpy.get.airr" title="scirpy.get.airr"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.get.airr()</span></code></a>, or temporarily add columns to <cite>obs</cite> using the <a class="reference internal" href="../generated/scirpy.get.airr_context.html#scirpy.get.airr_context" title="scirpy.get.airr_context"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.get.airr_context()</span></code></a> context
manager. For more information, see <a class="reference internal" href="../data-structure.html#accessing-airr-data"><span class="std std-ref">Accessing AIRR data</span></a>.</p>
</div>
<p>We can now extract information (e.g. CDR3-sequences) from a specific clonotype cluster by subsetting <cite>MuData</cite>.
By extracting the CDR3 sequences of clonotype cluster <cite>159</cite>, we retreive five different receptor configurations with different numbers of cells, corresponding to the five points in the graph.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;junction_aa&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;VJ_1&quot;</span><span class="p">,</span> <span class="s2">&quot;VDJ_1&quot;</span><span class="p">,</span> <span class="s2">&quot;VJ_2&quot;</span><span class="p">,</span> <span class="s2">&quot;VDJ_2&quot;</span><span class="p">]):</span>
    <span class="n">cdr3_ct_159</span> <span class="o">=</span> <span class="p">(</span>
        <span class="c1"># TODO astype(str) is required due to a bug in pandas ignoring `dropna=False`. It seems fixed in pandas 2.x</span>
        <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;159&quot;</span><span class="p">]</span>
        <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">.</span><span class="n">groupby</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="s2">&quot;VJ_1_junction_aa&quot;</span><span class="p">,</span>
                <span class="s2">&quot;VDJ_1_junction_aa&quot;</span><span class="p">,</span>
                <span class="s2">&quot;VJ_2_junction_aa&quot;</span><span class="p">,</span>
                <span class="s2">&quot;VDJ_2_junction_aa&quot;</span><span class="p">,</span>
                <span class="s2">&quot;airr:receptor_subtype&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">dropna</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;n_cells&quot;</span><span class="p">)</span>
    <span class="p">)</span>
<span class="n">cdr3_ct_159</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>VJ_1_junction_aa</th>
      <th>VDJ_1_junction_aa</th>
      <th>VJ_2_junction_aa</th>
      <th>VDJ_2_junction_aa</th>
      <th>airr:receptor_subtype</th>
      <th>n_cells</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CAGKAGNTGKLIF</td>
      <td>CASSYQGSTEAFF</td>
      <td>nan</td>
      <td>nan</td>
      <td>TRA+TRB</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>CAGKSGNTGKLIF</td>
      <td>CASSYQGATEAFF</td>
      <td>CATDPRRSTGNQFYF</td>
      <td>nan</td>
      <td>TRA+TRB</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>CAGKSGNTGKLIF</td>
      <td>CASSYQGATEAFF</td>
      <td>nan</td>
      <td>nan</td>
      <td>TRA+TRB</td>
      <td>12</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CATDPRRSTGNQFYF</td>
      <td>CASSYQGATEAFF</td>
      <td>CAGKSGNTGKLIF</td>
      <td>nan</td>
      <td>TRA+TRB</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CATDPRRSTGNQFYF</td>
      <td>CASSYQGATEAFF</td>
      <td>nan</td>
      <td>nan</td>
      <td>TRA+TRB</td>
      <td>11</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Including-the-V-gene-in-clonotype-definition">
<h3>Including the V-gene in clonotype definition<a class="headerlink" href="#Including-the-V-gene-in-clonotype-definition" title="Permalink to this headline"></a></h3>
<p>Using the paramter <cite>use_v_gene</cite> in <a class="reference internal" href="../generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">define_clonotypes()</span></code></a>, we can enforce
clonotypes (or clonotype clusters) to have the same <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">V-gene</span></a>, and, therefore, the same <a class="reference internal" href="../glossary.html#term-CDR"><span class="xref std std-term">CDR1 and 2</span></a>
regions. Let’s look for clonotype clusters with different V genes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotype_clusters</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;all&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;any&quot;</span><span class="p">,</span>
    <span class="n">same_v_gene</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;cc_aa_alignment_same_v&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing lookup tables.
Computing clonotype x clonotype distances.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 1549/1549 [00:07&lt;00:00, 199.31it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:cc_aa_alignment_same_v&#34;]`.
Stored result in `mdata.obs[&#34;airr:cc_aa_alignment_same_v_size&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># find clonotypes with more than one `clonotype_same_v`</span>
<span class="n">ct_different_v</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;airr:cc_aa_alignment_same_v&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span>
<span class="p">)</span>
<span class="n">ct_different_v</span> <span class="o">=</span> <span class="n">ct_different_v</span><span class="p">[</span><span class="n">ct_different_v</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">ct_different_v</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;280&#39;, &#39;765&#39;]
</pre></div></div>
</div>
<p>Here, we see that the clonotype clusters <code class="docutils literal notranslate"><span class="pre">280</span></code> and <code class="docutils literal notranslate"><span class="pre">765</span></code> get split into <code class="docutils literal notranslate"><span class="pre">(280,</span> <span class="pre">788)</span></code> and <code class="docutils literal notranslate"><span class="pre">(765,</span> <span class="pre">1071)</span></code>, respectively, when the <code class="docutils literal notranslate"><span class="pre">same_v_gene</span></code> flag is set.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;v_call&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;VJ_1&quot;</span><span class="p">,</span> <span class="s2">&quot;VDJ_1&quot;</span><span class="p">]):</span>
    <span class="n">ct_different_v_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ct_different_v</span><span class="p">),</span>
            <span class="p">[</span>
                <span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">,</span>
                <span class="s2">&quot;airr:cc_aa_alignment_same_v&quot;</span><span class="p">,</span>
                <span class="s2">&quot;VJ_1_v_call&quot;</span><span class="p">,</span>
                <span class="s2">&quot;VDJ_1_v_call&quot;</span><span class="p">,</span>
            <span class="p">],</span>
        <span class="p">]</span>
        <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="p">)</span>
<span class="n">ct_different_v_df</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>airr:cc_aa_alignment</th>
      <th>airr:cc_aa_alignment_same_v</th>
      <th>VJ_1_v_call</th>
      <th>VDJ_1_v_call</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>280</td>
      <td>280</td>
      <td>TRAV8-6</td>
      <td>TRBV6-6</td>
    </tr>
    <tr>
      <th>1</th>
      <td>280</td>
      <td>788</td>
      <td>TRAV8-3</td>
      <td>TRBV9</td>
    </tr>
    <tr>
      <th>2</th>
      <td>765</td>
      <td>765</td>
      <td>TRAV21</td>
      <td>TRBV6-6</td>
    </tr>
    <tr>
      <th>3</th>
      <td>765</td>
      <td>1071</td>
      <td>TRAV21</td>
      <td>TRBV6-5</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>
<section id="Clonotype-analysis">
<h2>Clonotype analysis<a class="headerlink" href="#Clonotype-analysis" title="Permalink to this headline"></a></h2>
<section id="Clonal-expansion">
<h3>Clonal expansion<a class="headerlink" href="#Clonal-expansion" title="Permalink to this headline"></a></h3>
<p>Let’s visualize the number of expanded clonotypes (i.e. clonotypes consisting
of more than one cell) by cell-type. The first option is to add a column with the <a class="reference internal" href="../generated/scirpy.tl.clonal_expansion.html#scirpy.tl.clonal_expansion" title="scirpy.tl.clonal_expansion"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.clonal_expansion()</span></code></a>
to <cite>adata.obs</cite> and overlay it on the UMAP plot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonal_expansion</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:clonal_expansion&#34;]`.
</pre></div></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">clonal_expansion</span></code> refers to expansion categories, i.e singleton clonotypes, clonotypes with 2 cells and more than 2 cells. The <code class="docutils literal notranslate"><span class="pre">clonotype_size</span></code> refers to the absolute number of cells in a clonotype.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;airr:clonal_expansion&quot;</span><span class="p">,</span> <span class="s2">&quot;airr:clone_id_size&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_68_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_68_1.png" style="width: 660px; height: 299px;" />
</div>
</div>
<p>The second option is to show the number of cells belonging to an expanded clonotype per category
in a stacked bar plot, using the <a class="reference internal" href="../generated/scirpy.pl.clonal_expansion.html#scirpy.pl.clonal_expansion" title="scirpy.pl.clonal_expansion"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.clonal_expansion()</span></code></a> plotting function.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonal_expansion</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;clone_id&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span> <span class="n">clip_at</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_70_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_70_0.png" style="width: 517px; height: 312px;" />
</div>
</div>
<p>The same plot, normalized to cluster size. Clonal expansion is a sign of positive selection for a certain, reactive T-cell clone. It, therefore, makes sense that CD8+ effector T-cells have the largest fraction of expanded clonotypes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonal_expansion</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;clone_id&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: &gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_72_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_72_1.png" style="width: 517px; height: 312px;" />
</div>
</div>
<p>Expectedly, the CD8+ effector T cells have the largest fraction of expanded clonotypes.</p>
<p>Consistent with this observation, they have the lowest <a class="reference internal" href="../generated/scirpy.pl.alpha_diversity.html#scirpy.pl.alpha_diversity" title="scirpy.pl.alpha_diversity"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.alpha_diversity()</span></code></a> of clonotypes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">alpha_diversity</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;normalized_shannon_entropy&quot;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_74_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_74_0.png" style="width: 396px; height: 334px;" />
</div>
</div>
</section>
<section id="Clonotype-abundance">
<h3>Clonotype abundance<a class="headerlink" href="#Clonotype-abundance" title="Permalink to this headline"></a></h3>
<p>The function <a class="reference internal" href="../generated/scirpy.pl.group_abundance.html#scirpy.pl.group_abundance" title="scirpy.pl.group_abundance"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.group_abundance()</span></code></a> allows us to create bar charts for
arbitrary categorial from <cite>obs</cite>. Here, we use it to show the distribution of the
ten largest clonotypes across the cell-type clusters.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span> <span class="n">max_cols</span><span class="o">=</span><span class="mi">10</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_77_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_77_0.png" style="width: 559px; height: 327px;" />
</div>
</div>
<p>It might be beneficial to normalize the counts to the number of cells per sample to mitigate biases due to different sample sizes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">,</span>
    <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span>
    <span class="n">max_cols</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">normalize</span><span class="o">=</span><span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_79_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_79_0.png" style="width: 563px; height: 327px;" />
</div>
</div>
<p>Coloring the bars by patient gives us information about public and private clonotypes: Some clonotypes are <em>private</em>, i.e. specific to a certain tissue, others are <em>public</em>, i.e. they are shared across different tissues.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:source&quot;</span><span class="p">,</span> <span class="n">max_cols</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_81_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_81_0.png" style="width: 709px; height: 365px;" />
</div>
</div>
<p>However, clonotypes that are shared between <em>patients</em> are rare:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">,</span>
    <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:patient&quot;</span><span class="p">,</span>
    <span class="n">max_cols</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_83_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_83_0.png" style="width: 713px; height: 365px;" />
</div>
</div>
</section>
<section id="Convergent-evolution">
<h3>Convergent evolution<a class="headerlink" href="#Convergent-evolution" title="Permalink to this headline"></a></h3>
<p>By comparing two levels of clonotype definitions (e.g. based on nucleotide sequences and based on amino-acid sequences),
we can identify receptors that are subject to <a class="reference internal" href="../glossary.html#term-Convergent-evolution-of-clonotypes"><span class="xref std std-term">convergent evolution</span></a>. By that,
we mean receptors that (likely) recognize the same antigen but have evolved from different clones.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_convergence</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">key_coarse</span><span class="o">=</span><span class="s2">&quot;cc_aa_alignment&quot;</span><span class="p">,</span> <span class="n">key_fine</span><span class="o">=</span><span class="s2">&quot;clone_id&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:is_convergent&#34;]`.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:is_convergent&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_87_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_87_1.png" style="width: 410px; height: 296px;" />
</div>
</div>
</section>
</section>
<section id="Gene-usage">
<h2>Gene usage<a class="headerlink" href="#Gene-usage" title="Permalink to this headline"></a></h2>
<p><a class="reference internal" href="../generated/scirpy.tl.group_abundance.html#scirpy.tl.group_abundance" title="scirpy.tl.group_abundance"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.group_abundance()</span></code></a> can also give us some information on VDJ usage.
We first add information about <cite>v_call</cite> to <cite>obs</cite> using <a class="reference internal" href="../generated/scirpy.get.airr_context.html#scirpy.get.airr_context" title="scirpy.get.airr_context"><code class="xref py py-func docutils literal notranslate"><span class="pre">airr_context()</span></code></a> and then generate a bar plot by the <cite>VJ_1_v_call</cite> variable.
We use <cite>max_col</cite> to limit the plot to the 10 most abundant V-genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[45]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;v_call&quot;</span><span class="p">):</span>
    <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
        <span class="n">mdata</span><span class="p">,</span>
        <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;VJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span>
        <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">max_cols</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_90_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_90_0.png" style="width: 563px; height: 359px;" />
</div>
</div>
<p>We can pre-select groups by filtering <code class="docutils literal notranslate"><span class="pre">adata</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[46]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;v_call&quot;</span><span class="p">):</span>
    <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">group_abundance</span><span class="p">(</span>
        <span class="n">mdata</span><span class="p">[</span>
            <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;VDJ_1_v_call&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="p">[</span><span class="s2">&quot;TRBV20-1&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV7-2&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV28&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV5-1&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV7-9&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="p">:,</span>
        <span class="p">],</span>
        <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span>
        <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;VDJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_92_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_92_0.png" style="width: 550px; height: 353px;" />
</div>
</div>
<p>The exact combinations of VDJ genes can be visualized as a Sankey-plot using <a class="reference internal" href="../generated/scirpy.pl.vdj_usage.html#scirpy.pl.vdj_usage" title="scirpy.pl.vdj_usage"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pl.vdj_usage()</span></code></a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">vdj_usage</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">full_combination</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">max_segments</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">max_ribbons</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
    <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;figsize&quot;</span><span class="p">:</span> <span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">)},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_94_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_94_0.png" style="width: 559px; height: 342px;" />
</div>
</div>
<p>We can also use this plot to investigate the exact VDJ composition of one (or several) clonotypes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">vdj_usage</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">[</span><span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;68&quot;</span><span class="p">,</span> <span class="s2">&quot;101&quot;</span><span class="p">,</span> <span class="s2">&quot;127&quot;</span><span class="p">,</span> <span class="s2">&quot;161&quot;</span><span class="p">]),</span> <span class="p">:],</span>
    <span class="n">max_ribbons</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">max_segments</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: &gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_96_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_96_1.png" style="width: 485px; height: 281px;" />
</div>
</div>
<section id="Spectratype-plots">
<h3>Spectratype plots<a class="headerlink" href="#Spectratype-plots" title="Permalink to this headline"></a></h3>
<p><a class="reference internal" href="../generated/scirpy.pl.spectratype.html#scirpy.pl.spectratype" title="scirpy.pl.spectratype"><code class="xref py py-func docutils literal notranslate"><span class="pre">spectratype()</span></code></a> plots give us information about the length distribution of CDR3 regions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span> <span class="n">viztype</span><span class="o">=</span><span class="s2">&quot;bar&quot;</span><span class="p">,</span> <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span> <span class="mi">120</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Spectratype of junction_aa by gex:cluster&#39;}, xlabel=&#39;junction_aa length&#39;, ylabel=&#39;Number of cells&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_99_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_99_1.png" style="width: 629px; height: 453px;" />
</div>
</div>
<p>The same chart visualized as “ridge”-plot:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span>
    <span class="n">viztype</span><span class="o">=</span><span class="s2">&quot;curve&quot;</span><span class="p">,</span>
    <span class="n">curve_layout</span><span class="o">=</span><span class="s2">&quot;shifted&quot;</span><span class="p">,</span>
    <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span> <span class="mi">120</span><span class="p">},</span>
    <span class="n">kde_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;kde_norm&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scirpy\pl\base.py:262: UserWarning: FixedFormatter should only be used together with FixedLocator
  ax.set_yticklabels(order)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Spectratype of junction_aa by gex:cluster&#39;}, xlabel=&#39;junction_aa length&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_101_2.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_101_2.png" style="width: 478px; height: 453px;" />
</div>
</div>
<p>A spectratype-plot by gene usage. To pre-select specific genes, we can simply filter the <code class="docutils literal notranslate"><span class="pre">mdata</span></code> object before plotting.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[51]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">airr_context</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;v_call&quot;</span><span class="p">):</span>
    <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spectratype</span><span class="p">(</span>
        <span class="n">mdata</span><span class="p">[</span>
            <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;VDJ_1_v_call&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
                <span class="p">[</span><span class="s2">&quot;TRBV20-1&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV7-2&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV28&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV5-1&quot;</span><span class="p">,</span> <span class="s2">&quot;TRBV7-9&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="p">:,</span>
        <span class="p">],</span>
        <span class="n">chain</span><span class="o">=</span><span class="s2">&quot;VDJ_1&quot;</span><span class="p">,</span>
        <span class="n">color</span><span class="o">=</span><span class="s2">&quot;VDJ_1_v_call&quot;</span><span class="p">,</span>
        <span class="n">normalize</span><span class="o">=</span><span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span>
        <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span> <span class="mi">120</span><span class="p">},</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_103_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_103_0.png" style="width: 602px; height: 453px;" />
</div>
</div>
</section>
</section>
<section id="Comparing-repertoires">
<h2>Comparing repertoires<a class="headerlink" href="#Comparing-repertoires" title="Permalink to this headline"></a></h2>
<section id="Repertoire-simlarity-and-overlaps">
<h3>Repertoire simlarity and overlaps<a class="headerlink" href="#Repertoire-simlarity-and-overlaps" title="Permalink to this headline"></a></h3>
<p>Overlaps in the adaptive immune receptor repertoire of samples or sample groups enables to pinpoint important clonotype groups, as well as to provide a measure of similarity between samples.
Running Scirpy’s <a class="reference internal" href="../generated/scirpy.tl.repertoire_overlap.html#scirpy.tl.repertoire_overlap" title="scirpy.tl.repertoire_overlap"><code class="xref py py-func docutils literal notranslate"><span class="pre">repertoire_overlap()</span></code></a> tool creates a matrix featuring the abundance of clonotypes in each sample. Additionally, it also computes a (Jaccard) distance matrix of samples as well as the linkage of hierarchical clustering.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df</span><span class="p">,</span> <span class="n">dst</span><span class="p">,</span> <span class="n">lk</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">repertoire_overlap</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[52]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>clone_id</th>
      <th>0</th>
      <th>1</th>
      <th>2</th>
      <th>3</th>
      <th>4</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
      <th>9</th>
      <th>...</th>
      <th>1516</th>
      <th>1517</th>
      <th>1518</th>
      <th>1519</th>
      <th>1520</th>
      <th>1521</th>
      <th>1522</th>
      <th>1523</th>
      <th>1524</th>
      <th>1525</th>
    </tr>
    <tr>
      <th>gex:sample</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>RN2</th>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>LN1</th>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>LN2</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>3.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>LN4</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>LT6</th>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>...</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 1526 columns</p>
</div></div>
</div>
<p>The distance matrix can be shown as a heatmap, while samples are reordered based on hierarchical clustering.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">repertoire_overlap</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span>
    <span class="n">heatmap_cats</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gex:patient&quot;</span><span class="p">,</span> <span class="s2">&quot;gex:source&quot;</span><span class="p">],</span>
    <span class="n">yticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">xticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_108_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_108_0.png" style="width: 795px; height: 783px;" />
</div>
</div>
<p>A specific pair of samples can be compared on a scatterplot, where dot size corresponds to the number of clonotypes at a given coordinate.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">repertoire_overlap</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span> <span class="n">pair_to_plot</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;LN2&quot;</span><span class="p">,</span> <span class="s2">&quot;LT2&quot;</span><span class="p">],</span> <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span> <span class="mi">120</span><span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Repertoire overlap between LN2 and LT2&#39;}, xlabel=&#39;Clonotype size in LN2&#39;, ylabel=&#39;Clonotype size in LT2&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_110_2.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_110_2.png" style="width: 528px; height: 453px;" />
</div>
</div>
</section>
</section>
<section id="Integrating-gene-expression-data">
<h2>Integrating gene expression data<a class="headerlink" href="#Integrating-gene-expression-data" title="Permalink to this headline"></a></h2>
<p>Leveraging the opportunity offered by close integeration with scanpy, transcriptomics-based data can be utilized alongside immune receptor data.</p>
<section id="Clonotype-modularity">
<h3>Clonotype modularity<a class="headerlink" href="#Clonotype-modularity" title="Permalink to this headline"></a></h3>
<p>Using the <a class="reference internal" href="../glossary.html#term-Clonotype-modularity"><span class="xref std std-term">Clonotype modularity</span></a> we can identify clonotypes consisting of
cells that are transcriptionally more similar than expected by random.</p>
<p>The clonotype modularity score represents the log2 fold change of the
number of edges in the cell-cell neighborhood graph compared to
the random background model. Clonotypes (or clonotype clusters) with
a high modularity score consist of cells that have a similar molecular phenotype.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_modularity</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">target_col</span><span class="o">=</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initalizing clonotype subgraphs...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 1487/1487 [00:00&lt;00:00, 31681.58it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing background distributions...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

100%|██████████| 1000/1000 [00:01&lt;00:00, 658.71it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:clonotype_modularity&#34;]`.
Stored result in `mdata.obs[&#34;airr:clonotype_modularity_fdr&#34;]`.
</pre></div></div>
</div>
<p>We can plot the clonotype modularity on top of a umap of clonotype network plot</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:clonotype_modularity&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_115_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_115_0.png" style="width: 289px; height: 296px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;clonotype_modularity&quot;</span><span class="p">,</span>
    <span class="n">label_fontsize</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span>
    <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
    <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_116_0.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_116_0.png" style="width: 546px; height: 418px;" />
</div>
</div>
<p>We can also visualize the clonotype modularity together with the associated FDR as a sort of “one sided volcano plot”:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_modularity</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;modularity score&#39;, ylabel=&#39;-log10(FDR)&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_118_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_118_1.png" style="width: 590px; height: 300px;" />
</div>
</div>
<p>Let’s further inspect the two top scoring candidates. We can extract that information from <code class="docutils literal notranslate"><span class="pre">mdata.obs[&quot;airr:clonotype_modularity&quot;]</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">clonotypes_top_modularity</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
    <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">)[</span><span class="s2">&quot;airr:clonotype_modularity&quot;</span><span class="p">]</span>
    <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="o">.</span><span class="n">values</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">test_ad</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:cc_aa_alignment&quot;</span><span class="p">,</span>
    <span class="n">groups</span><span class="o">=</span><span class="n">clonotypes_top_modularity</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="n">cycler</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">mpl_cm</span><span class="o">.</span><span class="n">Dark2_r</span><span class="o">.</span><span class="n">colors</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_121_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_121_1.png" style="width: 339px; height: 296px;" />
</div>
</div>
<p>We observe that they are (mostly) restricted to a single cluster. By leveraging scanpy’s differential expression module, we can compare the gene expression of the cells in the two clonotypes to the rest.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[61]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Since sc.tl.rank_genes_group does not support MuData, we need to temporarily add</span>
<span class="c1"># the AIRR columns to the gene expression AnnData object</span>
<span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_context</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="p">{</span><span class="s2">&quot;clone_id&quot;</span><span class="p">:</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">]}</span>
<span class="p">)</span> <span class="k">as</span> <span class="n">tmp_ad</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span>
        <span class="n">tmp_ad</span><span class="p">,</span>
        <span class="s2">&quot;clone_id&quot;</span><span class="p">,</span>
        <span class="n">groups</span><span class="o">=</span><span class="n">clonotypes_top_modularity</span><span class="p">,</span>
        <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;rest&quot;</span><span class="p">,</span>
        <span class="n">method</span><span class="o">=</span><span class="s2">&quot;wilcoxon&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">ct</span><span class="p">,</span> <span class="n">ax</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">clonotypes_top_modularity</span><span class="p">,</span> <span class="n">axs</span><span class="p">):</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_violin</span><span class="p">(</span>
            <span class="n">tmp_ad</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="n">ct</span><span class="p">],</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strip</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ranking genes
    finished (0:00:04)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_123_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_123_1.png" style="width: 550px; height: 413px;" />
</div>
</div>
</section>
<section id="Clonotype-imbalance-among-cell-clusters">
<h3>Clonotype imbalance among cell clusters<a class="headerlink" href="#Clonotype-imbalance-among-cell-clusters" title="Permalink to this headline"></a></h3>
<p>Using cell type annotation inferred from gene expression clusters, for example, clonotypes belonging to CD8+ effector T-cells and CD8+ tissue-resident memory T cells, can be compared.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[62]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">freq</span><span class="p">,</span> <span class="n">stat</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_imbalance</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">replicate_col</span><span class="o">=</span><span class="s2">&quot;gex:sample&quot;</span><span class="p">,</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span>
    <span class="n">case_label</span><span class="o">=</span><span class="s2">&quot;CD8_Teff&quot;</span><span class="p">,</span>
    <span class="n">control_label</span><span class="o">=</span><span class="s2">&quot;CD8_Trm&quot;</span><span class="p">,</span>
    <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">top_differential_clonotypes</span> <span class="o">=</span> <span class="n">stat</span><span class="p">[</span><span class="s2">&quot;clone_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Clonotype imbalance calculation depends on repertoire overlap. We could not detect any previous runs of repertoire_overlap, so the tool is running now...
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scirpy\tl\_clonotype_imbalance.py:278: RuntimeWarning: divide by zero encountered in log2
  logfoldchange = np.log2(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scirpy\tl\_clonotype_imbalance.py:279: RuntimeWarning: divide by zero encountered in double_scalars
  (case_mean_freq + global_minimum) / (control_mean_freq + global_minimum)
</pre></div></div>
</div>
<p>Showing top clonotypes on a UMAP clearly shows that clonotype 101 is featured by CD8+ tissue-resident memory T cells, while clonotype 68 by CD8+ effector and effector memory cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[63]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;wspace&quot;</span><span class="p">:</span> <span class="mf">0.6</span><span class="p">})</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">,</span>
    <span class="n">groups</span><span class="o">=</span><span class="n">top_differential_clonotypes</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">,</span>
    <span class="c1"># increase size of highlighted dots</span>
    <span class="n">size</span><span class="o">=</span><span class="p">[</span>
        <span class="mi">80</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">top_differential_clonotypes</span> <span class="k">else</span> <span class="mi">30</span>
        <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">][</span><span class="n">mdata</span><span class="o">.</span><span class="n">mod</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">obs_names</span><span class="p">]</span>
    <span class="p">],</span>
    <span class="n">palette</span><span class="o">=</span><span class="n">cycler</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">mpl_cm</span><span class="o">.</span><span class="n">Dark2_r</span><span class="o">.</span><span class="n">colors</span><span class="p">),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_127_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_127_1.png" style="width: 839px; height: 296px;" />
</div>
</div>
</section>
<section id="Repertoire-overlap-of-cell-types">
<h3>Repertoire overlap of cell types<a class="headerlink" href="#Repertoire-overlap-of-cell-types" title="Permalink to this headline"></a></h3>
<p>Just like comparing repertoire overlap among samples, Scirpy also offers comparison between gene expression clusters or cell subpopulations. As an example, repertoire overlap of the two cell types compared above is shown.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[64]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># ir.tl.repertoire_overlap(mdata, &quot;gex:cluster&quot;)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">repertoire_overlap</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;gex:cluster&quot;</span><span class="p">,</span> <span class="n">pair_to_plot</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD8_Teff&quot;</span><span class="p">,</span> <span class="s2">&quot;CD8_Trm&quot;</span><span class="p">],</span> <span class="n">fig_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;dpi&quot;</span><span class="p">:</span> <span class="mi">120</span><span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_129_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_129_1.png" style="width: 535px; height: 453px;" />
</div>
</div>
</section>
<section id="Marker-genes-in-top-clonotypes">
<h3>Marker genes in top clonotypes<a class="headerlink" href="#Marker-genes-in-top-clonotypes" title="Permalink to this headline"></a></h3>
<p>Gene expression of cells belonging to individual clonotypes can also be compared using Scanpy. As an example, differential gene expression of two clonotypes, found to be specific to cell type clusters can also be analysed.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[65]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">ir</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">obs_context</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="p">{</span><span class="s2">&quot;clone_id&quot;</span><span class="p">:</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;airr:clone_id&quot;</span><span class="p">]}</span>
<span class="p">)</span> <span class="k">as</span> <span class="n">tmp_ad</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span>
        <span class="n">tmp_ad</span><span class="p">,</span> <span class="s2">&quot;clone_id&quot;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;101&quot;</span><span class="p">],</span> <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;68&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;wilcoxon&quot;</span>
    <span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups_violin</span><span class="p">(</span><span class="n">tmp_ad</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="s2">&quot;101&quot;</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">15</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ranking genes
    finished (0:00:00)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_131_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_131_1.png" style="width: 301px; height: 371px;" />
</div>
</div>
</section>
</section>
<section id="Query-epitope-databases">
<h2>Query epitope databases<a class="headerlink" href="#Query-epitope-databases" title="Permalink to this headline"></a></h2>
<p>We can use scirpy to query reference databases or datasets to annotate <a class="reference internal" href="../glossary.html#term-IR"><span class="xref std std-term">IRs</span></a> with certain
features, such as epitope specificity. The reference database can be any dataset in <a class="reference internal" href="../data-structure.html#data-structure"><span class="std std-ref">scirpy’s AnnData format</span></a> and you can follow the instructions in the <a class="reference internal" href="tutorial_io.html#importing-custom-formats"><span class="std std-ref">data loading tutorial</span></a>
to build a custom reference database, if it is not available from <a class="reference internal" href="../api.html#module-scirpy.datasets" title="scirpy.datasets"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scirpy.datasets</span></code></a> yet.</p>
<p>Querying reference datasets uses the same logic as <a class="reference internal" href="#define-clonotypes-tutorial"><span class="std std-ref">defining clonotypes</span></a>:</p>
<table class="colwidths-given docutils align-default" id="id5">
<caption><span class="caption-text">Analysis steps on IR data</span><a class="headerlink" href="#id5" title="Permalink to this table"></a></caption>
<colgroup>
<col style="width: 40%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>scirpy function</p></th>
<th class="head"><p>objective</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.pp.ir_dist()</span></code></a></p></td>
<td><p>Compute sequence-based distance matrices .</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.tl.ir_query.html#scirpy.tl.ir_query" title="scirpy.tl.ir_query"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.ir_query()</span></code></a></p></td>
<td><p>For each cell, identify matching entries in a reference database.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.tl.ir_query_annotate.html#scirpy.tl.ir_query_annotate" title="scirpy.tl.ir_query_annotate"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.ir_query_annotate()</span></code></a></p></td>
<td><p>Transfer annotations from reference database to <cite>adata.obs</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.tl.ir_query_annotate_df.html#scirpy.tl.ir_query_annotate_df" title="scirpy.tl.ir_query_annotate_df"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.ir_query_annotate_df()</span></code></a></p></td>
<td><p>Return a dataframe with all matching annotations.</p></td>
</tr>
</tbody>
</table>
<p>Here, we obtain the <a class="reference internal" href="../generated/scirpy.datasets.vdjdb.html#scirpy.datasets.vdjdb" title="scirpy.datasets.vdjdb"><code class="xref py py-func docutils literal notranslate"><span class="pre">VDJDB</span></code></a> and annotate epitopes based
on amino acid sequence identity. For demonstration purposes on this toy dataset we use rather lenient settings: For a match, we specify that it is enough that either of the <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ</span></a> and <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VDJ</span></a> sequences, and either of the primary or secondary receptor chains matches the database.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[66]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vdjdb</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">vdjdb</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Downloading latest version of VDJDB
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scirpy\datasets\__init__.py:173: DtypeWarning: Columns (20,29,30) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(d / &#34;vdjdb_full.txt&#34;, sep=&#34;\t&#34;)
Processing VDJDB entries: 100%|██████████| 60055/60055 [00:17&lt;00:00, 3484.20it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Converting to AnnData object
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

100%|██████████| 60055/60055 [02:27&lt;00:00, 408.03it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[67]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">vdjdb</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VJ sequences.
Computing sequence x sequence distance matrix for VDJ sequences.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[68]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;any&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;any&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing lookup tables.
Computing clonotype x clonotype distances.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 1549/1549 [00:04&lt;00:00, 326.96it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored IR distance matrix in `adata.uns[&#34;ir_query_VDJDB_aa_identity&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p><a class="reference internal" href="../generated/scirpy.tl.ir_query_annotate_df.html#scirpy.tl.ir_query_annotate_df" title="scirpy.tl.ir_query_annotate_df"><code class="xref py py-func docutils literal notranslate"><span class="pre">ir_query_annotate_df()</span></code></a> allows us to retrieve <em>all pairs cells with their of matching entries</em>. If a cell matches multiple entires from the reference database, the resulting data frame will contain multiple rows for the same cell.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate_df</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">,</span> <span class="s2">&quot;antigen.gene&quot;</span><span class="p">],</span>
<span class="p">)</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[69]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>antigen.species</th>
      <th>antigen.gene</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>RT3_TCTCTAATCACAATGC-1</th>
      <td>CMV</td>
      <td>IE1</td>
    </tr>
    <tr>
      <th>RT3_TCTCTAATCACAATGC-1</th>
      <td>CMV</td>
      <td>IE1</td>
    </tr>
    <tr>
      <th>RT3_TCTCTAATCACAATGC-1</th>
      <td>CMV</td>
      <td>IE1</td>
    </tr>
    <tr>
      <th>RT3_TCTCTAATCACAATGC-1</th>
      <td>InfluenzaA</td>
      <td>NP</td>
    </tr>
    <tr>
      <th>RT3_TCTCTAATCACAATGC-1</th>
      <td>SARS-CoV-2</td>
      <td>ORF9b</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Alternatively, to break down the annotation to a single-value per cell, you can use
<a class="reference internal" href="../generated/scirpy.tl.ir_query_annotate.html#scirpy.tl.ir_query_annotate" title="scirpy.tl.ir_query_annotate"><code class="xref py py-func docutils literal notranslate"><span class="pre">ir_query_annotate()</span></code></a>. Depending on the specified <cite>strategy</cite> it will only label unambiguous matches, or use the most frequent value.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">ir_query_annotate</span><span class="p">(</span>
    <span class="n">mdata</span><span class="p">,</span>
    <span class="n">vdjdb</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;identity&quot;</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">include_ref_cols</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;antigen.species&quot;</span><span class="p">],</span>
    <span class="n">strategy</span><span class="o">=</span><span class="s2">&quot;most-frequent&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 315/315 [00:00&lt;00:00, 2879.92it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:antigen.species&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[71]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:antigen.species&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_3k_tcr_141_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_3k_tcr_141_1.png" style="width: 405px; height: 296px;" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../generated/scirpy.io.from_airr_cells.html" class="btn btn-neutral float-left" title="scirpy.io.from_airr_cells" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Gregor Sturm, Tamas Szabo..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>