

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  

  <title>API &mdash; scirpy  documentation</title>
  


  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/scanpy.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="scirpy.io.upgrade_schema" href="generated/scirpy.io.upgrade_schema.html" />
    <link rel="prev" title="Usage principles" href="usage-principles.html" />
 


</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.7.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage-principles.html">Usage principles</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.io">Input/Output: <code class="docutils literal notranslate"><span class="pre">io</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.upgrade_schema.html">scirpy.io.upgrade_schema</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.read_h5ad.html">scirpy.io.read_h5ad</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.read_10x_vdj.html">scirpy.io.read_10x_vdj</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.read_tracer.html">scirpy.io.read_tracer</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.read_bracer.html">scirpy.io.read_bracer</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.read_airr.html">scirpy.io.read_airr</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.from_dandelion.html">scirpy.io.from_dandelion</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.write_airr.html">scirpy.io.write_airr</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.to_dandelion.html">scirpy.io.to_dandelion</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.AirrCell.html">scirpy.io.AirrCell</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.from_airr_cells.html">scirpy.io.from_airr_cells</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.io.to_airr_cells.html">scirpy.io.to_airr_cells</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.pp">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.pp.merge_with_ir.html">scirpy.pp.merge_with_ir</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.pp.merge_airr_chains.html">scirpy.pp.merge_airr_chains</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.pp.ir_dist.html">scirpy.pp.ir_dist</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tools-tl">Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#generic">Generic</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.group_abundance.html">scirpy.tl.group_abundance</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#quality-control">Quality control</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.chain_qc.html">scirpy.tl.chain_qc</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#define-and-visualize-clonotypes">Define and visualize clonotypes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.define_clonotypes.html">scirpy.tl.define_clonotypes</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.define_clonotype_clusters.html">scirpy.tl.define_clonotype_clusters</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.clonotype_convergence.html">scirpy.tl.clonotype_convergence</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.clonotype_network.html">scirpy.tl.clonotype_network</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.clonotype_network_igraph.html">scirpy.tl.clonotype_network_igraph</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#analyse-clonal-diversity">Analyse clonal diversity</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.clonal_expansion.html">scirpy.tl.clonal_expansion</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.summarize_clonal_expansion.html">scirpy.tl.summarize_clonal_expansion</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.alpha_diversity.html">scirpy.tl.alpha_diversity</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.repertoire_overlap.html">scirpy.tl.repertoire_overlap</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.clonotype_imbalance.html">scirpy.tl.clonotype_imbalance</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#v-d-j-gene-usage">V(D)J gene usage</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.tl.spectratype.html">scirpy.tl.spectratype</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.pl">Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">Generic</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.embedding.html">scirpy.pl.embedding</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#tools">Tools</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.alpha_diversity.html">scirpy.pl.alpha_diversity</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.clonal_expansion.html">scirpy.pl.clonal_expansion</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.group_abundance.html">scirpy.pl.group_abundance</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.spectratype.html">scirpy.pl.spectratype</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.vdj_usage.html">scirpy.pl.vdj_usage</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.repertoire_overlap.html">scirpy.pl.repertoire_overlap</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.clonotype_imbalance.html">scirpy.pl.clonotype_imbalance</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.clonotype_network.html">scirpy.pl.clonotype_network</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-scirpy.pl.base">Base plotting functions: <code class="docutils literal notranslate"><span class="pre">pl.base</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.base.bar.html">scirpy.pl.base.bar</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.base.line.html">scirpy.pl.base.line</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.base.barh.html">scirpy.pl.base.barh</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.base.curve.html">scirpy.pl.base.curve</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-scirpy.pl.styling">Plot styling: <code class="docutils literal notranslate"><span class="pre">pl.styling</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.styling.apply_style_to_axes.html">scirpy.pl.styling.apply_style_to_axes</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.pl.styling.style_axes.html">scirpy.pl.styling.style_axes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.datasets">Datasets: <code class="docutils literal notranslate"><span class="pre">datasets</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.datasets.wu2020.html">scirpy.datasets.wu2020</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.datasets.wu2020_3k.html">scirpy.datasets.wu2020_3k</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.datasets.maynard2020.html">scirpy.datasets.maynard2020</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.util">Utility functions: <code class="docutils literal notranslate"><span class="pre">util</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.util.graph.layout_components.html">scirpy.util.graph.layout_components</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.util.graph.layout_fr_size_aware.html">scirpy.util.graph.layout_fr_size_aware</a></li>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.util.graph.igraph_from_sparse_matrix.html">scirpy.util.graph.igraph_from_sparse_matrix</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-scirpy.ir_dist">IR distance utilities: <code class="docutils literal notranslate"><span class="pre">ir_dist</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="generated/scirpy.ir_dist.sequence_dist.html">scirpy.ir_dist.sequence_dist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-scirpy.ir_dist.metrics">distance metrics</a><ul>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.DistanceCalculator.html">scirpy.ir_dist.metrics.DistanceCalculator</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.ParallelDistanceCalculator.html">scirpy.ir_dist.metrics.ParallelDistanceCalculator</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.IdentityDistanceCalculator.html">scirpy.ir_dist.metrics.IdentityDistanceCalculator</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.LevenshteinDistanceCalculator.html">scirpy.ir_dist.metrics.LevenshteinDistanceCalculator</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.HammingDistanceCalculator.html">scirpy.ir_dist.metrics.HammingDistanceCalculator</a></li>
<li class="toctree-l4"><a class="reference internal" href="generated/scirpy.ir_dist.metrics.AlignmentDistanceCalculator.html">scirpy.ir_dist.metrics.AlignmentDistanceCalculator</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">scirpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/icbi-lab/scirpy/blob/master/docs/api.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="api">
<h1>API<a class="headerlink" href="#api" title="Permalink to this headline">¶</a></h1>
<p>Import scirpy together with scanpy as</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">ir</span>
</pre></div>
</div>
<p>For consistency, the scirpy API tries to follow the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/api/index.html">scanpy API</a>
as closely as possible.</p>
<section id="module-scirpy.io">
<span id="input-output-io"></span><span id="api-io"></span><h2>Input/Output: <code class="docutils literal notranslate"><span class="pre">io</span></code><a class="headerlink" href="#module-scirpy.io" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In scirpy v0.7.0 the way VDJ data is stored in <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> has changed to
be fully compliant with the <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html#productive">AIRR Rearrangement</a>
schema. Please use <a class="reference internal" href="generated/scirpy.io.upgrade_schema.html#scirpy.io.upgrade_schema" title="scirpy.io.upgrade_schema"><code class="xref py py-func docutils literal notranslate"><span class="pre">upgrade_schema()</span></code></a> to make <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects
from previous scirpy versions compatible with the most recent scirpy workflow.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.upgrade_schema.html#scirpy.io.upgrade_schema" title="scirpy.io.upgrade_schema"><code class="xref py py-obj docutils literal notranslate"><span class="pre">upgrade_schema</span></code></a>(adata)</p></td>
<td><p>Update older versions of a scirpy anndata object to the latest schema.</p></td>
</tr>
</tbody>
</table>
</div>
<p>The following functions allow to import <a class="reference internal" href="glossary.html#term-V-D-J"><span class="xref std std-term">V(D)J</span></a> information from various
formats.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.read_h5ad.html#scirpy.io.read_h5ad" title="scirpy.io.read_h5ad"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_h5ad</span></code></a>(filename[, backed, as_sparse, …])</p></td>
<td><p>Read <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code>-formatted hdf5 file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.io.read_10x_vdj.html#scirpy.io.read_10x_vdj" title="scirpy.io.read_10x_vdj"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_10x_vdj</span></code></a>(path[, filtered])</p></td>
<td><p>Read <a class="reference internal" href="glossary.html#term-IR"><span class="xref std std-term">IR</span></a> data from 10x Genomics cell-ranger output.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.read_tracer.html#scirpy.io.read_tracer" title="scirpy.io.read_tracer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_tracer</span></code></a>(path)</p></td>
<td><p>Read data from <a class="reference external" href="https://github.com/Teichlab/tracer">TraCeR</a> (<span id="id1">[<a class="reference internal" href="zzz_bibliography.html#id12" title="Michael J T Stubbington, Tapio Lönnberg, Valentina Proserpio, Simon Clare, Anneliese O Speak, Gordon Dougan, and Sarah A Teichmann. T cell fate and clonality inference from single-cell transcriptomes. Nat. Methods, 13(4):329–332, April 2016.">SLonnbergP+16</a>]</span>).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.io.read_bracer.html#scirpy.io.read_bracer" title="scirpy.io.read_bracer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bracer</span></code></a>(path)</p></td>
<td><p>Read data from <a class="reference external" href="https://github.com/Teichlab/bracer">BraCeR</a> (<span id="id2">[<a class="reference internal" href="zzz_bibliography.html#id22" title="Ida Lindeman, Guy Emerton, Lira Mamanova, Omri Snir, Krzysztof Polanski, Shuo-Wang Qiao, Ludvig M. Sollid, Sarah A. Teichmann, and Michael J. T. Stubbington. BraCeR: b-cell-receptor reconstruction and clonality inference from single-cell RNA-seq. Nature Methods, 15(8):563–565, July 2018. URL: https://doi.org/10.1038/s41592-018-0082-3, doi:10.1038/s41592-018-0082-3.">LEM+18</a>]</span>).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.read_airr.html#scirpy.io.read_airr" title="scirpy.io.read_airr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_airr</span></code></a>(path[, use_umi_count_col, …])</p></td>
<td><p>Read data from <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR rearrangement</a> format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.io.from_dandelion.html#scirpy.io.from_dandelion" title="scirpy.io.from_dandelion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_dandelion</span></code></a>(dandelion[, transfer])</p></td>
<td><p>Import data from <a class="reference external" href="https://github.com/zktuong/dandelion">Dandelion</a> (<span id="id3">[<a class="reference internal" href="zzz_bibliography.html#id23" title="Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, Natsuhiko Kumasaka, Katarzyna Kania, Justin Engelbert, Bayanne Olabi, Jarmila Stremenova Spegarova, Nicola K Wilson, Nicole Mende, Laura Jardine, Louis CS Gardner, Issac Goh, Dave Horsfall, Jim McGrath, Simone Webb, Michael W. Mather, Rik GH Lindeboom, Emma Dann, Ni Huang, Krzysztof Polanski, Elena Prigmore, Florian Gothe, Jonathan Scott, Rebecca P Payne, Kenneth F Baker, Aidan T Hanrath, Ina CD Schim van der Loeff, Andrew S Barr, Amada Sanchez-Gonzalez, Laura Bergamaschi, Federica Mescia, Josephine L Barnes, Eliz Kilich, Angus de Wilton, Anita Saigal, Aarash Saleh, Sam M Janes, Claire M Smith, Nusayhah Gopee, Caroline Wilson, Paul Coupland, Jonathan M Coxhead, Vladimir Y Kiselev, Stijn van Dongen, Jaume Bacardit, Hamish W King, Anthony J Rostron, A John Simpson, Sophie Hambleton, Elisa Laurenti, Paul A Lyons, Kerstin B Meyer, Marko Z Nikolic, Christopher JA Duncan, Ken Smith, Sarah A Teichmann, Menna R Clatworthy, John C Marioni, Berthold Gottgens, and Muzlifah Haniffa and. The cellular immune response to COVID-19 deciphered by single cell multi-omics across three UK centres. BioRxiv, January 2021. URL: https://doi.org/10.1101/2021.01.13.21249725, doi:10.1101/2021.01.13.21249725.">SRB+21</a>]</span>).</p></td>
</tr>
</tbody>
</table>
<p>Scirpy can export data to the following formats:</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.write_airr.html#scirpy.io.write_airr" title="scirpy.io.write_airr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_airr</span></code></a>(adata, filename)</p></td>
<td><p>Export <a class="reference internal" href="glossary.html#term-IR"><span class="xref std std-term">IR</span></a> data to <a class="reference internal" href="glossary.html#term-AIRR"><span class="xref std std-term">AIRR</span></a> Rearrangement <code class="docutils literal notranslate"><span class="pre">tsv</span></code> format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.io.to_dandelion.html#scirpy.io.to_dandelion" title="scirpy.io.to_dandelion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_dandelion</span></code></a>(adata)</p></td>
<td><p><p>Export data to <a class="reference external" href="https://github.com/zktuong/dandelion">Dandelion</a> (<span id="id5">[<a class="reference internal" href="zzz_bibliography.html#id23" title="Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, Natsuhiko Kumasaka, Katarzyna Kania, Justin Engelbert, Bayanne Olabi, Jarmila Stremenova Spegarova, Nicola K Wilson, Nicole Mende, Laura Jardine, Louis CS Gardner, Issac Goh, Dave Horsfall, Jim McGrath, Simone Webb, Michael W. Mather, Rik GH Lindeboom, Emma Dann, Ni Huang, Krzysztof Polanski, Elena Prigmore, Florian Gothe, Jonathan Scott, Rebecca P Payne, Kenneth F Baker, Aidan T Hanrath, Ina CD Schim van der Loeff, Andrew S Barr, Amada Sanchez-Gonzalez, Laura Bergamaschi, Federica Mescia, Josephine L Barnes, Eliz Kilich, Angus de Wilton, Anita Saigal, Aarash Saleh, Sam M Janes, Claire M Smith, Nusayhah Gopee, Caroline Wilson, Paul Coupland, Jonathan M Coxhead, Vladimir Y Kiselev, Stijn van Dongen, Jaume Bacardit, Hamish W King, Anthony J Rostron, A John Simpson, Sophie Hambleton, Elisa Laurenti, Paul A Lyons, Kerstin B Meyer, Marko Z Nikolic, Christopher JA Duncan, Ken Smith, Sarah A Teichmann, Menna R Clatworthy, John C Marioni, Berthold Gottgens, and Muzlifah Haniffa and. The cellular immune response to COVID-19 deciphered by single cell multi-omics across three UK centres. BioRxiv, January 2021. URL: https://doi.org/10.1101/2021.01.13.21249725, doi:10.1101/2021.01.13.21249725.">SRB+21</a>]</span>).</p>
</p></td>
</tr>
</tbody>
</table>
<p>To convert own formats into the scirpy <a class="reference internal" href="usage-principles.html#data-structure"><span class="std std-ref">Data structure</span></a>, we recommend building
a list of <a class="reference internal" href="generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects first, and then converting them into
an <a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object using <a class="reference internal" href="generated/scirpy.io.from_airr_cells.html#scirpy.io.from_airr_cells" title="scirpy.io.from_airr_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">from_airr_cells()</span></code></a>.
For more details, check the <a class="reference internal" href="tutorials/tutorial_io.html#importing-data"><span class="std std-ref">Data loading tutorial</span></a>.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AirrCell</span></code></a>(cell_id[, cell_attribute_fields, …])</p></td>
<td><p>Data structure for a Cell with immune receptors.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.io.from_airr_cells.html#scirpy.io.from_airr_cells" title="scirpy.io.from_airr_cells"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_airr_cells</span></code></a>(airr_cells[, include_fields])</p></td>
<td><p>Convert a collection of <a class="reference internal" href="generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects to <a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.io.to_airr_cells.html#scirpy.io.to_airr_cells" title="scirpy.io.to_airr_cells"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_airr_cells</span></code></a>(adata)</p></td>
<td><p>Convert an adata object with IR information back to a list of <a class="reference internal" href="generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-scirpy.pp">
<span id="preprocessing-pp"></span><h2>Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code><a class="headerlink" href="#module-scirpy.pp" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pp.merge_with_ir.html#scirpy.pp.merge_with_ir" title="scirpy.pp.merge_with_ir"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_with_ir</span></code></a>(adata, adata_ir[, on])</p></td>
<td><p>Merge adaptive immune receptor (<a class="reference internal" href="glossary.html#term-IR"><span class="xref std std-term">IR</span></a>) data with transcriptomics data into a single <a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pp.merge_airr_chains.html#scirpy.pp.merge_airr_chains" title="scirpy.pp.merge_airr_chains"><code class="xref py py-obj docutils literal notranslate"><span class="pre">merge_airr_chains</span></code></a>(adata, adata2)</p></td>
<td><p>Merge two AnnData objects with <a class="reference internal" href="glossary.html#term-IR"><span class="xref std std-term">IR</span></a> information (e.g.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pp.ir_dist.html#scirpy.pp.ir_dist" title="scirpy.pp.ir_dist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ir_dist</span></code></a>(adata, *[, metric, cutoff, …])</p></td>
<td><p>Computes a sequence-distance metric between all unique <a class="reference internal" href="glossary.html#term-Chain-locus"><span class="xref std std-term">VJ</span></a> <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequences and between all unique <a class="reference internal" href="glossary.html#term-Chain-locus"><span class="xref std std-term">VDJ</span></a> <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequences.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="tools-tl">
<h2>Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code><a class="headerlink" href="#tools-tl" title="Permalink to this headline">¶</a></h2>
<p>Tools add an interpretable annotation to the <a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object
which usually can be visualized by a corresponding plotting function.</p>
<span class="target" id="module-scirpy.tl"></span><section id="generic">
<h3>Generic<a class="headerlink" href="#generic" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.group_abundance.html#scirpy.tl.group_abundance" title="scirpy.tl.group_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">group_abundance</span></code></a>(adata, groupby[, …])</p></td>
<td><p>Summarizes the number/fraction of cells of a certain category by a certain group.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="quality-control">
<h3>Quality control<a class="headerlink" href="#quality-control" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">chain_qc</span></code></a>(adata, *[, inplace, key_added])</p></td>
<td><p>Perform quality control based on the receptor-chain pairing configuration.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="define-and-visualize-clonotypes">
<h3>Define and visualize clonotypes<a class="headerlink" href="#define-and-visualize-clonotypes" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">define_clonotypes</span></code></a>(adata, *[, key_added, …])</p></td>
<td><p>Define <a class="reference internal" href="glossary.html#term-Clonotype"><span class="xref std std-term">clonotypes</span></a> based on <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> nucleic acid sequence identity.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.tl.define_clonotype_clusters.html#scirpy.tl.define_clonotype_clusters" title="scirpy.tl.define_clonotype_clusters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">define_clonotype_clusters</span></code></a>(adata, *[, …])</p></td>
<td><p>Define <a class="reference internal" href="glossary.html#term-Clonotype-cluster"><span class="xref std std-term">clonotype clusters</span></a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.clonotype_convergence.html#scirpy.tl.clonotype_convergence" title="scirpy.tl.clonotype_convergence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_convergence</span></code></a>(adata, *, key_coarse, …)</p></td>
<td><p>Finds evidence for <a class="reference internal" href="glossary.html#term-Convergent-evolution-of-clonotypes"><span class="xref std std-term">Convergent evolution of clonotypes</span></a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.tl.clonotype_network.html#scirpy.tl.clonotype_network" title="scirpy.tl.clonotype_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_network</span></code></a>(adata, *[, sequence, …])</p></td>
<td><p>Computes the layout of the clonotype network.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.clonotype_network_igraph.html#scirpy.tl.clonotype_network_igraph" title="scirpy.tl.clonotype_network_igraph"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_network_igraph</span></code></a>(adata[, basis])</p></td>
<td><p>Get an <code class="docutils literal notranslate"><span class="pre">igraph</span></code> object representing the clonotype network.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="analyse-clonal-diversity">
<h3>Analyse clonal diversity<a class="headerlink" href="#analyse-clonal-diversity" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.clonal_expansion.html#scirpy.tl.clonal_expansion" title="scirpy.tl.clonal_expansion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonal_expansion</span></code></a>(adata, *[, target_col, …])</p></td>
<td><p>Adds a column to <code class="docutils literal notranslate"><span class="pre">obs</span></code> recording which clonotypes are expanded.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.tl.summarize_clonal_expansion.html#scirpy.tl.summarize_clonal_expansion" title="scirpy.tl.summarize_clonal_expansion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">summarize_clonal_expansion</span></code></a>(adata, groupby, *)</p></td>
<td><p>Summarizes clonal expansion by a grouping variable.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.alpha_diversity.html#scirpy.tl.alpha_diversity" title="scirpy.tl.alpha_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alpha_diversity</span></code></a>(adata, groupby, *[, …])</p></td>
<td><p>Computes the alpha diversity of clonotypes within a group.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.tl.repertoire_overlap.html#scirpy.tl.repertoire_overlap" title="scirpy.tl.repertoire_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">repertoire_overlap</span></code></a>(adata, groupby, *[, …])</p></td>
<td><p>Compute distance between cell groups based on clonotype overlap.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.clonotype_imbalance.html#scirpy.tl.clonotype_imbalance" title="scirpy.tl.clonotype_imbalance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_imbalance</span></code></a>(adata, replicate_col, …)</p></td>
<td><p>Aims to find clonotypes that are the most enriched or depleted in a category.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="v-d-j-gene-usage">
<h3>V(D)J gene usage<a class="headerlink" href="#v-d-j-gene-usage" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.tl.spectratype.html#scirpy.tl.spectratype" title="scirpy.tl.spectratype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">spectratype</span></code></a>(adata[, groupby, combine_fun, …])</p></td>
<td><p>Summarizes the distribution of <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> region lengths.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-scirpy.pl">
<span id="plotting-pl"></span><h2>Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code><a class="headerlink" href="#module-scirpy.pl" title="Permalink to this headline">¶</a></h2>
<section id="id6">
<h3>Generic<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.embedding.html#scirpy.pl.embedding" title="scirpy.pl.embedding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">embedding</span></code></a>(adata, basis, *[, color, …])</p></td>
<td><p>A customized wrapper to the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.pl.embedding.html#scanpy.pl.embedding" title="(in Scanpy v1.8.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.pl.embedding()</span></code></a> function.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="tools">
<h3>Tools<a class="headerlink" href="#tools" title="Permalink to this headline">¶</a></h3>
<p>Every of these plotting functions has a corresponding <em>tool</em> in the <a class="reference internal" href="#module-scirpy.tl" title="scirpy.tl"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scirpy.tl</span></code></a>
section. Depending on the computational load, tools are either invoked on-the-fly
when calling the plotting function or need to be precomputed and stored in
<a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.6)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> previously.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.alpha_diversity.html#scirpy.pl.alpha_diversity" title="scirpy.pl.alpha_diversity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">alpha_diversity</span></code></a>(adata, groupby, *[, …])</p></td>
<td><p>Plot the alpha diversity per group.<br />`<img src="_static/api/scirpy.pl.alpha_diversity.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.clonal_expansion.html#scirpy.pl.clonal_expansion" title="scirpy.pl.clonal_expansion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonal_expansion</span></code></a>(adata, groupby, *[, …])</p></td>
<td><p>Visualize clonal expansion.<br />`<img src="_static/api/scirpy.pl.clonal_expansion.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.group_abundance.html#scirpy.pl.group_abundance" title="scirpy.pl.group_abundance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">group_abundance</span></code></a>(adata, groupby[, …])</p></td>
<td><p>Plots the number of cells per group, split up by a categorical variable.<br />`<img src="_static/api/scirpy.pl.group_abundance.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.spectratype.html#scirpy.pl.spectratype" title="scirpy.pl.spectratype"><code class="xref py py-obj docutils literal notranslate"><span class="pre">spectratype</span></code></a>(adata[, cdr3_col, combine_fun, …])</p></td>
<td><p>Show the distribution of CDR3 region lengths.<br />`<img src="_static/api/scirpy.pl.spectratype.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.vdj_usage.html#scirpy.pl.vdj_usage" title="scirpy.pl.vdj_usage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">vdj_usage</span></code></a>(adata, *[, vdj_cols, …])</p></td>
<td><p>Creates a ribbon plot of the most abundant VDJ combinations.<br />`<img src="_static/api/scirpy.pl.vdj_usage.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.repertoire_overlap.html#scirpy.pl.repertoire_overlap" title="scirpy.pl.repertoire_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">repertoire_overlap</span></code></a>(adata, groupby, *[, …])</p></td>
<td><p>Visualizes overlap betwen a pair of samples on a scatter plot or<br />`<img src="_static/api/scirpy.pl.repertoire_overlap.svg" style="width: 300px" /></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.clonotype_imbalance.html#scirpy.pl.clonotype_imbalance" title="scirpy.pl.clonotype_imbalance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_imbalance</span></code></a>(adata, replicate_col, …)</p></td>
<td><p>Aims to find clonotypes that are the most enriched or depleted in a category.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.clonotype_network.html#scirpy.pl.clonotype_network" title="scirpy.pl.clonotype_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">clonotype_network</span></code></a>(adata, *[, color, basis, …])</p></td>
<td><p>Plot the Clonotype network.<br />`<img src="_static/api/scirpy.pl.clonotype_network.png" style="width: 300px" /></p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-scirpy.pl.base">
<span id="base-plotting-functions-pl-base"></span><h3>Base plotting functions: <code class="docutils literal notranslate"><span class="pre">pl.base</span></code><a class="headerlink" href="#module-scirpy.pl.base" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.base.bar.html#scirpy.pl.base.bar" title="scirpy.pl.base.bar"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bar</span></code></a>(data, *[, ax, stacked, style, …])</p></td>
<td><p>Basic plotting function built on top of bar plot in Pandas.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.base.line.html#scirpy.pl.base.line" title="scirpy.pl.base.line"><code class="xref py py-obj docutils literal notranslate"><span class="pre">line</span></code></a>(data, *[, ax, style, style_kws, fig_kws])</p></td>
<td><p>Basic plotting function built on top of line plot in Pandas.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.base.barh.html#scirpy.pl.base.barh" title="scirpy.pl.base.barh"><code class="xref py py-obj docutils literal notranslate"><span class="pre">barh</span></code></a>(data, *[, ax, style, style_kws, fig_kws])</p></td>
<td><p>Basic plotting function built on top of bar plot in Pandas.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.base.curve.html#scirpy.pl.base.curve" title="scirpy.pl.base.curve"><code class="xref py py-obj docutils literal notranslate"><span class="pre">curve</span></code></a>(data, *[, ax, curve_layout, shade, …])</p></td>
<td><p>Basic plotting function for drawing KDE-smoothed curves.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-scirpy.pl.styling">
<span id="plot-styling-pl-styling"></span><h3>Plot styling: <code class="docutils literal notranslate"><span class="pre">pl.styling</span></code><a class="headerlink" href="#module-scirpy.pl.styling" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.pl.styling.apply_style_to_axes.html#scirpy.pl.styling.apply_style_to_axes" title="scirpy.pl.styling.apply_style_to_axes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">apply_style_to_axes</span></code></a>(ax, style, style_kws)</p></td>
<td><p>Apply a predefined style to an axis object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.pl.styling.style_axes.html#scirpy.pl.styling.style_axes" title="scirpy.pl.styling.style_axes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">style_axes</span></code></a>(ax[, title, legend_title, xlab, …])</p></td>
<td><p>Style an axes object.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="module-scirpy.datasets">
<span id="datasets-datasets"></span><h2>Datasets: <code class="docutils literal notranslate"><span class="pre">datasets</span></code><a class="headerlink" href="#module-scirpy.datasets" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.datasets.wu2020.html#scirpy.datasets.wu2020" title="scirpy.datasets.wu2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">wu2020</span></code></a>()</p></td>
<td><p>Return the dataset from <span id="id7">[<a class="reference internal" href="zzz_bibliography.html#id4" title="Thomas D. Wu, Shravan Madireddi, Patricia E. de Almeida, Romain Banchereau, Ying-Jiun J. Chen, Avantika S. Chitre, Eugene Y. Chiang, Hina Iftikhar, William E. O'Gorman, Amelia Au-Yeung, Chikara Takahashi, Leonard D. Goldstein, Chungkee Poon, Shilpa Keerthivasan, Denise E. de Almeida Nagata, Xiangnan Du, Hyang-Mi Lee, Karl L. Banta, Sanjeev Mariathasan, Meghna Das Thakur, Mahrukh A. Huseni, Marcus Ballinger, Ivette Estay, Patrick Caplazi, Zora Modrusan, Lélia Delamarre, Ira Mellman, Richard Bourgon, and Jane L. Grogan. Peripheral t cell expansion predicts tumour infiltration and clinical response. Nature, 579(7798):274–278, February 2020. URL: https://doi.org/10.1038/s41586-020-2056-8, doi:10.1038/s41586-020-2056-8.">WMdA+20</a>]</span> as AnnData object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.datasets.wu2020_3k.html#scirpy.datasets.wu2020_3k" title="scirpy.datasets.wu2020_3k"><code class="xref py py-obj docutils literal notranslate"><span class="pre">wu2020_3k</span></code></a>()</p></td>
<td><p>Return the dataset from <span id="id8">[<a class="reference internal" href="zzz_bibliography.html#id4" title="Thomas D. Wu, Shravan Madireddi, Patricia E. de Almeida, Romain Banchereau, Ying-Jiun J. Chen, Avantika S. Chitre, Eugene Y. Chiang, Hina Iftikhar, William E. O'Gorman, Amelia Au-Yeung, Chikara Takahashi, Leonard D. Goldstein, Chungkee Poon, Shilpa Keerthivasan, Denise E. de Almeida Nagata, Xiangnan Du, Hyang-Mi Lee, Karl L. Banta, Sanjeev Mariathasan, Meghna Das Thakur, Mahrukh A. Huseni, Marcus Ballinger, Ivette Estay, Patrick Caplazi, Zora Modrusan, Lélia Delamarre, Ira Mellman, Richard Bourgon, and Jane L. Grogan. Peripheral t cell expansion predicts tumour infiltration and clinical response. Nature, 579(7798):274–278, February 2020. URL: https://doi.org/10.1038/s41586-020-2056-8, doi:10.1038/s41586-020-2056-8.">WMdA+20</a>]</span> as AnnData object, downsampled to 3000 TCR-containing cells.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.datasets.maynard2020.html#scirpy.datasets.maynard2020" title="scirpy.datasets.maynard2020"><code class="xref py py-obj docutils literal notranslate"><span class="pre">maynard2020</span></code></a>()</p></td>
<td><p>Return the dataset from <span id="id9">[<a class="reference internal" href="zzz_bibliography.html#id2" title="Ashley Maynard, Caroline E. McCoach, Julia K. Rotow, Lincoln Harris, Franziska Haderk, D. Lucas Kerr, Elizabeth A. Yu, Erin L. Schenk, Weilun Tan, Alexander Zee, Michelle Tan, Philippe Gui, Tasha Lea, Wei Wu, Anatoly Urisman, Kirk Jones, Rene Sit, Pallav K. Kolli, Eric Seeley, Yaron Gesthalter, Daniel D. Le, Kevin A. Yamauchi, David M. Naeger, Sourav Bandyopadhyay, Khyati Shah, Lauren Cech, Nicholas J. Thomas, Anshal Gupta, Mayra Gonzalez, Hien Do, Lisa Tan, Bianca Bacaltos, Rafael Gomez-Sjoberg, Matthew Gubens, Thierry Jahan, Johannes R. Kratz, David Jablons, Norma Neff, Robert C. Doebele, Jonathan Weissman, Collin M. Blakely, Spyros Darmanis, and Trever G. Bivona. Therapy-induced evolution of human lung cancer revealed by single-cell RNA sequencing. Cell, 182(5):1232–1251.e22, September 2020. URL: https://doi.org/10.1016/j.cell.2020.07.017, doi:10.1016/j.cell.2020.07.017.">MMR+20</a>]</span> as AnnData object.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-scirpy.util">
<span id="utility-functions-util"></span><h2>Utility functions: <code class="docutils literal notranslate"><span class="pre">util</span></code><a class="headerlink" href="#module-scirpy.util" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.util.graph.layout_components.html#scirpy.util.graph.layout_components" title="scirpy.util.graph.layout_components"><code class="xref py py-obj docutils literal notranslate"><span class="pre">graph.layout_components</span></code></a>(graph[, …])</p></td>
<td><p>Compute a graph layout by layouting all connected components individually.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.util.graph.layout_fr_size_aware.html#scirpy.util.graph.layout_fr_size_aware" title="scirpy.util.graph.layout_fr_size_aware"><code class="xref py py-obj docutils literal notranslate"><span class="pre">graph.layout_fr_size_aware</span></code></a>(graph, *[, k, …])</p></td>
<td><p>Compute the Fruchterman-Reingold layout respecting node sizes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.util.graph.igraph_from_sparse_matrix.html#scirpy.util.graph.igraph_from_sparse_matrix" title="scirpy.util.graph.igraph_from_sparse_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">graph.igraph_from_sparse_matrix</span></code></a>(matrix, *[, …])</p></td>
<td><p>Get an igraph object from an adjacency or distance matrix.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="module-scirpy.ir_dist">
<span id="ir-distance-utilities-ir-dist"></span><h2>IR distance utilities: <code class="docutils literal notranslate"><span class="pre">ir_dist</span></code><a class="headerlink" href="#module-scirpy.ir_dist" title="Permalink to this headline">¶</a></h2>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.sequence_dist.html#scirpy.ir_dist.sequence_dist" title="scirpy.ir_dist.sequence_dist"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_dist</span></code></a>(seqs[, seqs2, metric, cutoff, …])</p></td>
<td><p>Calculate a sequence x sequence distance matrix.</p></td>
</tr>
</tbody>
</table>
<section id="module-scirpy.ir_dist.metrics">
<span id="distance-metrics"></span><h3>distance metrics<a class="headerlink" href="#module-scirpy.ir_dist.metrics" title="Permalink to this headline">¶</a></h3>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.DistanceCalculator.html#scirpy.ir_dist.metrics.DistanceCalculator" title="scirpy.ir_dist.metrics.DistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DistanceCalculator</span></code></a>(cutoff)</p></td>
<td><p>Abstract base class for a <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a>-sequence distance calculator.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.ParallelDistanceCalculator.html#scirpy.ir_dist.metrics.ParallelDistanceCalculator" title="scirpy.ir_dist.metrics.ParallelDistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ParallelDistanceCalculator</span></code></a>(cutoff, *[, …])</p></td>
<td><p>Abstract base class for a DistanceCalculator that computes distances in parallel.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.IdentityDistanceCalculator.html#scirpy.ir_dist.metrics.IdentityDistanceCalculator" title="scirpy.ir_dist.metrics.IdentityDistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">IdentityDistanceCalculator</span></code></a>([cutoff])</p></td>
<td><p>Calculates the Identity-distance between <a class="reference internal" href="glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequences.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.LevenshteinDistanceCalculator.html#scirpy.ir_dist.metrics.LevenshteinDistanceCalculator" title="scirpy.ir_dist.metrics.LevenshteinDistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">LevenshteinDistanceCalculator</span></code></a>([cutoff])</p></td>
<td><p>Calculates the Levenshtein edit-distance between sequences.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.HammingDistanceCalculator.html#scirpy.ir_dist.metrics.HammingDistanceCalculator" title="scirpy.ir_dist.metrics.HammingDistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">HammingDistanceCalculator</span></code></a>([cutoff])</p></td>
<td><p>Calculates the Hamming distance between sequences of identical length.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="generated/scirpy.ir_dist.metrics.AlignmentDistanceCalculator.html#scirpy.ir_dist.metrics.AlignmentDistanceCalculator" title="scirpy.ir_dist.metrics.AlignmentDistanceCalculator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AlignmentDistanceCalculator</span></code></a>([cutoff, …])</p></td>
<td><p>Calculates distance between sequences based on pairwise sequence alignment.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="generated/scirpy.io.upgrade_schema.html" class="btn btn-neutral float-right" title="scirpy.io.upgrade_schema" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="usage-principles.html" class="btn btn-neutral float-left" title="Usage principles" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2021, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>