<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <title>Loading adaptive immune receptor repertoire (AIRR)-sequencing data with Scirpy &mdash; scirpy  documentation</title>

      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/scanpy.css" type="text/css" />
      <link rel="stylesheet" href="../_static/typehints.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />

  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script src="../_static/js/theme.js"></script>


    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="scirpy.io.read_10x_vdj" href="../generated/scirpy.io.read_10x_vdj.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" />
 


</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/scirpy_logo_bright.png" class="logo" alt="Logo"/>
          </a>
              <div class="version">
                0.12.2.dev185+ge9f3a68
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../data-structure.html">Data structure and usage principles</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Loading adaptive immune receptor repertoire (<span class="xref std std-term">AIRR</span>)-sequencing data with Scirpy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Loading-data-from-estabilshed-analysis-pipelines-or-AIRR-compliant-tools">Loading data from estabilshed analysis pipelines or AIRR-compliant tools</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.read_10x_vdj.html">scirpy.io.read_10x_vdj</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.read_tracer.html">scirpy.io.read_tracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.read_bracer.html">scirpy.io.read_bracer</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.read_airr.html">scirpy.io.read_airr</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.read_bd_rhapsody.html">scirpy.io.read_bd_rhapsody</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.from_dandelion.html">scirpy.io.from_dandelion</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Read-10x-data">Read 10x data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Read-Smart-seq2-data-processed-with-TraCeR">Read Smart-seq2 data processed with TraCeR</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Read-an-AIRR-compliant-rearrangement-table">Read an AIRR-compliant rearrangement table</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Creating-AnnData-objects-from-other-formats">Creating AnnData objects from other formats</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.AirrCell.html">scirpy.io.AirrCell</a></li>
<li class="toctree-l4"><a class="reference internal" href="../generated/scirpy.io.from_airr_cells.html">scirpy.io.from_airr_cells</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Combining-multiple-samples">Combining multiple samples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_3k_tcr.html">Analysis of 3k T cells from cancer</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../ir-biology.html">Immune receptor (IR) model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../glossary.html">Glossary</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer_docs.html">Developer Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../versions.html">Other versions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../zzz_bibliography.html">Bibliography</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">scirpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../tutorials.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Loading adaptive immune receptor repertoire (<span class="xref std std-term">AIRR</span>)-sequencing data with Scirpy</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/scverse/scirpy/blob/master/docs/tutorials/tutorial_io.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Loading-adaptive-immune-receptor-repertoire-(AIRR)-sequencing-data-with-Scirpy">
<span id="importing-data"></span><h1>Loading adaptive immune receptor repertoire (<a class="reference internal" href="../glossary.html#term-AIRR"><span class="xref std std-term">AIRR</span></a>)-sequencing data with Scirpy<a class="headerlink" href="#Loading-adaptive-immune-receptor-repertoire-(AIRR)-sequencing-data-with-Scirpy" title="Permalink to this headline"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This cell is for development only. Don&#39;t copy this to your notebook.</span>
<span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">autoreload</span> 2
<span class="kn">import</span> <span class="nn">anndata</span>

<span class="n">anndata</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">anndata_logger</span><span class="o">.</span><span class="n">addFilter</span><span class="p">(</span>
    <span class="k">lambda</span> <span class="n">r</span><span class="p">:</span> <span class="ow">not</span> <span class="n">r</span><span class="o">.</span><span class="n">getMessage</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;storing&quot;</span><span class="p">)</span>
    <span class="ow">and</span> <span class="n">r</span><span class="o">.</span><span class="n">getMessage</span><span class="p">()</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;as categorical.&quot;</span><span class="p">)</span>
<span class="p">)</span>

<span class="c1"># Temporarily suppress FutureWarnings</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\experimental\pytorch\_annloader.py:18: UserWarning: Сould not load pytorch.
  warnings.warn(&#34;Сould not load pytorch.&#34;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scirpy</span> <span class="k">as</span> <span class="nn">ir</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">muon</span> <span class="k">as</span> <span class="nn">mu</span>
<span class="kn">from</span> <span class="nn">glob</span> <span class="kn">import</span> <span class="n">glob</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">tarfile</span>
<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># verbosity: errors (0), warnings (1), info (2), hints (3)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<p>In this notebook, we demonstrate how single-cell <a class="reference internal" href="../glossary.html#term-AIRR"><span class="xref std std-term">AIRR</span></a>-data can be imported into
an <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object and merged with gene expression data in a <a class="reference external" href="https://mudata.readthedocs.io/en/latest/api/generated/mudata.MuData.html#mudata.MuData" title="(in mudata)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a> container for the use with Scirpy.
To learn more about how AIRR data is represented in AnnData, check out the <a class="reference internal" href="../data-structure.html#data-structure"><span class="std std-ref">data structure</span></a> section.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p><strong>The scirpy data model</strong></p>
<p>Since v0.13, there are no restrictions on the AIRR data that can be stored in the scirpy data structure, except that
each receptor chain needs to be associated with a cell. However, for all analyses, the assumptions of the
<a class="reference internal" href="../ir-biology.html#receptor-model"><span class="std std-ref">receptor model</span></a> still apply:</p>
<blockquote>
<div><ul class="simple">
<li><p>BCR and TCR chains are supported. Chain loci must be valid <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">Chain locus</span></a>,
i.e. one of <cite>TRA</cite>, <cite>TRG</cite>, <cite>IGK</cite>, or <cite>IGL</cite> (chains with a <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VJ</span></a> junction) or
<cite>TRB</cite>, <cite>TRD</cite>, or <cite>IGH</cite> (chains with a <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">VDJ</span></a> junction).</p></li>
<li><p>Each cell can contain up to two <cite>VJ</cite> and two <cite>VDJ</cite> chains (<a class="reference internal" href="../glossary.html#term-Dual-IR"><span class="xref std std-term">Dual IR</span></a>).
Excess chains are ignored (those with lowest read count/<a class="reference internal" href="../glossary.html#term-UMI"><span class="xref std std-term">UMI</span></a> count)
and cells flagged as <a class="reference internal" href="../glossary.html#term-Multichain-cell"><span class="xref std std-term">Multichain-cell</span></a>.</p></li>
<li><p>Non-productive chains are ignored. <em>CellRanger</em>, <em>TraCeR</em>, and the <em>AIRR rearrangment format</em>
flag these cells appropriately. When reading <a class="reference internal" href="#importing-custom-formats"><span class="std std-ref">custom formats</span></a>,
you need to pass the flag explicitly or filter the chains beforehand.</p></li>
</ul>
</div></blockquote>
<p>The <a class="reference internal" href="../generated/scirpy.pp.index_chains.html#scirpy.pp.index_chains" title="scirpy.pp.index_chains"><code class="xref py py-func docutils literal notranslate"><span class="pre">index_chains()</span></code></a> function chooses the appropriate chains for each cell according to this model
and stores references to those chains in <cite>adata.obsm</cite>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>AIRR quality control</strong></p>
<ul>
<li><p>After importing the data, we recommend running the <a class="reference internal" href="../generated/scirpy.tl.chain_qc.html#scirpy.tl.chain_qc" title="scirpy.tl.chain_qc"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.chain_qc()</span></code></a> function.
It will</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>identify the <a class="reference internal" href="../glossary.html#term-Receptor-type"><span class="xref std std-term">Receptor type</span></a> and <a class="reference internal" href="../glossary.html#term-Receptor-subtype"><span class="xref std std-term">Receptor subtype</span></a> and flag cells
as <cite>ambiguous</cite> that cannot unambigously be assigned to a certain receptor (sub)type, and</p></li>
<li><p>flag cells with <a class="reference internal" href="../glossary.html#term-Orphan-chain"><span class="xref std std-term">orphan chains</span></a> (i.e. cells with only a single detected cell)
and <a class="reference internal" href="../glossary.html#term-Multichain-cell"><span class="xref std std-term">multichain-cells</span></a> (i.e. cells with more than two full pairs of VJ- and VDJ-chains).</p></li>
</ol>
</div></blockquote>
</li>
<li><p>We recommend excluding multichain- and ambiguous cells as these likely represent doublets</p></li>
<li><p>Based on the <em>orphan chain</em> flags, the corresponding cells can be excluded. Alternatively,
these cells can be matched to clonotypes on a single chain only, by using the <cite>receptor_arms=”any”</cite>
parameter when running <a class="reference internal" href="../generated/scirpy.tl.define_clonotypes.html#scirpy.tl.define_clonotypes" title="scirpy.tl.define_clonotypes"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.tl.define_clonotypes()</span></code></a>.</p></li>
</ul>
</div>
<section id="Loading-data-from-estabilshed-analysis-pipelines-or-AIRR-compliant-tools">
<h2>Loading data from estabilshed analysis pipelines or AIRR-compliant tools<a class="headerlink" href="#Loading-data-from-estabilshed-analysis-pipelines-or-AIRR-compliant-tools" title="Permalink to this headline"></a></h2>
<p>We provide convenience functions to load data from <em>10x CellRanger</em>, <em>BD Rhapsody</em>, <em>TraCeR</em>, or <em>BraCeR</em> with
a single function call. Moreover, we support importing data in the community-standard
<a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR rearrangement schema</a>.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.io.read_10x_vdj.html#scirpy.io.read_10x_vdj" title="scirpy.io.read_10x_vdj"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_10x_vdj</span></code></a>(path[, filtered, include_fields])</p></td>
<td><p>Read <a class="reference internal" href="../glossary.html#term-AIRR"><span class="xref std std-term">AIRR</span></a> data from 10x Genomics cell-ranger output.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.io.read_tracer.html#scirpy.io.read_tracer" title="scirpy.io.read_tracer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_tracer</span></code></a>(path, **kwargs)</p></td>
<td><p>Read data from <a class="reference external" href="https://github.com/Teichlab/tracer">TraCeR</a> (<span id="id1">[<a class="reference internal" href="../zzz_bibliography.html#id12" title="Michael J T Stubbington, Tapio Lönnberg, Valentina Proserpio, Simon Clare, Anneliese O Speak, Gordon Dougan, and Sarah A Teichmann. T cell fate and clonality inference from single-cell transcriptomes. Nat. Methods, 13(4):329–332, April 2016.">SLonnbergP+16</a>]</span>).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.io.read_bracer.html#scirpy.io.read_bracer" title="scirpy.io.read_bracer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bracer</span></code></a>(path, **kwargs)</p></td>
<td><p>Read data from <a class="reference external" href="https://github.com/Teichlab/bracer">BraCeR</a> (<span id="id2">[<a class="reference internal" href="../zzz_bibliography.html#id22" title="Ida Lindeman, Guy Emerton, Lira Mamanova, Omri Snir, Krzysztof Polanski, Shuo-Wang Qiao, Ludvig M. Sollid, Sarah A. Teichmann, and Michael J. T. Stubbington. BraCeR: b-cell-receptor reconstruction and clonality inference from single-cell RNA-seq. Nature Methods, 15(8):563–565, July 2018. URL: https://doi.org/10.1038/s41592-018-0082-3, doi:10.1038/s41592-018-0082-3.">LEM+18</a>]</span>).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.io.read_airr.html#scirpy.io.read_airr" title="scirpy.io.read_airr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_airr</span></code></a>(path[, use_umi_count_col, …])</p></td>
<td><p>Read data from <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html">AIRR rearrangement</a> format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.io.read_bd_rhapsody.html#scirpy.io.read_bd_rhapsody" title="scirpy.io.read_bd_rhapsody"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_bd_rhapsody</span></code></a>(path, **kwargs)</p></td>
<td><p>Read <a class="reference internal" href="../glossary.html#term-IR"><span class="xref std std-term">IR</span></a> data from the BD Rhapsody Analysis Pipeline.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.io.from_dandelion.html#scirpy.io.from_dandelion" title="scirpy.io.from_dandelion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_dandelion</span></code></a>(dandelion[, transfer])</p></td>
<td><p>Import data from <a class="reference external" href="https://github.com/zktuong/dandelion">Dandelion</a> (<span id="id3">[<a class="reference internal" href="../zzz_bibliography.html#id23" title="Emily Stephenson, Gary Reynolds, Rachel A Botting, Fernando J Calero-Nieto, Michael Morgan, Zewen Kelvin Tuong, Karsten Bach, Waradon Sungnak, Kaylee B Worlock, Masahiro Yoshida, Natsuhiko Kumasaka, Katarzyna Kania, Justin Engelbert, Bayanne Olabi, Jarmila Stremenova Spegarova, Nicola K Wilson, Nicole Mende, Laura Jardine, Louis CS Gardner, Issac Goh, Dave Horsfall, Jim McGrath, Simone Webb, Michael W. Mather, Rik GH Lindeboom, Emma Dann, Ni Huang, Krzysztof Polanski, Elena Prigmore, Florian Gothe, Jonathan Scott, Rebecca P Payne, Kenneth F Baker, Aidan T Hanrath, Ina CD Schim van der Loeff, Andrew S Barr, Amada Sanchez-Gonzalez, Laura Bergamaschi, Federica Mescia, Josephine L Barnes, Eliz Kilich, Angus de Wilton, Anita Saigal, Aarash Saleh, Sam M Janes, Claire M Smith, Nusayhah Gopee, Caroline Wilson, Paul Coupland, Jonathan M Coxhead, Vladimir Y Kiselev, Stijn van Dongen, Jaume Bacardit, Hamish W King, Anthony J Rostron, A John Simpson, Sophie Hambleton, Elisa Laurenti, Paul A Lyons, Kerstin B Meyer, Marko Z Nikolic, Christopher JA Duncan, Ken Smith, Sarah A Teichmann, Menna R Clatworthy, John C Marioni, Berthold Gottgens, and Muzlifah Haniffa and. Single-cell multi-omics analysis of the immune response in COVID-19. Nature Medicine, April 2021. URL: https://www.nature.com/articles/s41591-021-01329-2, doi:10.1038/s41591-021-01329-2.">SRB+21</a>]</span>).</p></td>
</tr>
</tbody>
</table>
<section id="Read-10x-data">
<h3>Read 10x data<a class="headerlink" href="#Read-10x-data" title="Permalink to this headline"></a></h3>
<p>With <a class="reference internal" href="../generated/scirpy.io.read_10x_vdj.html#scirpy.io.read_10x_vdj" title="scirpy.io.read_10x_vdj"><code class="xref py py-func docutils literal notranslate"><span class="pre">read_10x_vdj()</span></code></a> we can load <cite>filtered_contig_annotations.csv</cite> or <cite>contig_annotations.json</cite> files as they are produced by <em>CellRanger</em>.
Here, we demonstrate how to load paired single cell transcriptomics and TCR sequencing data from COVID19 patients
from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926">GSE145926</a> (<span id="id4">[<a class="reference internal" href="../zzz_bibliography.html#id18" title="Mingfeng Liao, Yang Liu, Jing Yuan, Yanling Wen, Gang Xu, Juanjuan Zhao, Lin Cheng, Jinxiu Li, Xin Wang, Fuxiang Wang, Lei Liu, Ido Amit, Shuye Zhang, and Zheng Zhang. Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19. Nature Medicine, May 2020. URL: https://doi.org/10.1038/s41591-020-0901-9, doi:10.1038/s41591-020-0901-9.">LLY+20</a>]</span>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the TCR data</span>
<span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_10x_vdj</span><span class="p">(</span>
    <span class="s2">&quot;example_data/liao-2019-covid19/GSM4385993_C144_filtered_contig_annotations.csv.gz&quot;</span>
<span class="p">)</span>

<span class="c1"># Load the associated transcriptomics data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span>
    <span class="s2">&quot;example_data/liao-2019-covid19/GSM4339772_C144_filtered_feature_bc_matrix.h5&quot;</span>
<span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
reading example_data/liao-2019-covid19/GSM4339772_C144_filtered_feature_bc_matrix.h5
 (0:00:00)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\aligned_mapping.py:54: ExperimentalFeatureWarning: Support for Awkward Arrays is currently experimental. Behavior may change in the future. Please report any issues you may encounter!
  warnings.warn(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<p>This particular sample only has a detected TCR for a small fraction of the cells:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(136, 0)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(3716, 33539)
</pre></div></div>
</div>
<p>Next, we integrate both the TCR and the transcriptomics data into a single <a class="reference external" href="https://mudata.readthedocs.io/en/latest/api/generated/mudata.MuData.html#mudata.MuData" title="(in mudata)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MuData</span></code></a> object. By convention,
gene expression data should be stored in the <cite>gex</cite> modality, immune receptor data in the <cite>airr</cite> modality.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_tcr</span><span class="p">})</span>
</pre></div>
</div>
</div>
<p>Now, we can use TCR-related variables together with the gene expression data. Here, we visualize the cells with a detected TCR on the UMAP plot. It is reassuring that the TCRs coincide with the T-cell marker gene CD3.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing PCA
    with n_comps=50
    finished (0:00:01)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
    finished (0:00:03)
computing UMAP
    finished (0:00:07)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 136/136 [00:00&lt;00:00, 791.39it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_type&#34;]`.
Stored result in `mdata.obs[&#34;airr:receptor_subtype&#34;]`.
Stored result in `mdata.obs[&#34;airr:chain_pairing&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;wspace&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">})</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD3E&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;airr:receptor_type&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_io_13_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_io_13_1.png" style="width: 719px; height: 296px;" />
</div>
</div>
</section>
<section id="Read-Smart-seq2-data-processed-with-TraCeR">
<h3>Read Smart-seq2 data processed with TraCeR<a class="headerlink" href="#Read-Smart-seq2-data-processed-with-TraCeR" title="Permalink to this headline"></a></h3>
<p><a class="reference external" href="https://github.com/Teichlab/tracer">TraCeR</a> (<span id="id5">[<a class="reference internal" href="../zzz_bibliography.html#id12" title="Michael J T Stubbington, Tapio Lönnberg, Valentina Proserpio, Simon Clare, Anneliese O Speak, Gordon Dougan, and Sarah A Teichmann. T cell fate and clonality inference from single-cell transcriptomes. Nat. Methods, 13(4):329–332, April 2016.">SLonnbergP+16</a>]</span>) is a method commonly used
to extract TCR sequences from data generated with Smart-seq2 or other full-length single-cell sequencing protocols.</p>
<p>The <a class="reference internal" href="../generated/scirpy.io.read_tracer.html#scirpy.io.read_tracer" title="scirpy.io.read_tracer"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.io.read_tracer()</span></code></a> function obtains its TCR information from the <cite>.pkl</cite> file
in the <cite>filtered_TCR_seqs</cite> folder TraCeR generates for each cell.</p>
<p>For this example, we load the ~500 cells from triple-negative breast cancer patients from
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75688">GSE75688</a> (<span id="id6">[<a class="reference internal" href="../zzz_bibliography.html#id19" title="Woosung Chung, Hye Hyeon Eum, Hae-Ock Lee, Kyung-Min Lee, Han-Byoel Lee, Kyu-Tae Kim, Han Suk Ryu, Sangmin Kim, Jeong Eon Lee, Yeon Hee Park, Zhengyan Kan, Wonshik Han, and Woong-Yang Park. Single-cell RNA-seq enables comprehensive tumour and immune cell profiling in primary breast cancer. Nature Communications, May 2017. URL: https://doi.org/10.1038/ncomms15081, doi:10.1038/ncomms15081.">CEL+17</a>]</span>).
The raw data has been processed using the <a class="reference external" href="https://github.com/nf-core/smartseq2/">Smart-seq2 pipeline</a> from nf-core.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># extract data</span>
<span class="k">with</span> <span class="n">tarfile</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="s2">&quot;example_data/chung-park-2017.tar.bz2&quot;</span><span class="p">,</span> <span class="s2">&quot;r:bz2&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tar</span><span class="p">:</span>
    <span class="n">tar</span><span class="o">.</span><span class="n">extractall</span><span class="p">(</span><span class="s2">&quot;example_data/chung-park-2017&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>First, we load the transcriptomics data from the <code class="docutils literal notranslate"><span class="pre">counts.tsv</span></code> file:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">expr_chung</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;example_data/chung-park-2017/counts.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># anndata needs genes in columns and samples in rows</span>
<span class="n">expr_chung</span> <span class="o">=</span> <span class="n">expr_chung</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Geneid&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">expr_chung</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(563, 23438)
</pre></div></div>
</div>
<p>Next, we load the TCR data and merge it with the transcriptomics data:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_tracer</span><span class="p">(</span><span class="s2">&quot;example_data/chung-park-2017/tracer/&quot;</span><span class="p">)</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_tcr</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:00)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
    finished (0:00:00)
computing UMAP
    finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 563/563 [00:00&lt;00:00, 1796.49it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_type&#34;]`.
Stored result in `mdata.obs[&#34;airr:receptor_subtype&#34;]`.
Stored result in `mdata.obs[&#34;airr:chain_pairing&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;wspace&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">})</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD3E&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;airr:receptor_type&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_io_21_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_io_21_1.png" style="width: 724px; height: 296px;" />
</div>
</div>
</section>
<section id="Read-an-AIRR-compliant-rearrangement-table">
<h3>Read an AIRR-compliant rearrangement table<a class="headerlink" href="#Read-an-AIRR-compliant-rearrangement-table" title="Permalink to this headline"></a></h3>
<p>We generated example data using <a class="reference external" href="https://immunesim.readthedocs.io/en/latest/">immuneSIM</a> (<span id="id7">[<a class="reference internal" href="../zzz_bibliography.html#id20" title="Cédric R Weber, Rahmad Akbar, Alexander Yermanos, Milena Pavlović, Igor Snapkov, Geir K Sandve, Sai T Reddy, and Victor Greiff. immuneSIM: tunable multi-feature simulation of b- and t-cell receptor repertoires for immunoinformatics benchmarking. Bioinformatics, April 2020. URL: https://doi.org/10.1093/bioinformatics/btaa158, doi:10.1093/bioinformatics/btaa158.">WAY+20</a>]</span>).
The data consists of 100 cells and does not include transcriptomics data.</p>
<p>The rearrangement tables are often organized into separate tables per chain. Therefore, <a class="reference internal" href="../generated/scirpy.io.read_airr.html#scirpy.io.read_airr" title="scirpy.io.read_airr"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.io.read_airr()</span></code></a> supports
specifiying multiple <cite>tsv</cite> files at once. This would have the same effect as concatenating them before
the import.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_airr</span><span class="p">(</span>
    <span class="p">[</span>
        <span class="s2">&quot;example_data/immunesim_airr/immunesim_tra.tsv&quot;</span><span class="p">,</span>
        <span class="s2">&quot;example_data/immunesim_airr/immunesim_trb.tsv&quot;</span><span class="p">,</span>
    <span class="p">]</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 100/100 [00:00&lt;00:00, 426.31it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `adata.obs[&#34;receptor_type&#34;]`.
Stored result in `adata.obs[&#34;receptor_subtype&#34;]`.
Stored result in `adata.obs[&#34;chain_pairing&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p>The dataset does not come with transcriptomics data. We can, therefore, not
show the UMAP plot highlighting cells with TCRs, but we can still use scirpy
to analyse it. Below, we visualize the clonotype network
connecting cells with similar <a class="reference internal" href="../glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequences.</p>
<p><strong>Note:</strong> The cutoff of 25 was chosen for demonstration purposes on this small sample dataset. Usually a smaller cutoff
is more approriate.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">ir_dist</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mi">25</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VJ sequences.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3/3 [00:18&lt;00:00,  6.16s/it]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Computing sequence x sequence distance matrix for VDJ sequences.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 3/3 [00:18&lt;00:00,  6.17s/it]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">define_clonotype_clusters</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span>
    <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">,</span>
    <span class="n">receptor_arms</span><span class="o">=</span><span class="s2">&quot;any&quot;</span><span class="p">,</span>
    <span class="n">dual_ir</span><span class="o">=</span><span class="s2">&quot;primary_only&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">size_aware</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s2">&quot;alignment&quot;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s2">&quot;aa&quot;</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">clonotype_network</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cc_aa_alignment&quot;</span><span class="p">,</span>
    <span class="n">base_size</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">panel_size</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span>
    <span class="n">show_legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">show_size_legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Initializing lookup tables.
Computing clonotype x clonotype distances.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 100/100 [00:00&lt;00:00, 710.79it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `adata.obs[&#34;cc_aa_alignment&#34;]`.
Stored result in `adata.obs[&#34;cc_aa_alignment_size&#34;]`.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;cc_aa_alignment&#39;}&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_io_26_5.png" class="no-scaled-link" src="../_images/tutorials_tutorial_io_26_5.png" style="width: 390px; height: 398px;" />
</div>
</div>
</section>
</section>
<section id="Creating-AnnData-objects-from-other-formats">
<span id="importing-custom-formats"></span><h2>Creating AnnData objects from other formats<a class="headerlink" href="#Creating-AnnData-objects-from-other-formats" title="Permalink to this headline"></a></h2>
<p>Often, immune receptor (IR) data are just provided as a simple table listing the <a class="reference internal" href="../glossary.html#term-CDR3"><span class="xref std std-term">CDR3</span></a> sequences for each cell.
We provide a generic data structure for cells with IRs, which can then be converted into
an <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a> object.</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-obj docutils literal notranslate"><span class="pre">AirrCell</span></code></a>(cell_id[, cell_attribute_fields, …])</p></td>
<td><p>Data structure for a Cell with immune receptors.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../generated/scirpy.io.from_airr_cells.html#scirpy.io.from_airr_cells" title="scirpy.io.from_airr_cells"><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_airr_cells</span></code></a>(airr_cells[, key_added])</p></td>
<td><p>Convert a collection of <a class="reference internal" href="../generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects to <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.8.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a>.</p></td>
</tr>
</tbody>
</table>
<p>If you believe you are working with a commonly used format, consider sending a <a class="reference external" href="https://github.com/icbi-lab/scirpy/issues">feature request</a>
for a <cite>read_XXX</cite> function.</p>
<p>For this example, we again load the triple-negative breast cancer data from <span id="id8">[<a class="reference internal" href="../zzz_bibliography.html#id19" title="Woosung Chung, Hye Hyeon Eum, Hae-Ock Lee, Kyung-Min Lee, Han-Byoel Lee, Kyu-Tae Kim, Han Suk Ryu, Sangmin Kim, Jeong Eon Lee, Yeon Hee Park, Zhengyan Kan, Wonshik Han, and Woong-Yang Park. Single-cell RNA-seq enables comprehensive tumour and immune cell profiling in primary breast cancer. Nature Communications, May 2017. URL: https://doi.org/10.1038/ncomms15081, doi:10.1038/ncomms15081.">CEL+17</a>]</span>. However, this
time, we retrieve the TCR data from a separate summary table containing the TCR information
(we generated this table for the sake of the example, but it could very well have been a
supplementary file from the paper).</p>
<p>Such a table typically contains information about</p>
<blockquote>
<div><ul class="simple">
<li><p>CDR3 sequences (amino acid and/or nucleotide)</p></li>
<li><p>The <a class="reference internal" href="../glossary.html#term-Chain-locus"><span class="xref std std-term">Chain locus</span></a>, e.g. <cite>TRA</cite>, <cite>TRB</cite>, or <cite>IGH</cite>.</p></li>
<li><p>expression of the receptor chain (e.g. count, <a class="reference internal" href="../glossary.html#term-UMI"><span class="xref std std-term">UMI</span></a>, transcripts per million (TPM))</p></li>
<li><p>the <a class="reference internal" href="../glossary.html#term-V-D-J"><span class="xref std std-term">V(D)J</span></a> genes for each chain</p></li>
<li><p>information if the chain is <a class="reference internal" href="../glossary.html#term-Productive-chain"><span class="xref std std-term">productive</span></a>.</p></li>
</ul>
</div></blockquote>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_table</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
    <span class="s2">&quot;example_data/chung-park-2017/tcr_table.tsv&quot;</span><span class="p">,</span>
    <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">na_values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;None&quot;</span><span class="p">],</span>
    <span class="n">true_values</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">tcr_table</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>cell_id</th>
      <th>cdr3_alpha</th>
      <th>cdr3_nt_alpha</th>
      <th>count_alpha</th>
      <th>v_alpha</th>
      <th>j_alpha</th>
      <th>cdr3_beta</th>
      <th>cdr3_nt_beta</th>
      <th>count_beta</th>
      <th>v_beta</th>
      <th>d_beta</th>
      <th>j_beta</th>
      <th>productive_alpha</th>
      <th>productive_beta</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SRR2973278</td>
      <td>AVSDIHASGGSYIPT</td>
      <td>GCTGTTTCGGATATTCATGCATCAGGAGGAAGCTACATACCTACA</td>
      <td>9.29463</td>
      <td>TRAV21</td>
      <td>TRAJ6</td>
      <td>ASSWWQNTEAF</td>
      <td>GCCAGCAGCTGGTGGCAGAACACTGAAGCTTTC</td>
      <td>37.5984</td>
      <td>TRBV5-1</td>
      <td>NaN</td>
      <td>TRBJ1-1</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SRR2973305</td>
      <td>AVVTGANSKLT</td>
      <td>GCTGTGGTAACTGGAGCCAATAGTAAGCTGACA</td>
      <td>89.45740</td>
      <td>TRAV22</td>
      <td>TRAJ56</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SRR2973371</td>
      <td>ALKRTGNTPLV</td>
      <td>GCTCTGAAAAGAACAGGAAACACACCTCTTGTC</td>
      <td>431.96500</td>
      <td>TRAV9-2</td>
      <td>TRAJ29</td>
      <td>ASRSRDSGEPQH</td>
      <td>GCCAGCAGGAGCAGGGACAGCGGAGAGCCCCAGCAT</td>
      <td>952.0230</td>
      <td>TRBV10-2</td>
      <td>TRBD1</td>
      <td>TRBJ1-5</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SRR2973377</td>
      <td>ATDPETSGSRLT</td>
      <td>GCTACGGACCCAGAAACCAGTGGCTCTAGGTTGACC</td>
      <td>772.43600</td>
      <td>TRAV17</td>
      <td>TRAJ58</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>4</th>
      <td>SRR2973403</td>
      <td>AVRGATDSWGKFQ</td>
      <td>GCTGTGAGAGGAGCAACTGACAGCTGGGGGAAATTCCAG</td>
      <td>95.63640</td>
      <td>TRAV3</td>
      <td>TRAJ24</td>
      <td>SVQTSEYEQY</td>
      <td>AGCGTCCAGACTAGCGAGTACGAGCAGTAC</td>
      <td>205.8330</td>
      <td>TRBV29-1</td>
      <td>TRBD2</td>
      <td>TRBJ2-7</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>85</th>
      <td>SRR5023618</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ASSDSPFSSYNEQF</td>
      <td>GCCAGCAGTGACTCGCCCTTTAGCTCCTACAATGAGCAGTTC</td>
      <td>864.4550</td>
      <td>TRBV6-4</td>
      <td>NaN</td>
      <td>TRBJ2-1</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>86</th>
      <td>SRR5023621</td>
      <td>AENSGGSNYKLT</td>
      <td>GCAGAGAATAGTGGAGGTAGCAACTATAAACTGACA</td>
      <td>512.63000</td>
      <td>TRAV13-2</td>
      <td>TRAJ53</td>
      <td>ASSPDGGGGYT</td>
      <td>GCCAGCAGCCCTGATGGGGGAGGGGGCTACACC</td>
      <td>805.2010</td>
      <td>TRBV7-3</td>
      <td>TRBD2</td>
      <td>TRBJ1-2</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>87</th>
      <td>SRR5023626</td>
      <td>ALRIGSNYKLT</td>
      <td>GCTCTGAGAATCGGTAGCAACTATAAACTGACA</td>
      <td>12.51630</td>
      <td>TRAV9-2</td>
      <td>TRAJ53</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>88</th>
      <td>SRR5023633</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ASGLGQSVGGTQY</td>
      <td>GCTAGTGGCCTAGGGCAGTCGGTAGGAGGGACCCAGTAC</td>
      <td>18.4273</td>
      <td>TRBV12-5</td>
      <td>TRBD2</td>
      <td>TRBJ2-5</td>
      <td>True</td>
      <td>True</td>
    </tr>
    <tr>
      <th>89</th>
      <td>SRR5023639</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>ASSKGSLGPAGELF</td>
      <td>GCCAGCAGCAAAGGATCGCTGGGGCCCGCCGGGGAGCTGTTT</td>
      <td>905.9260</td>
      <td>TRBV21-1</td>
      <td>TRBD1</td>
      <td>TRBJ2-2</td>
      <td>True</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
<p>90 rows × 14 columns</p>
</div></div>
</div>
<p>Our task is now to dissect the table into <a class="reference internal" href="../generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects.
Each <a class="reference internal" href="../generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> can have an arbitrary number of chains. A chain is simply represented as a Python
dictionary following the <a class="reference external" href="https://docs.airr-community.org/en/latest/datarep/rearrangements.html#fields">AIRR Rearrangement Schema</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_cells</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">tcr_table</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="n">cell</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">AirrCell</span><span class="p">(</span><span class="n">cell_id</span><span class="o">=</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;cell_id&quot;</span><span class="p">])</span>
    <span class="c1"># some fields are mandatory according to the Rearrangement standard, but can be set to NULL</span>
    <span class="c1"># the `empty_chain_dict()` function provides a dictionary with all mandatory fields, but set to NULL.</span>
    <span class="n">alpha_chain</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">AirrCell</span><span class="o">.</span><span class="n">empty_chain_dict</span><span class="p">()</span>
    <span class="n">beta_chain</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">AirrCell</span><span class="o">.</span><span class="n">empty_chain_dict</span><span class="p">()</span>
    <span class="n">alpha_chain</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;TRA&quot;</span><span class="p">,</span>
            <span class="s2">&quot;junction_aa&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;cdr3_alpha&quot;</span><span class="p">],</span>
            <span class="s2">&quot;junction&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;cdr3_nt_alpha&quot;</span><span class="p">],</span>
            <span class="s2">&quot;consensus_count&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;count_alpha&quot;</span><span class="p">],</span>
            <span class="s2">&quot;v_call&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;v_alpha&quot;</span><span class="p">],</span>
            <span class="s2">&quot;j_call&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;j_alpha&quot;</span><span class="p">],</span>
            <span class="s2">&quot;productive&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;productive_alpha&quot;</span><span class="p">],</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">beta_chain</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="s2">&quot;TRB&quot;</span><span class="p">,</span>
            <span class="s2">&quot;junction_aa&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;cdr3_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;junction&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;cdr3_nt_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;consensus_count&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;count_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;v_call&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;v_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;d_call&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;d_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;j_call&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;j_beta&quot;</span><span class="p">],</span>
            <span class="s2">&quot;productive&quot;</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;productive_beta&quot;</span><span class="p">],</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">cell</span><span class="o">.</span><span class="n">add_chain</span><span class="p">(</span><span class="n">alpha_chain</span><span class="p">)</span>
    <span class="n">cell</span><span class="o">.</span><span class="n">add_chain</span><span class="p">(</span><span class="n">beta_chain</span><span class="p">)</span>
    <span class="n">tcr_cells</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now, we can convert the list of <a class="reference internal" href="../generated/scirpy.io.AirrCell.html#scirpy.io.AirrCell" title="scirpy.io.AirrCell"><code class="xref py py-class docutils literal notranslate"><span class="pre">AirrCell</span></code></a> objects using <a class="reference internal" href="../generated/scirpy.io.from_airr_cells.html#scirpy.io.from_airr_cells" title="scirpy.io.from_airr_cells"><code class="xref py py-func docutils literal notranslate"><span class="pre">scirpy.io.from_airr_cells()</span></code></a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">from_airr_cells</span><span class="p">(</span><span class="n">tcr_cells</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># We can re-use the transcriptomics data from above...</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">expr_chung</span><span class="p">)</span>
<span class="c1"># ... and merge it with the TCR data</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata</span><span class="p">,</span> <span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_tcr</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;cell_ranger&quot;</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
If you pass `n_top_genes`, all cutoffs are ignored.
extracting highly variable genes
    finished (0:00:00)
computing PCA
    on highly variable genes
    with n_comps=50
    finished (0:00:00)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
    finished (0:00:00)
computing UMAP
    finished (0:00:01)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 90/90 [00:00&lt;00:00, 287.99it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_type&#34;]`.
Stored result in `mdata.obs[&#34;airr:receptor_subtype&#34;]`.
Stored result in `mdata.obs[&#34;airr:chain_pairing&#34;]`.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;wspace&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">})</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;CD3E&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;airr:receptor_type&quot;</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_io_35_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_io_35_1.png" style="width: 719px; height: 296px;" />
</div>
</div>
</section>
<section id="Combining-multiple-samples">
<span id="combining-samples"></span><h2>Combining multiple samples<a class="headerlink" href="#Combining-multiple-samples" title="Permalink to this headline"></a></h2>
<p>It is quite common that the sequncing data is split up in multiple samples.
To combine them into a single object, we load each sample independently using one of the approaches described
in this document. Then, we combine them using <a class="reference external" href="https://anndata.readthedocs.io/en/stable/generated/anndata.concat.html#anndata.concat" title="(in anndata v0.8.0)"><code class="xref py py-func docutils literal notranslate"><span class="pre">anndata.concat()</span></code></a>.</p>
<p>MuData currently does not implement a <cite>concat</cite> function (see <a class="reference external" href="https://github.com/scverse/mudata/issues/20">scverse/mudata#20</a>). Therefore, we
need to first concatenate gene expression and AIRR data separately and create the MuData object as the last step.</p>
<p>Here is a full example loading and combining three samples from the COVID19 study by <span id="id9">[<a class="reference internal" href="../zzz_bibliography.html#id18" title="Mingfeng Liao, Yang Liu, Jing Yuan, Yanling Wen, Gang Xu, Juanjuan Zhao, Lin Cheng, Jinxiu Li, Xin Wang, Fuxiang Wang, Lei Liu, Ido Amit, Shuye Zhang, and Zheng Zhang. Single-cell landscape of bronchoalveolar immune cells in patients with COVID-19. Nature Medicine, May 2020. URL: https://doi.org/10.1038/s41591-020-0901-9, doi:10.1038/s41591-020-0901-9.">LLY+20</a>]</span>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># define sample metadata. Usually read from a file.</span>
<span class="n">samples</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;C144&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;mild&quot;</span><span class="p">},</span>
    <span class="s2">&quot;C146&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;severe&quot;</span><span class="p">},</span>
    <span class="s2">&quot;C149&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;healthy control&quot;</span><span class="p">},</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a list of AnnData objects (one for each sample)</span>
<span class="n">adatas_tcr</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">adatas_gex</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">sample_meta</span> <span class="ow">in</span> <span class="n">samples</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">gex_file</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;example_data/liao-2019-covid19/*</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">*.h5&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">tcr_file</span> <span class="o">=</span> <span class="n">glob</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;example_data/liao-2019-covid19/*</span><span class="si">{</span><span class="n">sample</span><span class="si">}</span><span class="s2">*.csv.gz&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">adata_gex</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_10x_h5</span><span class="p">(</span><span class="n">gex_file</span><span class="p">)</span>
    <span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">ir</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">read_10x_vdj</span><span class="p">(</span><span class="n">tcr_file</span><span class="p">)</span>
    <span class="c1"># concatenation only works with unique gene names</span>
    <span class="n">adata_gex</span><span class="o">.</span><span class="n">var_names_make_unique</span><span class="p">()</span>
    <span class="n">adatas_tcr</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_tcr</span>
    <span class="n">adatas_gex</span><span class="p">[</span><span class="n">sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata_gex</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
reading example_data/liao-2019-covid19\GSM4339772_C144_filtered_feature_bc_matrix.h5
 (0:00:00)
reading example_data/liao-2019-covid19\GSM4339774_C146_filtered_feature_bc_matrix.h5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 (0:00:00)
reading example_data/liao-2019-covid19\GSM4475052_C149_filtered_feature_bc_matrix.h5
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
 (0:00:00)
WARNING: Non-standard locus name: Multi
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\anndata\_core\anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates(&#34;var&#34;)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge anndata objects</span>
<span class="n">adata_gex</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">adatas_gex</span><span class="p">,</span> <span class="n">index_unique</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="n">adata_tcr</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">adatas_tcr</span><span class="p">,</span> <span class="n">index_unique</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
<span class="n">mdata</span> <span class="o">=</span> <span class="n">mu</span><span class="o">.</span><span class="n">MuData</span><span class="p">({</span><span class="s2">&quot;gex&quot;</span><span class="p">:</span> <span class="n">adata_gex</span><span class="p">,</span> <span class="s2">&quot;airr&quot;</span><span class="p">:</span> <span class="n">adata_tcr</span><span class="p">})</span>

<span class="c1"># Set global metadata on `mdata.obs`</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs_names</span><span class="o">.</span><span class="n">to_series</span><span class="p">()</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;group&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;sample&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">samples</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s2">&quot;group&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>The data is now integrated in a single object. Again, the detected TCRs coincide with <code class="docutils literal notranslate"><span class="pre">CD3E</span></code> gene expression. We clearly observe batch effects between the samples – for a meaningful downstream analysis further processing steps such as highly-variable gene filtering and batch correction are necessary.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<pre>MuData object with n_obs × n_vars = 10715 × 33539
  obs:      &#x27;sample&#x27;, &#x27;group&#x27;
  2 modalities
    gex:    10706 x 33539
    airr:   581 x 0
      obs:  &#x27;high_confidence&#x27;, &#x27;is_cell&#x27;
      obsm: &#x27;airr&#x27;</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">],</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s2">&quot;arpack&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">mdata</span><span class="p">[</span><span class="s2">&quot;gex&quot;</span><span class="p">])</span>
<span class="n">ir</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">index_chains</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
<span class="n">ir</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">chain_qc</span><span class="p">(</span><span class="n">mdata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing PCA
    with n_comps=50
    finished (0:00:07)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
    finished (0:00:20)
computing UMAP
    finished (0:00:09)
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 581/581 [00:00&lt;00:00, 826.13it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_type&#34;]`.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Stored result in `mdata.obs[&#34;airr:receptor_subtype&#34;]`.
Stored result in `mdata.obs[&#34;airr:chain_pairing&#34;]`.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax0</span><span class="p">,</span> <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;wspace&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">})</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;CD3E&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax0</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">mu</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">mdata</span><span class="p">,</span> <span class="n">basis</span><span class="o">=</span><span class="s2">&quot;gex:umap&quot;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;airr:receptor_type&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
C:\hostedtoolcache\windows\Python\3.9.13\x64\lib\site-packages\scanpy\plotting\_tools\scatterplots.py:392: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_tutorial_io_43_1.png" class="no-scaled-link" src="../_images/tutorials_tutorial_io_43_1.png" style="width: 1031px; height: 296px;" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../generated/scirpy.io.read_10x_vdj.html" class="btn btn-neutral float-right" title="scirpy.io.read_10x_vdj" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Gregor Sturm, Tamas Szabo..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>